<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<meta property="og:title" content="Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data." />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pycallingcards.readthedocs.io/en/latest/tutorials/notebooks/Brd4_bulk.html" />
<meta property="og:site_name" content="Pycallingcards" />
<meta property="og:description" content="These are brd4 bulk callingcards datasets in glioblastoma(GBM) in mouse brain of both female and male from Kfoury et al., PNAS. (2021) and could be downloaded from GEO. In this tutorial, we will ca..." />
<meta name="description" content="These are brd4 bulk callingcards datasets in glioblastoma(GBM) in mouse brain of both female and male from Kfoury et al., PNAS. (2021) and could be downloaded from GEO. In this tutorial, we will ca..." />
<link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Tutorial: SP1 bindings in Cre-driver mouse lines." href="SP1_bulk.html" /><link rel="prev" title="Tutorial: K562 HCT116 SP1 single-cell callingcards data." href="K562HCT116_SP1.html" />

    <meta name="generator" content="sphinx-6.0.0, furo 2022.09.29"/>
        <title>Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data. - pycallingcards-tools</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?digest=d81277517bee4d6b0349d71bb2661d4890b5617c" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/override.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pycallingcards-tools</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Mouse_cortex_Example.html">Tutorial: Mouse cortex single-cell callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="K562HCT116_SP1.html">Tutorial: K562 HCT116 SP1 single-cell callingcards data.</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="SP1_bulk.html">Tutorial: SP1 bindings in Cre-driver mouse lines.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mousecortex_data.html">pycallingcards.datasets.mousecortex_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_K562HCT116_data.html">pycallingcards.datasets.SP1_K562HCT116_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mouse_brd4_data.html">pycallingcards.datasets.mouse_brd4_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_Cre_data.html">pycallingcards.datasets.SP1_Cre_data</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/reading.html">Reading</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_qbed.html">pycallingcards.reading.read_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_rna.html">pycallingcards.reading.read_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.combine_qbed.html">pycallingcards.reading.combine_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_qbed.html">pycallingcards.reading.save_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_bed.html">pycallingcards.reading.save_bed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/preprocessing.html">Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.callpeaks.html">pycallingcards.preprocessing.callpeaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.annotation.html">pycallingcards.preprocessing.annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.makeAnndata.html">pycallingcards.preprocessing.makeAnndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.filter_peaks.html">pycallingcards.preprocessing.filter_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.adata_insertions.html">pycallingcards.preprocessing.adata_insertions</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/tools.html">Tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.call_motif.html">pycallingcards.tools.call_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups.html">pycallingcards.tools.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bygroup.html">pycallingcards.tools.diff2group_bygroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bysample.html">pycallingcards.tools.diff2group_bysample</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.DE_pvalue.html">pycallingcards.tools.DE_pvalue</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_bulk.html">pycallingcards.tools.pair_peak_gene_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc.html">pycallingcards.tools.pair_peak_gene_sc</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area.html">pycallingcards.plotting.draw_area</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.whole_peaks.html">pycallingcards.plotting.whole_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.calculate_signal.html">pycallingcards.plotting.calculate_signal</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_plot.html">pycallingcards.plotting.signal_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_heatmap.html">pycallingcards.plotting.signal_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.rank_peak_groups.html">pycallingcards.plotting.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.volcano_plot.html">pycallingcards.plotting.volcano_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.heatmap.html">pycallingcards.plotting.heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_bulk.html">pycallingcards.plotting.dotplot_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc.html">pycallingcards.plotting.dotplot_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.sankey.html">pycallingcards.plotting.sankey</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Tutorial:-Brd4-bindings-for-Male/Female-mouses-bulk-callingcards-data.">
<h1>Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data.<a class="headerlink" href="#Tutorial:-Brd4-bindings-for-Male/Female-mouses-bulk-callingcards-data." title="Permalink to this heading">#</a></h1>
<p>These are brd4 bulk callingcards datasets in glioblastoma(GBM) in mouse brain of both female and male from <a class="reference external" href="https://www.pnas.org/doi/10.1073/pnas.2017148118">Kfoury et al., PNAS. (2021)</a> and could be downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE156821">GEO</a>.</p>
<p>In this tutorial, we will call peaks, make annotation, do differential peak analysis, pair bindings with gene expression. There are 668525 insertions in female ccf data and 613728 insertions in male ccf data. For bulk RNA-seq data, there are 6 samples in total: 3 female samples and 3 male samples.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pycallingcards as cc
import numpy as np
import pandas as pd
import scanpy as sc
</pre></div>
</div>
</div>
<p>We start by reading ccf data. In this data, one row is one insertion and columns indecate the chromosome, start point and end point, read number, direction and borcode of each insertion. For example, the first row means one insertion is in Chromosome 1 and starts from 3478112 to 3478115. The read number is 1 with direction from 3’ to 5’. The barcode of the cell is TTTGTCCAA. We add another column as group.</p>
<p>You could use <code class="docutils literal notranslate"><span class="pre">cc.rd.read_ccf(filename)</span></code> to load your own ccf data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>F6_Brd4 = cc.datasets.mouse_brd4_data(data=&quot;Female_Brd4&quot;)
F6_Brd4[&#39;group&#39;] = &quot;F6_Brd4&quot;
F6_Brd4
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3478112</td>
      <td>3478115</td>
      <td>1</td>
      <td>+</td>
      <td>TTTGTCCAA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3514554</td>
      <td>3514557</td>
      <td>1</td>
      <td>-</td>
      <td>TCGCTGGCT</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3516833</td>
      <td>3516836</td>
      <td>2</td>
      <td>+</td>
      <td>TGTCGTGCA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3630031</td>
      <td>3630034</td>
      <td>7</td>
      <td>+</td>
      <td>TTTGTCCAA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3904411</td>
      <td>3904414</td>
      <td>1</td>
      <td>-</td>
      <td>TTTGTCCAA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>668520</th>
      <td>chrX</td>
      <td>169982399</td>
      <td>169982402</td>
      <td>1</td>
      <td>-</td>
      <td>TATACTCTA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>668521</th>
      <td>chrY</td>
      <td>42512547</td>
      <td>42512550</td>
      <td>3</td>
      <td>-</td>
      <td>TGTCGTGCA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>668522</th>
      <td>chrY</td>
      <td>90801372</td>
      <td>90801375</td>
      <td>1</td>
      <td>-</td>
      <td>TGACTTAAT</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>668523</th>
      <td>chrY</td>
      <td>90803580</td>
      <td>90803583</td>
      <td>1</td>
      <td>-</td>
      <td>TGTCGTGCA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>668524</th>
      <td>chrY</td>
      <td>90804430</td>
      <td>90804433</td>
      <td>1</td>
      <td>-</td>
      <td>TATACTCTA</td>
      <td>F6_Brd4</td>
    </tr>
  </tbody>
</table>
<p>668525 rows × 7 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>M6_Brd4 = cc.datasets.mouse_brd4_data(data=&quot;Male_Brd4&quot;)
M6_Brd4[&#39;group&#39;] = &quot;M6_Brd4&quot;
M6_Brd4
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3493276</td>
      <td>3493279</td>
      <td>1</td>
      <td>+</td>
      <td>TACTTTCCG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3493342</td>
      <td>3493345</td>
      <td>8</td>
      <td>+</td>
      <td>ATGAAAGCA</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3493644</td>
      <td>3493647</td>
      <td>3</td>
      <td>+</td>
      <td>TCTACTGCC</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3494018</td>
      <td>3494021</td>
      <td>1</td>
      <td>+</td>
      <td>TCTACTGCC</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3508902</td>
      <td>3508905</td>
      <td>1</td>
      <td>+</td>
      <td>TACTTTCCG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>613723</th>
      <td>chrY</td>
      <td>90801152</td>
      <td>90801155</td>
      <td>2</td>
      <td>-</td>
      <td>GATCTTATG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>613724</th>
      <td>chrY</td>
      <td>90803734</td>
      <td>90803737</td>
      <td>13</td>
      <td>-</td>
      <td>TACTTTCCG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>613725</th>
      <td>chrY</td>
      <td>90803734</td>
      <td>90803737</td>
      <td>9</td>
      <td>-</td>
      <td>CACGGCGGA</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>613726</th>
      <td>chrY</td>
      <td>90803734</td>
      <td>90803737</td>
      <td>17</td>
      <td>-</td>
      <td>GATCTTATG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>613727</th>
      <td>chrY</td>
      <td>90803775</td>
      <td>90803778</td>
      <td>4</td>
      <td>-</td>
      <td>ATGAAAGCA</td>
      <td>M6_Brd4</td>
    </tr>
  </tbody>
</table>
<p>613728 rows × 7 columns</p>
</div></div>
</div>
<p>Combine the two data to call peaks together.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>Brd4 = cc.rd.combine_qbed([F6_Brd4,M6_Brd4])
Brd4
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3478112</td>
      <td>3478115</td>
      <td>1</td>
      <td>+</td>
      <td>TTTGTCCAA</td>
      <td>F6_Brd4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3493276</td>
      <td>3493279</td>
      <td>1</td>
      <td>+</td>
      <td>TACTTTCCG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3493342</td>
      <td>3493345</td>
      <td>8</td>
      <td>+</td>
      <td>ATGAAAGCA</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3493644</td>
      <td>3493647</td>
      <td>3</td>
      <td>+</td>
      <td>TCTACTGCC</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3494018</td>
      <td>3494021</td>
      <td>1</td>
      <td>+</td>
      <td>TCTACTGCC</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1282248</th>
      <td>chrY</td>
      <td>90803734</td>
      <td>90803737</td>
      <td>13</td>
      <td>-</td>
      <td>TACTTTCCG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>1282249</th>
      <td>chrY</td>
      <td>90803734</td>
      <td>90803737</td>
      <td>9</td>
      <td>-</td>
      <td>CACGGCGGA</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>1282250</th>
      <td>chrY</td>
      <td>90803734</td>
      <td>90803737</td>
      <td>17</td>
      <td>-</td>
      <td>GATCTTATG</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>1282251</th>
      <td>chrY</td>
      <td>90803775</td>
      <td>90803778</td>
      <td>4</td>
      <td>-</td>
      <td>ATGAAAGCA</td>
      <td>M6_Brd4</td>
    </tr>
    <tr>
      <th>1282252</th>
      <td>chrY</td>
      <td>90804430</td>
      <td>90804433</td>
      <td>1</td>
      <td>-</td>
      <td>TATACTCTA</td>
      <td>F6_Brd4</td>
    </tr>
  </tbody>
</table>
<p>1282253 rows × 7 columns</p>
</div></div>
</div>
<p>Because insertions are discrete, we now call peaks to deduce potential binding sites. Three different methods (<em>CCcaller,ccf_tools,Blockify</em>) are available along with there different species (<em>hg38,mm10,yeast</em>).</p>
<p>In this setting, we use <strong>CCcaller</strong> in <em>mouse</em>(‘mm10’) data. <em>maxbetween</em> is the most important parameter for <em>CCcaller</em>, it would control the maximum distance bwtween two nearby insertions, in another word, that is the minimun distance between two peaks. 800-2500 is a good parameter for <em>maxbetween</em>. <em>pvalue_cutoff</em> is also an important parameter and a number below 0.05 is strongly advised.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_data = cc.pp.callpeaks(Brd4, method = &quot;CCcaller&quot;, reference = &quot;mm10&quot;, pvalue_cutoffbg = 0.1, maxbetween = 2000,
                  pvalue_cutoffTTAA = 0.001, lam_win_size = 1000000,  pseudocounts = 0.1, record = True, save = &quot;peak.bed&quot;)
peak_data
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the CCcaller method without background, [expdata, reference, pvalue_cutoff, lam_win_size, pseudocounts, minlen, extend, maxbetween, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 21/21 [00:27&lt;00:00,  1.32s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3514354</td>
      <td>3517037</td>
      <td>5</td>
      <td>24</td>
      <td>0.148659</td>
      <td>1.319915e-08</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>4196845</td>
      <td>4200095</td>
      <td>13</td>
      <td>25</td>
      <td>0.208190</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>4213840</td>
      <td>4217642</td>
      <td>5</td>
      <td>23</td>
      <td>0.199920</td>
      <td>7.473307e-08</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>4783763</td>
      <td>4787501</td>
      <td>27</td>
      <td>31</td>
      <td>1.185859</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>4806769</td>
      <td>4809050</td>
      <td>16</td>
      <td>21</td>
      <td>0.835582</td>
      <td>1.110223e-16</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>22980</th>
      <td>chrY</td>
      <td>1009770</td>
      <td>1011742</td>
      <td>17</td>
      <td>22</td>
      <td>0.478808</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>22981</th>
      <td>chrY</td>
      <td>1241882</td>
      <td>1246464</td>
      <td>23</td>
      <td>50</td>
      <td>0.870485</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>22982</th>
      <td>chrY</td>
      <td>1267935</td>
      <td>1271185</td>
      <td>5</td>
      <td>27</td>
      <td>0.520359</td>
      <td>1.768918e-05</td>
    </tr>
    <tr>
      <th>22983</th>
      <td>chrY</td>
      <td>1282449</td>
      <td>1287505</td>
      <td>25</td>
      <td>60</td>
      <td>1.035953</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>22984</th>
      <td>chrY</td>
      <td>90803380</td>
      <td>90804634</td>
      <td>6</td>
      <td>21</td>
      <td>0.151565</td>
      <td>3.193262e-10</td>
    </tr>
  </tbody>
</table>
<p>22985 rows × 7 columns</p>
</div></div>
</div>
<p>In order to choose suitable method and parameters for peak calling, please take a look at genome areas. We stongly advise to adjust the parameters for <code class="docutils literal notranslate"><span class="pre">cc.pp.callpeaks()</span></code> to call better peaks.</p>
<p>In this plot, the first part is insertions and their read counts. One dot is an insertion and the height is log(reads+1). The second part is the distribution of insertions. The third part represents reference genes and peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr1&quot;,4856929,4863861,100000,peak_data, Brd4, &quot;mm10&quot;,
                figsize = (30,6),peak_line = 2,save = False, example_length = 5000)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_11_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_11_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr2&quot;,6136575,6144043,75000, peak_data, Brd4, &quot;mm10&quot;,
                figsize = (30,5),peak_line = 1,save = False, example_length = 5000)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_12_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_12_0.png" />
</div>
</div>
<p>We could also visualize our data in <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a> directly. Notice that this link only valid for 24h, please rerun it if you want to use it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>qbed = {&quot;Brd4&quot;:Brd4}
bed = {&quot;peak&quot;:peak_data}
cc.pl.WashU_browser_url(qbed = qbed,bed = bed,genome = &#39;mm10&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=mm10&amp;hub=https://companion.epigenomegateway.org//task/df6b7b1d2f2cf4a86cfd3ed227129fe6/output//datahub.json
</pre></div></div>
</div>
<p>We advise to look at the whole chromosome and see the distribution of peaks to make sure everything is going right.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.whole_peaks(peak_data, reference = &quot;mm10&quot;, linewidth = 1)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_16_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_16_0.png" />
</div>
</div>
<p>Here, we first combine the data and then call peaks together. Although not recommended, you could also try calling peaks seperately and then merge the peaks by <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>. Below are the codes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="n">peak_data1</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">callpeaks</span><span class="p">(</span><span class="n">F6_Brd4</span><span class="p">,</span>  <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;CCcaller&quot;</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">pvalue_cutoffbg</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">maxbetween</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
                  <span class="n">pvalue_cutoffTTAA</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span> <span class="n">lam_win_size</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span>  <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">peak_data2</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">callpeaks</span><span class="p">(</span><span class="n">M6_Brd4</span><span class="p">,</span>  <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;CCcaller&quot;</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span>  <span class="n">pvalue_cutoffbg</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">maxbetween</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
                  <span class="n">pvalue_cutoffTTAA</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span> <span class="n">lam_win_size</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span>  <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">rd</span><span class="o">.</span><span class="n">combine_ccf</span><span class="p">([</span><span class="n">peak_data1</span><span class="p">,</span><span class="n">peak_data2</span><span class="p">])</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="o">.</span><span class="n">from_dataframe</span><span class="p">(</span><span class="n">peak</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">peak_data</span> <span class="o">=</span> <span class="n">peak</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chrom&quot;</span><span class="p">:</span><span class="s2">&quot;Chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span><span class="s2">&quot;End&quot;</span><span class="p">})</span>
</pre></div>
</div>
<p>In the next step, we find out the closest genes for each peak to annotate using <a class="reference external" href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>. Make sure they are all previously installed before using.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_annotation = cc.pp.annotation(peak_data, reference = &quot;mm10&quot;)
peak_annotation = cc.pp.combine_annotation(peak_data,peak_annotation)
peak_annotation
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In the bedtools method, we would use bedtools in the default path. Set bedtools path by &#39;bedtools_path&#39; if needed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
      <th>Nearest Refseq1</th>
      <th>Gene Name1</th>
      <th>Nearest Refseq2</th>
      <th>Gene Name2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3514354</td>
      <td>3517037</td>
      <td>5</td>
      <td>24</td>
      <td>0.148659</td>
      <td>1.319915e-08</td>
      <td>NM_001011874</td>
      <td>Xkr4</td>
      <td>NM_001195662</td>
      <td>Rp1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>4196845</td>
      <td>4200095</td>
      <td>13</td>
      <td>25</td>
      <td>0.208190</td>
      <td>0.000000e+00</td>
      <td>NM_001195662</td>
      <td>Rp1</td>
      <td>NM_011283</td>
      <td>Rp1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>4213840</td>
      <td>4217642</td>
      <td>5</td>
      <td>23</td>
      <td>0.199920</td>
      <td>7.473307e-08</td>
      <td>NM_001195662</td>
      <td>Rp1</td>
      <td>NM_011283</td>
      <td>Rp1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>4783763</td>
      <td>4787501</td>
      <td>27</td>
      <td>31</td>
      <td>1.185859</td>
      <td>0.000000e+00</td>
      <td>NR_033530</td>
      <td>Mrpl15</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>4806769</td>
      <td>4809050</td>
      <td>16</td>
      <td>21</td>
      <td>0.835582</td>
      <td>1.110223e-16</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
      <td>NR_033530</td>
      <td>Mrpl15</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>22980</th>
      <td>chrY</td>
      <td>1009770</td>
      <td>1011742</td>
      <td>17</td>
      <td>22</td>
      <td>0.478808</td>
      <td>0.000000e+00</td>
      <td>NM_012011</td>
      <td>Eif2s3y</td>
      <td>NR_027507</td>
      <td>Tspy-ps</td>
    </tr>
    <tr>
      <th>22981</th>
      <td>chrY</td>
      <td>1241882</td>
      <td>1246464</td>
      <td>23</td>
      <td>50</td>
      <td>0.870485</td>
      <td>0.000000e+00</td>
      <td>NM_009484</td>
      <td>Uty</td>
      <td>NM_012008</td>
      <td>Ddx3y</td>
    </tr>
    <tr>
      <th>22982</th>
      <td>chrY</td>
      <td>1267935</td>
      <td>1271185</td>
      <td>5</td>
      <td>27</td>
      <td>0.520359</td>
      <td>1.768918e-05</td>
      <td>NM_012008</td>
      <td>Ddx3y</td>
      <td>NM_009484</td>
      <td>Uty</td>
    </tr>
    <tr>
      <th>22983</th>
      <td>chrY</td>
      <td>1282449</td>
      <td>1287505</td>
      <td>25</td>
      <td>60</td>
      <td>1.035953</td>
      <td>0.000000e+00</td>
      <td>NM_012008</td>
      <td>Ddx3y</td>
      <td>NM_148943</td>
      <td>Usp9y</td>
    </tr>
    <tr>
      <th>22984</th>
      <td>chrY</td>
      <td>90803380</td>
      <td>90804634</td>
      <td>6</td>
      <td>21</td>
      <td>0.151565</td>
      <td>3.193262e-10</td>
      <td>NR_137282</td>
      <td>Erdr1</td>
      <td>NR_137283</td>
      <td>G530011O06Rik</td>
    </tr>
  </tbody>
</table>
<p>22985 rows × 11 columns</p>
</div></div>
</div>
<p>Use ccf data, peak data and barcode data to make a cell by peak <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">Anndata</a> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf = cc.pp.makeAnndata(Brd4, peak_annotation, [&quot;F6_Brd4&quot;,&quot;M6_Brd4&quot;], key = &#39;group&#39;)
adata_ccf
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 21/21 [00:11&lt;00:00,  1.88it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2 × 22985
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf.obs
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
    <tr>
      <th>Index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>F6_Brd4</th>
    </tr>
    <tr>
      <th>M6_Brd4</th>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Differential peak analysis would find out the significant binding for each group. In this example, we use <em>fisher exact test</em> to find out.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.tl.rank_peak_groups(adata_ccf,&quot;Index&quot;,method = &#39;fisher_exact&#39;,key_added = &#39;fisher_exact&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2/2 [01:58&lt;00:00, 59.13s/it]
</pre></div></div>
</div>
<p>Plot the results for differential peak analysis.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_ccf</span><span class="p">,</span><span class="s2">&quot;Index&quot;</span><span class="p">,</span><span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span><span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span><span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_ccf</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span><span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.rank_peak_groups(adata_ccf, key = &#39;fisher_exact&#39;, rankby = &#39;pvalues&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_26_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_26_0.png" />
</div>
</div>
<p>Then, we take a look at the genome for the highly differentiated peaks. The colored ones are the insertions for specific cluster and the grey ones are the total insertions for the whole dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr9&quot;,91366534,91425373,200000,peak_data,Brd4,&quot;mm10&quot;,adata = adata_ccf,
                name = &quot;F6_Brd4&quot;,key = &quot;Index&quot;,insertionkey = &quot;group&quot;, bins = 400, figsize = (30,5), peak_line = 3,color = &quot;blue&quot;)
cc.pl.draw_area(&quot;chr9&quot;,91366534,91425373,200000,peak_data, Brd4,&quot;mm10&quot;,adata = adata_ccf,
                name = &quot;M6_Brd4&quot;,key = &quot;Index&quot;, insertionkey = &quot;group&quot;,bins = 400, figsize = (30,5),peak_line = 3)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_28_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_28_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_28_1.png" src="../../_images/tutorials_notebooks_Brd4_bulk_28_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr12&quot;,56516453,56538107,200000,peak_data,Brd4,&quot;mm10&quot;,adata = adata_ccf,
                name = &quot;F6_Brd4&quot;,key = &quot;Index&quot;,insertionkey = &quot;group&quot;, figsize = (30,6),peak_line = 3,color = &quot;blue&quot;)
cc.pl.draw_area(&quot;chr12&quot;,56516453,56538107,200000,peak_data, Brd4,&quot;mm10&quot;,adata = adata_ccf,
                name = &quot;M6_Brd4&quot;,key = &quot;Index&quot;, insertionkey = &quot;group&quot;, figsize = (30,6),peak_line = 3)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_29_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_29_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_29_1.png" src="../../_images/tutorials_notebooks_Brd4_bulk_29_1.png" />
</div>
</div>
<p>Plot the volcano plot for differential binding sites. labelright and labellest are the location points of the labels.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.volcano_plot(adata_ccf,figsize = (6,10),labelright = (5,220),labelleft = (-9,220))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_31_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_31_0.png" />
</div>
</div>
<p>Read bulk RNA-seq data. Note that this RNA-seq data is normalized by <a class="reference external" href="https://wiki.arrayserver.com/wiki/index.php?title=RPKM">RPKM</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna =  cc.datasets.mouse_brd4_data(data=&quot;RNA&quot;)
rna
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample.f6_dmso_3</th>
      <th>sample.f6_dmso_2</th>
      <th>sample.f6_dmso_1</th>
      <th>sample.m6_dmso_3</th>
      <th>sample.m6_dmso_2</th>
      <th>sample.m6_dmso_1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>mt-Tf</th>
      <td>0.000000</td>
      <td>0.639272</td>
      <td>0.000000</td>
      <td>1.156308</td>
      <td>0.000000</td>
      <td>0.687531</td>
    </tr>
    <tr>
      <th>mt-Rnr1</th>
      <td>127.907281</td>
      <td>184.396768</td>
      <td>172.544723</td>
      <td>130.417054</td>
      <td>192.566581</td>
      <td>114.212201</td>
    </tr>
    <tr>
      <th>mt-Tv</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.761150</td>
      <td>0.000000</td>
      <td>0.510777</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>mt-Rnr2</th>
      <td>247.267306</td>
      <td>527.855824</td>
      <td>324.345998</td>
      <td>675.031187</td>
      <td>684.376279</td>
      <td>418.345603</td>
    </tr>
    <tr>
      <th>mt-Tl1</th>
      <td>44.418710</td>
      <td>117.660098</td>
      <td>70.481211</td>
      <td>167.217607</td>
      <td>115.129081</td>
      <td>127.789076</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Gm16897</th>
      <td>0.281553</td>
      <td>0.349020</td>
      <td>0.463442</td>
      <td>0.078913</td>
      <td>0.070742</td>
      <td>0.112610</td>
    </tr>
    <tr>
      <th>A330023F24Rik</th>
      <td>0.118678</td>
      <td>0.055749</td>
      <td>0.058441</td>
      <td>0.029411</td>
      <td>0.018833</td>
      <td>0.049965</td>
    </tr>
    <tr>
      <th>Cd46</th>
      <td>0.127930</td>
      <td>0.095151</td>
      <td>0.075997</td>
      <td>0.061467</td>
      <td>0.055103</td>
      <td>0.087715</td>
    </tr>
    <tr>
      <th>Cr1l</th>
      <td>9.993119</td>
      <td>14.027801</td>
      <td>13.717564</td>
      <td>13.586101</td>
      <td>13.240219</td>
      <td>11.089902</td>
    </tr>
    <tr>
      <th>Cr2</th>
      <td>0.018832</td>
      <td>0.056028</td>
      <td>0.019578</td>
      <td>0.012668</td>
      <td>0.017034</td>
      <td>0.015064</td>
    </tr>
  </tbody>
</table>
<p>21430 rows × 6 columns</p>
</div></div>
</div>
<p>Plot heatmap for calling cards data and gene expression data.</p>
<p>The top lines are the relative binding expression for the two groups. The buttom lines are the gene expression of the top two nearest genes separately.</p>
<p>We could see a relationship between bindings and gene expression. Bindings tend to have possitive effects for gene expression and have stronger effects for the nearest one than the second nearest one.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.heatmap(adata_ccf,rna, rnalabels = [&quot;Female1&quot;, &quot;Female2&quot;, &quot;Female3&quot;,&quot;Male1&quot;, &quot;Male2&quot;,&quot;Male3&quot;])
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Please make sure that the samples in adata_ccf and rna are in the same order.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/scratch/rmlab/rmlab_shared/g.juanru/pycallingcards/pycallingcards/plotting/_heatmap_ccfrna.py:107: RuntimeWarning: invalid value encountered in divide
  data[groupnumber:secondnum,:]/data[groupnumber:secondnum,:].max(axis = 0).reshape((1,size)),
/scratch/rmlab/rmlab_shared/g.juanru/pycallingcards/pycallingcards/plotting/_heatmap_ccfrna.py:108: RuntimeWarning: invalid value encountered in divide
  data[secondnum:,:]/data[secondnum:,:].max(axis = 0).reshape((1,size))), axis = 0)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_35_2.png" src="../../_images/tutorials_notebooks_Brd4_bulk_35_2.png" />
</div>
</div>
<p>Find out the co-differential peak-gene pairs. It would look into all the differential peaks and then see if the annotated genes are also significant expressed. We could set the pvalue and score/log foldchange cutoff easily.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>result = cc.tl.pair_peak_gene_bulk(adata_ccf,&quot;https://github.com/The-Mitra-Lab/pycallingcards_data/releases/download/data/deseq_MF.csv&quot;,name_ccf = &#39;logfoldchanges&#39;,
                                   pvalue_cutoff_ccf = 0.001, pvalue_cutoff_rna = 0.001, lfc_cutoff_rna = 5, lfc_cutoff_ccf = 5)
result
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Peak</th>
      <th>logfoldchanges_peak</th>
      <th>Pvalue_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_peak</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1_72823300_72830641</td>
      <td>5.636525</td>
      <td>2.131181e-08</td>
      <td>Igfbp2</td>
      <td>-5.658549</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1_72823300_72830641</td>
      <td>5.636525</td>
      <td>2.131181e-08</td>
      <td>Igfbp5</td>
      <td>-7.779615</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1_174913218_174921560</td>
      <td>6.044294</td>
      <td>5.270706e-11</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr2_93615717_93619375</td>
      <td>5.141725</td>
      <td>4.589712e-06</td>
      <td>Alx4</td>
      <td>-10.292504</td>
      <td>5.458739e-18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr2_93642718_93648163</td>
      <td>5.065991</td>
      <td>9.012038e-06</td>
      <td>Alx4</td>
      <td>-10.292504</td>
      <td>5.458739e-18</td>
    </tr>
    <tr>
      <th>5</th>
      <td>chr2_147695485_147699121</td>
      <td>5.213681</td>
      <td>2.339309e-06</td>
      <td>Pax1</td>
      <td>-5.196625</td>
      <td>8.986249e-04</td>
    </tr>
    <tr>
      <th>6</th>
      <td>chr2_167232402_167237320</td>
      <td>5.410235</td>
      <td>3.111009e-07</td>
      <td>Ptgis</td>
      <td>-5.078295</td>
      <td>7.799586e-104</td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr2_167232402_167237320</td>
      <td>5.410235</td>
      <td>3.111009e-07</td>
      <td>Kcnb1</td>
      <td>-5.613772</td>
      <td>2.285981e-05</td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr3_33120861_33127487</td>
      <td>5.065991</td>
      <td>9.012038e-06</td>
      <td>Pex5l</td>
      <td>-7.590236</td>
      <td>2.351122e-24</td>
    </tr>
    <tr>
      <th>9</th>
      <td>chr3_33139996_33142829</td>
      <td>5.347646</td>
      <td>6.090580e-07</td>
      <td>Pex5l</td>
      <td>-7.590236</td>
      <td>2.351122e-24</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr3_132085670_132108436</td>
      <td>5.380378</td>
      <td>6.624415e-129</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
    </tr>
    <tr>
      <th>11</th>
      <td>chr4_6811845_6813820</td>
      <td>5.347646</td>
      <td>6.090580e-07</td>
      <td>Tox</td>
      <td>-5.641062</td>
      <td>2.178520e-12</td>
    </tr>
    <tr>
      <th>12</th>
      <td>chr4_6860745_6867727</td>
      <td>5.410235</td>
      <td>3.111009e-07</td>
      <td>Tox</td>
      <td>-5.641062</td>
      <td>2.178520e-12</td>
    </tr>
    <tr>
      <th>13</th>
      <td>chr4_68545354_68551370</td>
      <td>7.202023</td>
      <td>8.393449e-24</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
    </tr>
    <tr>
      <th>14</th>
      <td>chr5_13001870_13004057</td>
      <td>6.121146</td>
      <td>1.394584e-11</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
    </tr>
    <tr>
      <th>15</th>
      <td>chr5_13001870_13004057</td>
      <td>6.121146</td>
      <td>1.394584e-11</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
    </tr>
    <tr>
      <th>16</th>
      <td>chr5_13124523_13131816</td>
      <td>5.863196</td>
      <td>5.914808e-26</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr5_13124523_13131816</td>
      <td>5.863196</td>
      <td>5.914808e-26</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
    </tr>
    <tr>
      <th>18</th>
      <td>chr5_13276480_13283561</td>
      <td>7.368891</td>
      <td>1.015699e-26</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
    </tr>
    <tr>
      <th>19</th>
      <td>chr5_13276480_13283561</td>
      <td>7.368891</td>
      <td>1.015699e-26</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
    </tr>
    <tr>
      <th>20</th>
      <td>chr5_16764617_16770523</td>
      <td>5.527812</td>
      <td>8.132726e-08</td>
      <td>Hgf</td>
      <td>-7.513756</td>
      <td>2.543270e-09</td>
    </tr>
    <tr>
      <th>21</th>
      <td>chr5_16764617_16770523</td>
      <td>5.527812</td>
      <td>8.132726e-08</td>
      <td>Hgf</td>
      <td>-7.513756</td>
      <td>2.543270e-09</td>
    </tr>
    <tr>
      <th>22</th>
      <td>chr5_17080322_17085124</td>
      <td>5.832087</td>
      <td>1.472838e-09</td>
      <td>Hgf</td>
      <td>-7.513756</td>
      <td>2.543270e-09</td>
    </tr>
    <tr>
      <th>23</th>
      <td>chr6_22863927_22871301</td>
      <td>5.213681</td>
      <td>2.339309e-06</td>
      <td>Ptprz1</td>
      <td>-10.218818</td>
      <td>2.926910e-23</td>
    </tr>
    <tr>
      <th>24</th>
      <td>chr6_22863927_22871301</td>
      <td>5.213681</td>
      <td>2.339309e-06</td>
      <td>Aass</td>
      <td>-7.706441</td>
      <td>9.787575e-10</td>
    </tr>
    <tr>
      <th>25</th>
      <td>chr6_23077160_23084853</td>
      <td>5.141725</td>
      <td>4.589712e-06</td>
      <td>Aass</td>
      <td>-7.706441</td>
      <td>9.787575e-10</td>
    </tr>
    <tr>
      <th>26</th>
      <td>chr6_23077160_23084853</td>
      <td>5.141725</td>
      <td>4.589712e-06</td>
      <td>Ptprz1</td>
      <td>-10.218818</td>
      <td>2.926910e-23</td>
    </tr>
    <tr>
      <th>27</th>
      <td>chr7_18803120_18806567</td>
      <td>5.282218</td>
      <td>1.193209e-06</td>
      <td>Psg17</td>
      <td>-6.192943</td>
      <td>7.226941e-06</td>
    </tr>
    <tr>
      <th>28</th>
      <td>chr7_55291506_55293906</td>
      <td>5.527812</td>
      <td>8.132726e-08</td>
      <td>Luzp2</td>
      <td>-7.732798</td>
      <td>2.515372e-10</td>
    </tr>
    <tr>
      <th>29</th>
      <td>chr7_56523379_56528437</td>
      <td>6.673896</td>
      <td>9.230699e-17</td>
      <td>Gabrg3</td>
      <td>-5.960560</td>
      <td>2.341610e-05</td>
    </tr>
    <tr>
      <th>30</th>
      <td>chr8_82255689_82262262</td>
      <td>5.065991</td>
      <td>9.012038e-06</td>
      <td>Il15</td>
      <td>-5.185933</td>
      <td>8.096707e-04</td>
    </tr>
    <tr>
      <th>31</th>
      <td>chr8_82329337_82331285</td>
      <td>5.282218</td>
      <td>1.193209e-06</td>
      <td>Il15</td>
      <td>-5.185933</td>
      <td>8.096707e-04</td>
    </tr>
    <tr>
      <th>32</th>
      <td>chr8_102778665_102784309</td>
      <td>6.121146</td>
      <td>1.394584e-11</td>
      <td>Cdh11</td>
      <td>-6.729272</td>
      <td>1.932941e-232</td>
    </tr>
    <tr>
      <th>33</th>
      <td>chr10_57783900_57788071</td>
      <td>5.737618</td>
      <td>5.597023e-09</td>
      <td>Fabp7</td>
      <td>-6.577521</td>
      <td>5.567989e-17</td>
    </tr>
    <tr>
      <th>34</th>
      <td>chr10_57792844_57795824</td>
      <td>5.347646</td>
      <td>6.090580e-07</td>
      <td>Fabp7</td>
      <td>-6.577521</td>
      <td>5.567989e-17</td>
    </tr>
    <tr>
      <th>35</th>
      <td>chr10_103023599_103031365</td>
      <td>5.410235</td>
      <td>3.111009e-07</td>
      <td>Alx1</td>
      <td>-6.624259</td>
      <td>3.878401e-22</td>
    </tr>
    <tr>
      <th>36</th>
      <td>chr11_46057069_46059464</td>
      <td>5.785626</td>
      <td>2.870476e-09</td>
      <td>Adam19</td>
      <td>-5.523519</td>
      <td>1.429582e-165</td>
    </tr>
    <tr>
      <th>37</th>
      <td>chr11_46079112_46085758</td>
      <td>5.213681</td>
      <td>2.339309e-06</td>
      <td>Adam19</td>
      <td>-5.523519</td>
      <td>1.429582e-165</td>
    </tr>
    <tr>
      <th>38</th>
      <td>chr11_72279273_72287600</td>
      <td>5.141725</td>
      <td>4.589712e-06</td>
      <td>Slc13a5</td>
      <td>-5.455965</td>
      <td>1.073016e-06</td>
    </tr>
    <tr>
      <th>39</th>
      <td>chr11_96341059_96343217</td>
      <td>5.065991</td>
      <td>9.012038e-06</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
    </tr>
    <tr>
      <th>40</th>
      <td>chr11_96341059_96343217</td>
      <td>5.065991</td>
      <td>9.012038e-06</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
    </tr>
    <tr>
      <th>41</th>
      <td>chr12_56507583_56514677</td>
      <td>-8.854576</td>
      <td>2.799289e-76</td>
      <td>Nkx2-1</td>
      <td>9.134857</td>
      <td>1.061953e-60</td>
    </tr>
    <tr>
      <th>42</th>
      <td>chr14_88460574_88466755</td>
      <td>5.785626</td>
      <td>2.870476e-09</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
    </tr>
    <tr>
      <th>43</th>
      <td>chr14_88898126_88902522</td>
      <td>-6.101044</td>
      <td>8.655851e-12</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
    </tr>
    <tr>
      <th>44</th>
      <td>chr15_11743635_11745457</td>
      <td>6.483268</td>
      <td>9.473693e-15</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>45</th>
      <td>chr15_11743635_11745457</td>
      <td>6.483268</td>
      <td>9.473693e-15</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>46</th>
      <td>chr15_11781285_11785784</td>
      <td>6.329821</td>
      <td>2.600647e-13</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>47</th>
      <td>chr15_11781285_11785784</td>
      <td>6.329821</td>
      <td>2.600647e-13</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>48</th>
      <td>chr15_11823522_11836910</td>
      <td>5.994369</td>
      <td>1.648918e-28</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>49</th>
      <td>chr15_11823522_11836910</td>
      <td>5.994369</td>
      <td>1.648918e-28</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>50</th>
      <td>chr15_11898885_11914841</td>
      <td>5.095947</td>
      <td>8.454135e-15</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
    </tr>
    <tr>
      <th>51</th>
      <td>chr18_53465634_53470566</td>
      <td>5.282218</td>
      <td>1.193209e-06</td>
      <td>Prdm6</td>
      <td>-9.029699</td>
      <td>4.482978e-14</td>
    </tr>
    <tr>
      <th>52</th>
      <td>chr19_59158212_59161670</td>
      <td>-6.180078</td>
      <td>2.019017e-12</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
    </tr>
    <tr>
      <th>53</th>
      <td>chrY_1009770_1011742</td>
      <td>-5.081308</td>
      <td>4.234086e-06</td>
      <td>Eif2s3y</td>
      <td>11.645099</td>
      <td>9.414457e-23</td>
    </tr>
    <tr>
      <th>54</th>
      <td>chrY_1241882_1246464</td>
      <td>-5.506247</td>
      <td>5.373930e-08</td>
      <td>Uty</td>
      <td>8.369064</td>
      <td>1.224906e-22</td>
    </tr>
    <tr>
      <th>55</th>
      <td>chrY_1241882_1246464</td>
      <td>-5.506247</td>
      <td>5.373930e-08</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
    </tr>
    <tr>
      <th>56</th>
      <td>chrY_1282449_1287505</td>
      <td>-5.624000</td>
      <td>1.253524e-08</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Plot the above results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.dotplot_bulk(adata_ccf,rna,selected_list =np.unique(np.array(result[&quot;Peak&quot;])), num_list = [3,3], cmap = &#39;plasma&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_39_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_39_0.png" />
</div>
</div>
<p>We could see from the results that all bindings are differential and at least one of the nearest genes is differeintial. For most of the situations, bindings have a positive influence on gene expression. We could make many hypotheses to test.</p>
<p>Save the file if needed.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf.write(&quot;Brd4_bindings_bulk.h5ad&quot;)
</pre></div>
</div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="SP1_bulk.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tutorial: SP1 bindings in Cre-driver mouse lines.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="K562HCT116_SP1.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorial: K562 HCT116 SP1 single-cell callingcards data.</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script integrity="sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK" src="../../_static/jquery.js"></script>
    <script integrity="sha384-lSZeSIVKp9myfKbDQ3GkN/KHjUc+mzg17VKDN4Y2kUeBSJioB9QSM639vM9fuY//" src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/scripts/furo.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>