<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<meta property="og:title" content="Tutorial: K562 HCT116 SP1 single-cell callingcards data." />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pycallingcards.readthedocs.io/en/latest/tutorials/notebooks/K562HCT116_SP1.html" />
<meta property="og:site_name" content="Pycallingcards" />
<meta property="og:description" content="In this data, we test transcription factor SP1 bindings in celline K562 and HCT116 by single-cell(sc) calling cards techenology. The data is from Moudgil et al., Cell. (2020) and could be download ..." />
<meta name="description" content="In this data, we test transcription factor SP1 bindings in celline K562 and HCT116 by single-cell(sc) calling cards techenology. The data is from Moudgil et al., Cell. (2020) and could be download ..." />
<link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data." href="Brd4_bulk.html" /><link rel="prev" title="Tutorial: Mouse cortex single-cell callingcards data." href="Mouse_cortex_Example.html" />

    <meta name="generator" content="sphinx-6.0.0, furo 2022.09.29"/>
        <title>Tutorial: K562 HCT116 SP1 single-cell callingcards data. - pycallingcards-tools</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?digest=d81277517bee4d6b0349d71bb2661d4890b5617c" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/override.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pycallingcards-tools</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Mouse_cortex_Example.html">Tutorial: Mouse cortex single-cell callingcards data.</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Tutorial: K562 HCT116 SP1 single-cell callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="Brd4_bulk.html">Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="SP1_bulk.html">Tutorial: SP1 bindings in Cre-driver mouse lines.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mousecortex_data.html">pycallingcards.datasets.mousecortex_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_K562HCT116_data.html">pycallingcards.datasets.SP1_K562HCT116_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mouse_brd4_data.html">pycallingcards.datasets.mouse_brd4_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_Cre_data.html">pycallingcards.datasets.SP1_Cre_data</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/reading.html">Reading</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_qbed.html">pycallingcards.reading.read_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_rna.html">pycallingcards.reading.read_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.combine_qbed.html">pycallingcards.reading.combine_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_qbed.html">pycallingcards.reading.save_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_bed.html">pycallingcards.reading.save_bed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/preprocessing.html">Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.callpeaks.html">pycallingcards.preprocessing.callpeaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.annotation.html">pycallingcards.preprocessing.annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.makeAnndata.html">pycallingcards.preprocessing.makeAnndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.filter_peaks.html">pycallingcards.preprocessing.filter_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.adata_insertions.html">pycallingcards.preprocessing.adata_insertions</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/tools.html">Tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.call_motif.html">pycallingcards.tools.call_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups.html">pycallingcards.tools.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bygroup.html">pycallingcards.tools.diff2group_bygroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bysample.html">pycallingcards.tools.diff2group_bysample</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.DE_pvalue.html">pycallingcards.tools.DE_pvalue</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_bulk.html">pycallingcards.tools.pair_peak_gene_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc.html">pycallingcards.tools.pair_peak_gene_sc</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area.html">pycallingcards.plotting.draw_area</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.whole_peaks.html">pycallingcards.plotting.whole_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.calculate_signal.html">pycallingcards.plotting.calculate_signal</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_plot.html">pycallingcards.plotting.signal_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_heatmap.html">pycallingcards.plotting.signal_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.rank_peak_groups.html">pycallingcards.plotting.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.volcano_plot.html">pycallingcards.plotting.volcano_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.heatmap.html">pycallingcards.plotting.heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_bulk.html">pycallingcards.plotting.dotplot_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc.html">pycallingcards.plotting.dotplot_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.sankey.html">pycallingcards.plotting.sankey</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Tutorial:-K562-HCT116-SP1-single-cell-callingcards-data.">
<h1>Tutorial: K562 HCT116 SP1 single-cell callingcards data.<a class="headerlink" href="#Tutorial:-K562-HCT116-SP1-single-cell-callingcards-data." title="Permalink to this heading">#</a></h1>
<p>In this data, we test transcription factor SP1 bindings in celline K562 and HCT116 by single-cell(sc) calling cards techenology. The data is from <a class="reference external" href="https://doi.org/10.1016/j.cell.2020.06.037">Moudgil et al., Cell. (2020)</a> and could be download from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE148448">GEO</a>.</p>
<p>In this tutorial, we will call peaks with backgound, make annotation, compare peaks with Chip-seq reference data and do differential peak analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pycallingcards as cc
import numpy as np
import pandas as pd
import scanpy as sc
</pre></div>
</div>
</div>
<p>We start by reading ccf data. In this data, one row is one insertion and columns indecate the chromosome, start point and end point, read number, direction and cell borcode of each insertion. For example, the first row means one insertion is in Chromosome 1 and starts from 30116 to 30120. The read number is 12 with direction from 5’ to 3’. The barcode of the cell is CCCAATCCATCGGTTA-1. Note that the barcodes of the cell would also be consistent with scRNA-seq data.</p>
<p>You could use <code class="docutils literal notranslate"><span class="pre">cc.rd.read_ccf(filename)</span></code> to read your own ccf data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># read experiement data
HCT116_SP1 = cc.datasets.SP1_K562HCT116_data(data=&quot;HCT116_SP1_ccf&quot;)
HCT116_SP1
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>30116</td>
      <td>30120</td>
      <td>5</td>
      <td>-</td>
      <td>CCCAATCCATCGGTTA-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>34568</td>
      <td>34572</td>
      <td>3</td>
      <td>-</td>
      <td>CCTTCGAAGGGCTTCC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>36736</td>
      <td>36740</td>
      <td>29</td>
      <td>+</td>
      <td>ACGAGCCGTATAGGTA-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>42447</td>
      <td>42451</td>
      <td>3</td>
      <td>-</td>
      <td>CTCTACGTCGGAGCAA-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>89697</td>
      <td>89701</td>
      <td>119</td>
      <td>-</td>
      <td>AGCTCTCGTTTGTTTC-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77205</th>
      <td>chrY</td>
      <td>25518788</td>
      <td>25518792</td>
      <td>2</td>
      <td>+</td>
      <td>TGGGCGTTCGAACGGA-1</td>
    </tr>
    <tr>
      <th>77206</th>
      <td>chrY</td>
      <td>56987633</td>
      <td>56987637</td>
      <td>13</td>
      <td>+</td>
      <td>CAGTCCTAGGCACATG-1</td>
    </tr>
    <tr>
      <th>77207</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>17</td>
      <td>+</td>
      <td>CGGAGCTCATCGACGC-1</td>
    </tr>
    <tr>
      <th>77208</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>7</td>
      <td>+</td>
      <td>GTAACGTAGTTACGGG-1</td>
    </tr>
    <tr>
      <th>77209</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>9</td>
      <td>+</td>
      <td>TCAGCAAGTTGAACTC-1</td>
    </tr>
  </tbody>
</table>
<p>77210 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># read backgound data
HCT116_brd4 = cc.datasets.SP1_K562HCT116_data(data=&quot;HCT116_brd4_ccf&quot;)
HCT116_brd4
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>89697</td>
      <td>89701</td>
      <td>14</td>
      <td>+</td>
      <td>TCTGAGACAATGGTCT-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>89697</td>
      <td>89701</td>
      <td>8</td>
      <td>+</td>
      <td>CAGCGACCAAATACAG-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>203932</td>
      <td>203936</td>
      <td>99</td>
      <td>+</td>
      <td>TTCTCCTTCTACTTAC-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>204063</td>
      <td>204067</td>
      <td>5</td>
      <td>-</td>
      <td>TGTTCCGGTGTAAGTA-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>204063</td>
      <td>204067</td>
      <td>7</td>
      <td>-</td>
      <td>CAAGATCTCGACCAGC-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>37769</th>
      <td>chrY</td>
      <td>18037315</td>
      <td>18037319</td>
      <td>9</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37770</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>168</td>
      <td>+</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37771</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>508</td>
      <td>+</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
    <tr>
      <th>37772</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>13</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37773</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>32</td>
      <td>-</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
  </tbody>
</table>
<p>37774 rows × 6 columns</p>
</div></div>
</div>
<p>Because insertions are discrete, we now call peaks to deduce potential binding sites. Three different methods (<em>CCcaller,ccf_tools,Blockify</em>) are available along with there different species (<em>hg38,mm10,yeast</em>).</p>
<p>In this setting, we use <strong>ccf_tools</strong> in <em>human</em>(‘hg38’) data. <em>window_size</em> is the most important parameter for <em>ccf_tools</em>, it is highly related to the length of a peak. 2000-3500 is a good fit for <em>window_size</em>. <em>step_size</em> is another important paramenter and it controls how careful we are looking into the chromosomes. 500-1000 is good for <em>step_size</em>. <em>pvalue_cutoffTTAA</em> is the pvalue cutoff for TTAA data and <em>pvalue_cutoffbg</em> is pvalue cutoff for the background ccf data. Normally, the
setting for <em>pvalue_cutoffbg</em> is higher than <em>pvalue_cutoffTTAA</em>, but both of them should not exceed 0.1.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_data_HCT116 = cc.pp.callpeaks(HCT116_SP1, HCT116_brd4, method = &quot;ccf_tools&quot;, reference = &quot;hg38&quot;,  window_size = 2000, step_size = 500,
                  pvalue_cutoffTTAA = 0.001, pvalue_cutoffbg = 0.1, lam_win_size = None,  pseudocounts = 0.1, record = True, save = &quot;peak.bed&quot;)
peak_data_HCT116
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the ccf_tools method with background, [expdata, background, reference, pvalue_cutoffbg, pvalue_cutoffTTAA, lam_win_size, window_size, step_size, extend, pseudocounts, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 25/25 [00:57&lt;00:00,  2.32s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Center</th>
      <th>Experiment Insertions</th>
      <th>Background insertions</th>
      <th>Reference Insertions</th>
      <th>pvalue Reference</th>
      <th>pvalue Background</th>
      <th>Fraction Experiment</th>
      <th>TPH Experiment</th>
      <th>Fraction background</th>
      <th>TPH background</th>
      <th>TPH background subtracted</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>906689</td>
      <td>907160</td>
      <td>906957.0</td>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>3.099334e-09</td>
      <td>1.546531e-04</td>
      <td>0.000065</td>
      <td>6475.845098</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>6475.845098</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>999921</td>
      <td>1000324</td>
      <td>1000121.0</td>
      <td>20</td>
      <td>0</td>
      <td>1</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>0.000259</td>
      <td>25903.380391</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>25903.380391</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1156947</td>
      <td>1157863</td>
      <td>1157660.0</td>
      <td>11</td>
      <td>0</td>
      <td>2</td>
      <td>0.000000e+00</td>
      <td>1.274899e-09</td>
      <td>0.000142</td>
      <td>14246.859215</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>14246.859215</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1692740</td>
      <td>1693542</td>
      <td>1693339.0</td>
      <td>6</td>
      <td>0</td>
      <td>3</td>
      <td>5.135270e-11</td>
      <td>1.546531e-04</td>
      <td>0.000078</td>
      <td>7771.014117</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>7771.014117</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1744492</td>
      <td>1746808</td>
      <td>1746605.0</td>
      <td>11</td>
      <td>0</td>
      <td>7</td>
      <td>0.000000e+00</td>
      <td>1.274899e-09</td>
      <td>0.000142</td>
      <td>14246.859215</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>14246.859215</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3009</th>
      <td>chrY</td>
      <td>1245503</td>
      <td>1247177</td>
      <td>1245703.0</td>
      <td>5</td>
      <td>0</td>
      <td>7</td>
      <td>1.652897e-09</td>
      <td>4.678840e-03</td>
      <td>0.000065</td>
      <td>6475.845098</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>6475.845098</td>
    </tr>
    <tr>
      <th>3010</th>
      <td>chrY</td>
      <td>1280372</td>
      <td>1281989</td>
      <td>1281786.0</td>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>1.426995e-09</td>
      <td>4.678840e-03</td>
      <td>0.000065</td>
      <td>6475.845098</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>6475.845098</td>
    </tr>
    <tr>
      <th>3011</th>
      <td>chrY</td>
      <td>1586317</td>
      <td>1587733</td>
      <td>1587530.0</td>
      <td>7</td>
      <td>0</td>
      <td>8</td>
      <td>3.370637e-13</td>
      <td>1.546531e-04</td>
      <td>0.000091</td>
      <td>9066.183137</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>9066.183137</td>
    </tr>
    <tr>
      <th>3012</th>
      <td>chrY</td>
      <td>2391936</td>
      <td>2392440</td>
      <td>2392237.0</td>
      <td>6</td>
      <td>1</td>
      <td>2</td>
      <td>1.985434e-11</td>
      <td>9.958372e-02</td>
      <td>0.000078</td>
      <td>7771.014117</td>
      <td>0.000026</td>
      <td>2647.323556</td>
      <td>5123.690561</td>
    </tr>
    <tr>
      <th>3013</th>
      <td>chrY</td>
      <td>2608793</td>
      <td>2610054</td>
      <td>2608993.0</td>
      <td>8</td>
      <td>0</td>
      <td>2</td>
      <td>2.775558e-15</td>
      <td>1.546531e-04</td>
      <td>0.000104</td>
      <td>10361.352156</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>10361.352156</td>
    </tr>
  </tbody>
</table>
<p>3014 rows × 14 columns</p>
</div></div>
</div>
<p>In order to choose suitable method and parameters for peak calling, please take a look at genome areas. We stongly advise to adjust the parameters for <code class="docutils literal notranslate"><span class="pre">cc.pp.callpeaks()</span></code> to call better peaks.</p>
<p>In this plot, the colored ones are the experiment ccf data and the gray ones are the background data. The first part is insertions and their read counts. One dot is an insertion and the height is log(reads+1). The second part is the distribution of insertions. The third part represents reference genes and peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr1&quot;,999921,1000324,50000,peak_data_HCT116, HCT116_SP1, &quot;hg38&quot;, HCT116_brd4,
                figsize = (30,6),peak_line = 1,save = False,plotsize = [1,1,4], example_length = 1000)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_9_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_9_0.png" />
</div>
</div>
<p>We could also visualize our data in <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a> directly. Notice that this link only valid for 24h, please rerun it if you want to use it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>qbed = {&quot;SP1&quot;:HCT116_SP1, &quot;Brd4&quot;: HCT116_brd4}
bed = {&quot;peak&quot;:peak_data_HCT116}
cc.pl.WashU_browser_url(qbed = qbed,bed = bed,genome = &#39;hg38&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=hg38&amp;hub=https://companion.epigenomegateway.org//task/a61a33350deaae2be3bd1db2266adfac/output//datahub.json
</pre></div></div>
</div>
<p>We advise to look at the whole chromosome and see the distribution of peaks to make sure everything is going right.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.whole_peaks(peak_data_HCT116, reference = &quot;hg38&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_13_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_13_0.png" />
</div>
</div>
<p>We then check the peak with reference Chip-seq data of SP1 in HCT116 from <a class="reference external" href="https://www.encodeproject.org/experiments/ENCSR000BSF/">ENCSR000BSF</a> (use the bigWig file <a class="reference external" href="https://www.encodeproject.org/files/ENCFF587ZMX/">ENCFF587ZMX</a> generated by it).</p>
<p>Download the data if needed by:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!wget https://www.encodeproject.org/files/ENCFF587ZMX/@@download/ENCFF587ZMX.bigWig
</pre></div>
</div>
<p>We first calculate the signal of the Chip-seq signal around the peak.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mtx_HCT116 = cc.pl.calculate_signal(peak_data_HCT116,
                                    chipseq_signal = &quot;ENCFF587ZMX.bigWig&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3014/3014 [00:22&lt;00:00, 136.66it/s]
</pre></div></div>
</div>
<p>Visualized by the plotting the siganl values.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.signal_plot(mtx_HCT116, alpha = 0.05)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_18_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_18_0.png" />
</div>
</div>
<p>Visualized by the plotting the signal heatmap plot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.signal_heatmap(mtx_HCT116,pad = 0.035)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_20_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_20_0.png" />
</div>
</div>
<p>Do the exact same thing for K562 SP1 data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># read experiment data
K562_SP1 = cc.datasets.SP1_K562HCT116_data(data=&quot;K562_SP1_ccf&quot;)
K562_SP1
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>16529</td>
      <td>16533</td>
      <td>163</td>
      <td>-</td>
      <td>GCTCCTAAGTACGTTC-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>10</td>
      <td>+</td>
      <td>CTCACACCAGACGCTC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>155</td>
      <td>+</td>
      <td>TGGCCAGCACCCATTC-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>285</td>
      <td>+</td>
      <td>GTGGGTCCACGGCCAT-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>7</td>
      <td>+</td>
      <td>CGTCTACTCAACACGT-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>327460</th>
      <td>chrY</td>
      <td>57061562</td>
      <td>57061566</td>
      <td>6</td>
      <td>+</td>
      <td>CTCATTATCATCATTC-1</td>
    </tr>
    <tr>
      <th>327461</th>
      <td>chrY</td>
      <td>57061562</td>
      <td>57061566</td>
      <td>67</td>
      <td>+</td>
      <td>TGCGTGGCATTAGGCT-1</td>
    </tr>
    <tr>
      <th>327462</th>
      <td>chrY</td>
      <td>57145084</td>
      <td>57145088</td>
      <td>2</td>
      <td>-</td>
      <td>ACATACGTCGCGCCAA-1</td>
    </tr>
    <tr>
      <th>327463</th>
      <td>chrY</td>
      <td>57148630</td>
      <td>57148634</td>
      <td>2</td>
      <td>-</td>
      <td>TATGCCCGTACAGTTC-1</td>
    </tr>
    <tr>
      <th>327464</th>
      <td>chrY</td>
      <td>57183913</td>
      <td>57183917</td>
      <td>228</td>
      <td>-</td>
      <td>AAACCTGGTCCTGCTT-1</td>
    </tr>
  </tbody>
</table>
<p>327465 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># read background data
K562_brd4 =cc.datasets.SP1_K562HCT116_data(data=&quot;K562_brd4_ccf&quot;)
K562_brd4
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>30238</td>
      <td>30242</td>
      <td>3</td>
      <td>+</td>
      <td>TTTACTGCATAAAGGT-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>2</td>
      <td>-</td>
      <td>ATCACGAAGAGTAATC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>70</td>
      <td>+</td>
      <td>TTGAACGCAAATCCGT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>31101</td>
      <td>31105</td>
      <td>2</td>
      <td>+</td>
      <td>CCTCAGTCATCAGTAC-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>32116</td>
      <td>32120</td>
      <td>5</td>
      <td>+</td>
      <td>CTAGTGAAGACAAAGG-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>107380</th>
      <td>chrY</td>
      <td>57080210</td>
      <td>57080214</td>
      <td>9</td>
      <td>-</td>
      <td>AAGGAGCCAGTATAAG-1</td>
    </tr>
    <tr>
      <th>107381</th>
      <td>chrY</td>
      <td>57087785</td>
      <td>57087789</td>
      <td>24</td>
      <td>-</td>
      <td>CGAGCCAGTCTCTCTG-1</td>
    </tr>
    <tr>
      <th>107382</th>
      <td>chrY</td>
      <td>57144853</td>
      <td>57144857</td>
      <td>5</td>
      <td>+</td>
      <td>GAAGCAGTCCCATTTA-1</td>
    </tr>
    <tr>
      <th>107383</th>
      <td>chrY</td>
      <td>57183772</td>
      <td>57183776</td>
      <td>2</td>
      <td>-</td>
      <td>TCTTTCCTCTTGCCGT-1</td>
    </tr>
    <tr>
      <th>107384</th>
      <td>chrY</td>
      <td>57204853</td>
      <td>57204857</td>
      <td>369</td>
      <td>-</td>
      <td>ATAACGCAGTTTGCGT-1</td>
    </tr>
  </tbody>
</table>
<p>107385 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_data_K562 = cc.pp.callpeaks(K562_SP1, K562_brd4, method = &quot;ccf_tools&quot;, reference = &quot;hg38&quot;, window_size = 2000, step_size = 500,
                  pvalue_cutoffTTAA = 0.0001,  pvalue_cutoffbg = 0.1, lam_win_size = None,  pseudocounts = 0.1, record = True, save = &quot;peak_k562.bed&quot;)
peak_data_K562
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the ccf_tools method with background, [expdata, background, reference, pvalue_cutoffbg, pvalue_cutoffTTAA, lam_win_size, window_size, step_size, extend, pseudocounts, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 24/24 [02:11&lt;00:00,  5.47s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Center</th>
      <th>Experiment Insertions</th>
      <th>Background insertions</th>
      <th>Reference Insertions</th>
      <th>pvalue Reference</th>
      <th>pvalue Background</th>
      <th>Fraction Experiment</th>
      <th>TPH Experiment</th>
      <th>Fraction background</th>
      <th>TPH background</th>
      <th>TPH background subtracted</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>29684</td>
      <td>30087</td>
      <td>29884.0</td>
      <td>6</td>
      <td>0</td>
      <td>1</td>
      <td>8.878753e-11</td>
      <td>1.546531e-04</td>
      <td>0.000018</td>
      <td>1832.256882</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1832.256882</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>36239</td>
      <td>38107</td>
      <td>37578.0</td>
      <td>24</td>
      <td>2</td>
      <td>15</td>
      <td>0.000000e+00</td>
      <td>1.486029e-03</td>
      <td>0.000073</td>
      <td>7329.027530</td>
      <td>0.000019</td>
      <td>1862.457513</td>
      <td>5466.570017</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>198893</td>
      <td>201208</td>
      <td>200869.0</td>
      <td>28</td>
      <td>2</td>
      <td>11</td>
      <td>0.000000e+00</td>
      <td>6.927041e-05</td>
      <td>0.000086</td>
      <td>8550.532118</td>
      <td>0.000019</td>
      <td>1862.457513</td>
      <td>6688.074605</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>203351</td>
      <td>207161</td>
      <td>205004.0</td>
      <td>92</td>
      <td>13</td>
      <td>22</td>
      <td>0.000000e+00</td>
      <td>4.337485e-05</td>
      <td>0.000281</td>
      <td>28094.605530</td>
      <td>0.000121</td>
      <td>12105.973832</td>
      <td>15988.631698</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>265549</td>
      <td>266336</td>
      <td>265749.0</td>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>3.731359e-08</td>
      <td>4.678840e-03</td>
      <td>0.000015</td>
      <td>1526.880735</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1526.880735</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>9871</th>
      <td>chrY</td>
      <td>15158250</td>
      <td>15158653</td>
      <td>15158450.0</td>
      <td>11</td>
      <td>0</td>
      <td>1</td>
      <td>0.000000e+00</td>
      <td>1.546531e-04</td>
      <td>0.000034</td>
      <td>3359.137618</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>3359.137618</td>
    </tr>
    <tr>
      <th>9872</th>
      <td>chrY</td>
      <td>16985442</td>
      <td>16985845</td>
      <td>16985642.0</td>
      <td>5</td>
      <td>0</td>
      <td>2</td>
      <td>1.806731e-09</td>
      <td>4.678840e-03</td>
      <td>0.000015</td>
      <td>1526.880735</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1526.880735</td>
    </tr>
    <tr>
      <th>9873</th>
      <td>chrY</td>
      <td>19753311</td>
      <td>19753714</td>
      <td>19753511.0</td>
      <td>33</td>
      <td>0</td>
      <td>1</td>
      <td>0.000000e+00</td>
      <td>2.269296e-13</td>
      <td>0.000101</td>
      <td>10077.412853</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>10077.412853</td>
    </tr>
    <tr>
      <th>9874</th>
      <td>chrY</td>
      <td>21011133</td>
      <td>21011828</td>
      <td>21011333.0</td>
      <td>5</td>
      <td>0</td>
      <td>4</td>
      <td>2.510448e-09</td>
      <td>4.678840e-03</td>
      <td>0.000015</td>
      <td>1526.880735</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1526.880735</td>
    </tr>
    <tr>
      <th>9875</th>
      <td>chrY</td>
      <td>56952574</td>
      <td>56957328</td>
      <td>56953707.0</td>
      <td>40</td>
      <td>1</td>
      <td>37</td>
      <td>0.000000e+00</td>
      <td>2.427052e-06</td>
      <td>0.000122</td>
      <td>12215.045883</td>
      <td>0.000009</td>
      <td>931.228756</td>
      <td>11283.817126</td>
    </tr>
  </tbody>
</table>
<p>9876 rows × 14 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr10&quot;,3048452,3049913,60000,peak_data_K562,K562_SP1, &quot;hg38&quot;, K562_brd4 ,
                figsize = (30,6),peak_line = 4,save = False,bins =400, plotsize = [1,1,3], example_length = 1000)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_25_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_25_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>qbed = {&quot;SP1&quot;:K562_SP1, &quot;Brd4&quot;: K562_brd4}
bed = {&quot;peak&quot;:peak_data_K562}
cc.pl.WashU_browser_url(qbed = qbed,bed = bed,genome = &#39;hg38&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=hg38&amp;hub=https://companion.epigenomegateway.org//task/c2af4fdb5f4a1a57fabfd851b62d370e/output//datahub.json
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.whole_peaks(peak_data_K562, reference = &quot;hg38&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_27_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_27_0.png" />
</div>
</div>
<p>We could see that SP1 binds much more frequently in K562 than HCT116.</p>
<p>We could then check with reference Chip-seq data of SP1 in K562 from <a class="reference external" href="https://www.encodeproject.org/experiments/ENCSR372IML/">ENCSR372IML</a> (and use the bigWig file <a class="reference external" href="https://www.encodeproject.org/files/ENCFF588UII/">ENCFF588UII</a> generated by it）.</p>
<p>Down load the data if needed:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!wget https://www.encodeproject.org/files/ENCFF588UII/@@download/ENCFF588UII.bigWig
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mtx_K562 = cc.pl.calculate_signal(peak_data_K562, chipseq_signal = &quot;ENCFF588UII.bigWig&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 9876/9876 [01:04&lt;00:00, 152.73it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.signal_plot(mtx_K562, alpha = 0.05)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_31_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_31_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.signal_heatmap(mtx_K562,pad =  0.023, belowlength = 100)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/scratch/rmlab/rmlab_shared/g.juanru/pycallingcards/pycallingcards/plotting/_Chipseq.py:219: RuntimeWarning: Mean of empty slice
  cf = axis.pcolormesh(np.log2(signalmtx[np.argsort(np.nanmean(signalmtx,axis = 1)),:]+1),cmap = colormap, vmin=colormap_vmin, vmax=colormap_vmax)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_32_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_32_1.png" />
</div>
</div>
<p>We could see that calling cards peaks are consistent with Chip-seq data. Peak centers tend to have higher signal and the sigmal goes lower as the distance get larger.</p>
<p>Because we call peaks for two data separately, we need to combine two peaks together using <a class="reference external" href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pybedtools
peak = cc.rd.combine_qbed([peak_data_HCT116, peak_data_K562])
peak = pybedtools.BedTool.from_dataframe(peak).merge().to_dataframe()
peak_data = peak.rename(columns={&quot;chrom&quot;:&quot;Chr&quot;, &quot;start&quot;:&quot;Start&quot;, &quot;end&quot;:&quot;End&quot;})
peak_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>29684</td>
      <td>30087</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>36239</td>
      <td>38107</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>198893</td>
      <td>201208</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>203351</td>
      <td>207161</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>265549</td>
      <td>266336</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>10888</th>
      <td>chrY</td>
      <td>15158250</td>
      <td>15158653</td>
    </tr>
    <tr>
      <th>10889</th>
      <td>chrY</td>
      <td>16985442</td>
      <td>16985845</td>
    </tr>
    <tr>
      <th>10890</th>
      <td>chrY</td>
      <td>19753311</td>
      <td>19753714</td>
    </tr>
    <tr>
      <th>10891</th>
      <td>chrY</td>
      <td>21011133</td>
      <td>21011828</td>
    </tr>
    <tr>
      <th>10892</th>
      <td>chrY</td>
      <td>56952574</td>
      <td>56957328</td>
    </tr>
  </tbody>
</table>
<p>10893 rows × 3 columns</p>
</div></div>
</div>
<p>Check for the peaks again.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr1&quot;,999921,1000324,15000,peak_data, HCT116_SP1, &quot;hg38&quot;, HCT116_brd4,
                figsize = (30,5),peak_line = 2,save = False,plotsize = [1,1,3],  example_length = 1000)
cc.pl.draw_area(&quot;chr1&quot;,999921,1000324,15000,peak_data, K562_SP1, &quot;hg38&quot;, K562_brd4,
                figsize = (30,5),peak_line = 2,save = False,plotsize = [1,1,3],  example_length = 1000)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_37_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_37_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_37_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_37_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr10&quot;,3048452,3049913,60000,peak_data, HCT116_SP1, &quot;hg38&quot;, HCT116_brd4,
                figsize = (30,7),peak_line = 3,save = False, bins = 200, plotsize = [1,1,4], example_length = 1000)
cc.pl.draw_area(&quot;chr10&quot;,3048452,3049913,60000,peak_data, K562_SP1, &quot;hg38&quot;, K562_brd4,
                figsize = (30,7),peak_line = 3,save = False, bins = 200, plotsize = [1,1,4],  example_length = 1000)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_38_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_38_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_38_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_38_1.png" />
</div>
</div>
<p>Results seem to be good! Congratulations! Find out the closest genes for each peak for annotation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_annotation = cc.pp.annotation(peak_data, reference = &quot;hg38&quot;)
peak_annotation = cc.pp.combine_annotation(peak_data,peak_annotation)
peak_annotation
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In the bedtools method, we would use bedtools in the default path. Set bedtools path by &#39;bedtools_path&#39; if needed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Nearest Refseq1</th>
      <th>Gene Name1</th>
      <th>Nearest Refseq2</th>
      <th>Gene Name2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>29684</td>
      <td>30087</td>
      <td>NR_036051</td>
      <td>MIR1302-2</td>
      <td>NR_024540</td>
      <td>WASH7P</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>36239</td>
      <td>38107</td>
      <td>NR_026818</td>
      <td>FAM138A</td>
      <td>NR_036051</td>
      <td>MIR1302-2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>198893</td>
      <td>201208</td>
      <td>NR_026823</td>
      <td>FAM138D</td>
      <td>NR_107063</td>
      <td>MIR6859-3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>203351</td>
      <td>207161</td>
      <td>NR_026823</td>
      <td>FAM138D</td>
      <td>NR_107063</td>
      <td>MIR6859-3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>265549</td>
      <td>266336</td>
      <td>NR_026823</td>
      <td>FAM138D</td>
      <td>NR_107063</td>
      <td>MIR6859-3</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>10888</th>
      <td>chrY</td>
      <td>15158250</td>
      <td>15158653</td>
      <td>NM_001206850</td>
      <td>NLGN4Y</td>
      <td>NR_046504</td>
      <td>NLGN4Y-AS1</td>
    </tr>
    <tr>
      <th>10889</th>
      <td>chrY</td>
      <td>16985442</td>
      <td>16985845</td>
      <td>NR_028083</td>
      <td>FAM41AY1</td>
      <td>NR_002160</td>
      <td>FAM224B</td>
    </tr>
    <tr>
      <th>10890</th>
      <td>chrY</td>
      <td>19753311</td>
      <td>19753714</td>
      <td>NM_001146706</td>
      <td>KDM5D</td>
      <td>NR_045128</td>
      <td>TXLNGY</td>
    </tr>
    <tr>
      <th>10891</th>
      <td>chrY</td>
      <td>21011133</td>
      <td>21011828</td>
      <td>NM_001039567</td>
      <td>RPS4Y2</td>
      <td>NM_001282471</td>
      <td>PRORY</td>
    </tr>
    <tr>
      <th>10892</th>
      <td>chrY</td>
      <td>56952574</td>
      <td>56957328</td>
      <td>NM_005840</td>
      <td>SPRY3</td>
      <td>NM_001145149</td>
      <td>VAMP7</td>
    </tr>
  </tbody>
</table>
<p>10893 rows × 7 columns</p>
</div></div>
</div>
<p>Combine the two experiment ccf files to make anndata object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>exp_ccf = pd.concat([K562_SP1,HCT116_SP1])
exp_ccf
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>16529</td>
      <td>16533</td>
      <td>163</td>
      <td>-</td>
      <td>GCTCCTAAGTACGTTC-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>10</td>
      <td>+</td>
      <td>CTCACACCAGACGCTC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>155</td>
      <td>+</td>
      <td>TGGCCAGCACCCATTC-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>285</td>
      <td>+</td>
      <td>GTGGGTCCACGGCCAT-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>7</td>
      <td>+</td>
      <td>CGTCTACTCAACACGT-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77205</th>
      <td>chrY</td>
      <td>25518788</td>
      <td>25518792</td>
      <td>2</td>
      <td>+</td>
      <td>TGGGCGTTCGAACGGA-1</td>
    </tr>
    <tr>
      <th>77206</th>
      <td>chrY</td>
      <td>56987633</td>
      <td>56987637</td>
      <td>13</td>
      <td>+</td>
      <td>CAGTCCTAGGCACATG-1</td>
    </tr>
    <tr>
      <th>77207</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>17</td>
      <td>+</td>
      <td>CGGAGCTCATCGACGC-1</td>
    </tr>
    <tr>
      <th>77208</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>7</td>
      <td>+</td>
      <td>GTAACGTAGTTACGGG-1</td>
    </tr>
    <tr>
      <th>77209</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>9</td>
      <td>+</td>
      <td>TCAGCAAGTTGAACTC-1</td>
    </tr>
  </tbody>
</table>
<p>404675 rows × 6 columns</p>
</div></div>
</div>
<p>Read the barocede file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>barcodes = cc.datasets.SP1_K562HCT116_data(data = &quot;barcodes&quot;)
barcodes = barcodes.drop_duplicates(subset=[&#39;Index&#39;])
barcodes
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Index</th>
      <th>cluster</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AAACCTGAGAAAGTGG-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AAACCTGAGACCGGAT-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAACCTGAGACTAGAT-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAACCTGAGAGCTTCT-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AAACCTGAGAGTACCG-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>52206</th>
      <td>TTTGTCATCTCCGGTT-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>52207</th>
      <td>TTTGTCATCTCGATGA-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>52208</th>
      <td>TTTGTCATCTCTAAGG-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>52209</th>
      <td>TTTGTCATCTGGAGCC-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>52210</th>
      <td>TTTGTCATCTTGGGTA-1</td>
      <td>HCT116</td>
    </tr>
  </tbody>
</table>
<p>51079 rows × 2 columns</p>
</div></div>
</div>
<p>Use ccf data, peak data and barcodes data to make a cell by peak anndata object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf = cc.pp.makeAnndata(exp_ccf, peak_annotation, barcodes)
adata_ccf
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 24/24 [00:02&lt;00:00,  8.05it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 51079 × 10893
    obs: &#39;cluster&#39;
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;
</pre></div></div>
</div>
<p>Although one peak should have many insertions, but there is a chance that all the cells from the peak was filtered by the RNA preprocesssing. In this case, we advice to filter the peaks. Additioanlly we also filter cells has very few insertions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pp.filter_peaks(adata_ccf, min_counts=5)
cc.pp.filter_peaks(adata_ccf, min_cells=5)
adata_ccf
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 51079 × 10887
    obs: &#39;cluster&#39;
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;n_counts&#39;, &#39;n_cells&#39;
</pre></div></div>
</div>
<p>Differential peak analysis would find out the significant binding for each group. In this example, we use <em>fisher exact test</em> to find out.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.tl.rank_peak_groups(adata_ccf,&quot;cluster&quot;,method = &#39;fisher_exact&#39;,key_added = &#39;fisher_exact&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2/2 [02:17&lt;00:00, 68.87s/it]
</pre></div></div>
</div>
<p>Plot the results for differential peak analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.rank_peak_groups(adata_ccf, key = &#39;fisher_exact&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_52_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_52_0.png" />
</div>
</div>
<p>Then, we take a look at the genome for highly differentiated peaks. The colored ones are the insertions for specific cluster and the grey ones are the backgound insertion information. We could see some huge difference between SP1 bindings in HCT116 and SP1 bindings in K562.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>bg_ccf = pd.concat([K562_brd4,HCT116_brd4])
bg_ccf
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>30238</td>
      <td>30242</td>
      <td>3</td>
      <td>+</td>
      <td>TTTACTGCATAAAGGT-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>2</td>
      <td>-</td>
      <td>ATCACGAAGAGTAATC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>70</td>
      <td>+</td>
      <td>TTGAACGCAAATCCGT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>31101</td>
      <td>31105</td>
      <td>2</td>
      <td>+</td>
      <td>CCTCAGTCATCAGTAC-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>32116</td>
      <td>32120</td>
      <td>5</td>
      <td>+</td>
      <td>CTAGTGAAGACAAAGG-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>37769</th>
      <td>chrY</td>
      <td>18037315</td>
      <td>18037319</td>
      <td>9</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37770</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>168</td>
      <td>+</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37771</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>508</td>
      <td>+</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
    <tr>
      <th>37772</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>13</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37773</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>32</td>
      <td>-</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
  </tbody>
</table>
<p>145159 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr15&quot;,70763238,70765267,200000,peak_data,exp_ccf,&quot;hg38&quot;,adata = adata_ccf,  bins = 150,
                name = &quot;HCT116&quot;,key =&#39;cluster&#39;,figsize = (30,6),peak_line = 4,color = &quot;purple&quot;, plotsize = [1,1,5], title = &quot;chr15_70763238_70765267_HCT116&quot;)
cc.pl.draw_area(&quot;chr15&quot;,70763238,70765267,200000,peak_data,exp_ccf,&quot;hg38&quot;,adata = adata_ccf, bins = 150,
                name = &quot;K562&quot;,key =&#39;cluster&#39;,figsize = (30,6),peak_line = 4,color = &quot;red&quot;, plotsize = [1,1,5], title = &quot;chr15_70763238_70765267_K562&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_55_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_55_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_55_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_55_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr7&quot;,143706539,143718962,100000,peak_data,exp_ccf,&quot;hg38&quot;,adata = adata_ccf, bins = 150,
                name = &quot;K562&quot;,key =&#39;cluster&#39;,figsize = (30,6),peak_line = 4,color = &quot;red&quot;, plotsize = [1,1,5], title = &quot;chr7_143706539_143718962_K562&quot;)
cc.pl.draw_area(&quot;chr7&quot;,143706539,143718962,100000,peak_data,exp_ccf,&quot;hg38&quot;,adata = adata_ccf,  bins = 150,
                name = &quot;HCT116&quot;,key =&#39;cluster&#39;,figsize = (30,6),peak_line = 4,color = &quot;purple&quot;, plotsize = [1,1,5], title = &quot;chr7_143706539_143718962_HCT116&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_56_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_56_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_56_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_56_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr8&quot;,17799370,17802353,100000,peak_data,exp_ccf,&quot;hg38&quot;,adata = adata_ccf,  bins = 150,
                name = &quot;K562&quot;,key =&#39;cluster&#39;,figsize = (30,7),peak_line = 6,color = &quot;red&quot;, plotsize = [1,1,6], title = &quot;chr19_9818643_9820060&quot;)
cc.pl.draw_area(&quot;chr8&quot;,17799370,17802353,100000,peak_data,exp_ccf,&quot;hg38&quot;,adata = adata_ccf,  bins = 150,
                name = &quot;HCT116&quot;,key =&#39;cluster&#39;,figsize = (30,7),peak_line = 6,color = &quot;purple&quot;, plotsize = [1,1,6], title = &quot;chr19_9818643_9820060&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_57_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_57_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_57_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_57_1.png" />
</div>
</div>
<p>Saved the file if needed.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf.write(&quot;SP1_CCF.h5ad&quot;)
</pre></div>
</div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="Brd4_bulk.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="Mouse_cortex_Example.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorial: Mouse cortex single-cell callingcards data.</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script integrity="sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK" src="../../_static/jquery.js"></script>
    <script integrity="sha384-lSZeSIVKp9myfKbDQ3GkN/KHjUc+mzg17VKDN4Y2kUeBSJioB9QSM639vM9fuY//" src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/scripts/furo.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>