<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<meta property="og:title" content="Tutorial: Mouse cortex single-cell callingcards data." />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pycallingcards.readthedocs.io/en/latest/tutorials/notebooks/Mouse_cortex_Example.html" />
<meta property="og:site_name" content="Pycallingcards" />
<meta property="og:description" content="This is a brd4 single-cell(sc) callingcards dataset in mouse cortex from Moudgil et al., Cell. (2020) and could be downloaded from GEO. In this tutorial, we will call peaks, make annotation, do dif..." />
<meta name="description" content="This is a brd4 single-cell(sc) callingcards dataset in mouse cortex from Moudgil et al., Cell. (2020) and could be downloaded from GEO. In this tutorial, we will call peaks, make annotation, do dif..." />
<link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Tutorial: K562 HCT116 SP1 single-cell callingcards data." href="K562HCT116_SP1.html" /><link rel="prev" title="Tutorials" href="../index.html" />

    <meta name="generator" content="sphinx-6.0.0, furo 2022.09.29"/>
        <title>Tutorial: Mouse cortex single-cell callingcards data. - pycallingcards-tools</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?digest=d81277517bee4d6b0349d71bb2661d4890b5617c" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/override.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pycallingcards-tools</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Tutorial: Mouse cortex single-cell callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="K562HCT116_SP1.html">Tutorial: K562 HCT116 SP1 single-cell callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="Brd4_bulk.html">Tutorial: Brd4 bindings for Male/Female mouses bulk callingcards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="SP1_bulk.html">Tutorial: SP1 bindings in Cre-driver mouse lines.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mousecortex_data.html">pycallingcards.datasets.mousecortex_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_K562HCT116_data.html">pycallingcards.datasets.SP1_K562HCT116_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mouse_brd4_data.html">pycallingcards.datasets.mouse_brd4_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_Cre_data.html">pycallingcards.datasets.SP1_Cre_data</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/reading.html">Reading</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_qbed.html">pycallingcards.reading.read_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_rna.html">pycallingcards.reading.read_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.combine_qbed.html">pycallingcards.reading.combine_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_qbed.html">pycallingcards.reading.save_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_bed.html">pycallingcards.reading.save_bed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/preprocessing.html">Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.callpeaks.html">pycallingcards.preprocessing.callpeaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.annotation.html">pycallingcards.preprocessing.annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.makeAnndata.html">pycallingcards.preprocessing.makeAnndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.filter_peaks.html">pycallingcards.preprocessing.filter_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.adata_insertions.html">pycallingcards.preprocessing.adata_insertions</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/tools.html">Tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.call_motif.html">pycallingcards.tools.call_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups.html">pycallingcards.tools.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bygroup.html">pycallingcards.tools.diff2group_bygroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bysample.html">pycallingcards.tools.diff2group_bysample</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.DE_pvalue.html">pycallingcards.tools.DE_pvalue</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_bulk.html">pycallingcards.tools.pair_peak_gene_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc.html">pycallingcards.tools.pair_peak_gene_sc</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area.html">pycallingcards.plotting.draw_area</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.whole_peaks.html">pycallingcards.plotting.whole_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.calculate_signal.html">pycallingcards.plotting.calculate_signal</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_plot.html">pycallingcards.plotting.signal_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_heatmap.html">pycallingcards.plotting.signal_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.rank_peak_groups.html">pycallingcards.plotting.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.volcano_plot.html">pycallingcards.plotting.volcano_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.heatmap.html">pycallingcards.plotting.heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_bulk.html">pycallingcards.plotting.dotplot_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc.html">pycallingcards.plotting.dotplot_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.sankey.html">pycallingcards.plotting.sankey</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Tutorial:-Mouse-cortex-single-cell-callingcards-data.">
<h1>Tutorial: Mouse cortex single-cell callingcards data.<a class="headerlink" href="#Tutorial:-Mouse-cortex-single-cell-callingcards-data." title="Permalink to this heading">#</a></h1>
<p>This is a brd4 single-cell(sc) callingcards dataset in mouse cortex from <a class="reference external" href="https://doi.org/10.1016/j.cell.2020.06.037">Moudgil et al., Cell. (2020)</a> and could be downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE148448">GEO</a>.</p>
<p>In this tutorial, we will call peaks, make annotation, do differential peak analysis, pair peaks with genes. In this dataset, there are 111382 insertions and 35950 cells in tatal. However, many cells are filted in scRNA-seq analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pycallingcards as cc
import numpy as np
import pandas as pd
import scanpy as sc
</pre></div>
</div>
</div>
<p>We start by reading ccf data. In this data, one row is one insertion and columns indecate the chromosome, start point and end point, read number, direction and cell borcode of each insertion. For example, the first row means one insertion is on Chromosome 1 and starts from 3112541 to 3112545. The read number is 12 with the direction from 3’ to 5’. The barcode of the cell is GATGAAAAGAGTTGGC-1. Note that the barcodes of cells would be consistent with scRNA-seq data.</p>
<p>You could use <code class="docutils literal notranslate"><span class="pre">cc.rd.read_ccf(filename)</span></code> to read your own ccf data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ccf_data = cc.datasets.mousecortex_data(data=&quot;ccf&quot;)
ccf_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3112541</td>
      <td>3112545</td>
      <td>12</td>
      <td>+</td>
      <td>GATGAAAAGAGTTGGC-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3121337</td>
      <td>3121341</td>
      <td>6</td>
      <td>-</td>
      <td>CGATCGGCACATTTCT-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3199281</td>
      <td>3199285</td>
      <td>7</td>
      <td>+</td>
      <td>GTCCTCATCTCCGGTT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3211433</td>
      <td>3211437</td>
      <td>22</td>
      <td>-</td>
      <td>CGAGAAGAGGAATCGC-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3245859</td>
      <td>3245863</td>
      <td>149</td>
      <td>+</td>
      <td>TTTACTGCATCCGCGA-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>111377</th>
      <td>chrY</td>
      <td>90807968</td>
      <td>90807972</td>
      <td>200</td>
      <td>-</td>
      <td>ACGGAGAGTCGCATAT-1</td>
    </tr>
    <tr>
      <th>111378</th>
      <td>chrY</td>
      <td>90833531</td>
      <td>90833535</td>
      <td>51</td>
      <td>-</td>
      <td>TAGCCGGTCCTGTACC-1</td>
    </tr>
    <tr>
      <th>111379</th>
      <td>chrY</td>
      <td>90833600</td>
      <td>90833604</td>
      <td>13</td>
      <td>-</td>
      <td>TTGGCAAAGAATTGTG-1</td>
    </tr>
    <tr>
      <th>111380</th>
      <td>chrY</td>
      <td>90840262</td>
      <td>90840266</td>
      <td>8</td>
      <td>-</td>
      <td>GTGCATAGTACCAGTT-1</td>
    </tr>
    <tr>
      <th>111381</th>
      <td>chrY</td>
      <td>90840262</td>
      <td>90840266</td>
      <td>95</td>
      <td>+</td>
      <td>TTGTAGGTCGAATCCA-1</td>
    </tr>
  </tbody>
</table>
<p>111382 rows × 6 columns</p>
</div></div>
</div>
<p>Because insertions are discrete, we now call peaks to deduce potential binding sites. Three different methods (<em>CCcaller,ccf_tools,Blockify</em>) are available along with there different species (<em>hg38,mm10,yeast</em>).</p>
<p>In this setting, we use <strong>CCcaller</strong> in <em>mouse</em>(‘mm10’) data. <em>maxbetween</em> is the most important parameter for <em>CCcaller</em>, it would control the maximum distance bwtween two nearby insertions, in another word, that is the minimun distance between two peaks. 800-2500 is a good parameter for <em>maxbetween</em>. <em>pvalue_cutoff</em> is also an important parameter and a number below 0.05 is strongly advised.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_data = cc.pp.callpeaks(ccf_data, method = &quot;CCcaller&quot;, reference = &quot;mm10&quot;,  maxbetween = 2000, pvalue_cutoff = 0.01,
                  lam_win_size = 1000000,  pseudocounts = 1, record = True)
peak_data
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the CCcaller method without background, [expdata, reference, pvalue_cutoff, lam_win_size, pseudocounts, minlen, extend, maxbetween, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 21/21 [00:22&lt;00:00,  1.06s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3399868</td>
      <td>3400272</td>
      <td>5</td>
      <td>4</td>
      <td>1.031353</td>
      <td>6.965061e-04</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>4806673</td>
      <td>4809049</td>
      <td>12</td>
      <td>20</td>
      <td>1.120541</td>
      <td>2.498336e-10</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>9300144</td>
      <td>9300901</td>
      <td>5</td>
      <td>5</td>
      <td>1.037149</td>
      <td>7.168254e-04</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>14302176</td>
      <td>14310895</td>
      <td>14</td>
      <td>92</td>
      <td>1.523252</td>
      <td>1.015569e-10</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>15287495</td>
      <td>15288141</td>
      <td>8</td>
      <td>4</td>
      <td>1.029800</td>
      <td>1.427167e-06</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2166</th>
      <td>chrX</td>
      <td>166172262</td>
      <td>166174518</td>
      <td>6</td>
      <td>10</td>
      <td>1.137755</td>
      <td>1.824630e-04</td>
    </tr>
    <tr>
      <th>2167</th>
      <td>chrX</td>
      <td>166241178</td>
      <td>166243587</td>
      <td>8</td>
      <td>18</td>
      <td>1.257844</td>
      <td>7.050309e-06</td>
    </tr>
    <tr>
      <th>2168</th>
      <td>chrX</td>
      <td>166345453</td>
      <td>166350005</td>
      <td>11</td>
      <td>35</td>
      <td>1.507674</td>
      <td>7.204418e-08</td>
    </tr>
    <tr>
      <th>2169</th>
      <td>chrX</td>
      <td>168117814</td>
      <td>168120255</td>
      <td>7</td>
      <td>21</td>
      <td>1.162110</td>
      <td>2.957329e-05</td>
    </tr>
    <tr>
      <th>2170</th>
      <td>chrX</td>
      <td>169842873</td>
      <td>169845831</td>
      <td>9</td>
      <td>28</td>
      <td>1.314399</td>
      <td>1.292113e-06</td>
    </tr>
  </tbody>
</table>
<p>2171 rows × 7 columns</p>
</div></div>
</div>
<p>In order to choose suitable method and parameters for peak calling, please take a look at genome areas. We stongly advise to adjust the parameters for <code class="docutils literal notranslate"><span class="pre">cc.pp.callpeaks()</span></code> to call better peaks.</p>
<p>In this plot, the first part is insertions and their read counts. One dot is an insertion and the height is log(reads+1). The second part is the distribution of insertions. The third part represents reference genes and peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr12&quot;,50102917,50124960,400000,peak_data,ccf_data,&quot;mm10&quot;,
                figsize = (30,5),peak_line = 4,save = True, title = &quot;chr12_50102917_50124960&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_8_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_8_0.png" />
</div>
</div>
<p>We could also visualize our data in <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a> directly. Notice that this link only valid for 24h, please rerun it if you want to use it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>qbed = {&quot;ccf_data&quot;: ccf_data}
bed = {&quot;peak&quot;:peak_data}
cc.pl.WashU_browser_url(qbed = qbed,bed = bed,genome = &#39;mm10&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=mm10&amp;hub=https://companion.epigenomegateway.org//task/f5b4d403b57cfdac7be9dc0f6795ffaa/output//datahub.json
</pre></div></div>
</div>
<p>We advise to look at the whole chromosome and see the distribution of peaks to make sure everything is going right.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.whole_peaks(peak_data, reference = &quot;mm10&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_12_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_12_0.png" />
</div>
</div>
<p>In the next step, we annotate the peaks by their closest genes using <a class="reference external" href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>. Make sure they are all previously installed before using.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>peak_annotation = cc.pp.annotation(peak_data, reference = &quot;mm10&quot;)
peak_annotation = cc.pp.combine_annotation(peak_data,peak_annotation)
peak_annotation
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In the bedtools method, we would use bedtools in the default path. Set bedtools path by &#39;bedtools_path&#39; if needed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
      <th>Nearest Refseq1</th>
      <th>Gene Name1</th>
      <th>Nearest Refseq2</th>
      <th>Gene Name2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3399868</td>
      <td>3400272</td>
      <td>5</td>
      <td>4</td>
      <td>1.031353</td>
      <td>6.965061e-04</td>
      <td>NM_001011874</td>
      <td>Xkr4</td>
      <td>NM_001195662</td>
      <td>Rp1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>4806673</td>
      <td>4809049</td>
      <td>12</td>
      <td>20</td>
      <td>1.120541</td>
      <td>2.498336e-10</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
      <td>NR_033530</td>
      <td>Mrpl15</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>9300144</td>
      <td>9300901</td>
      <td>5</td>
      <td>5</td>
      <td>1.037149</td>
      <td>7.168254e-04</td>
      <td>NM_001290390</td>
      <td>Sntg1</td>
      <td>NM_021511</td>
      <td>Rrs1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>14302176</td>
      <td>14310895</td>
      <td>14</td>
      <td>92</td>
      <td>1.523252</td>
      <td>1.015569e-10</td>
      <td>NM_010164</td>
      <td>Eya1</td>
      <td>NM_010827</td>
      <td>Msc</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>15287495</td>
      <td>15288141</td>
      <td>8</td>
      <td>4</td>
      <td>1.029800</td>
      <td>1.427167e-06</td>
      <td>NM_001098528</td>
      <td>Kcnb2</td>
      <td>NM_177781</td>
      <td>Trpa1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2166</th>
      <td>chrX</td>
      <td>166172262</td>
      <td>166174518</td>
      <td>6</td>
      <td>10</td>
      <td>1.137755</td>
      <td>1.824630e-04</td>
      <td>NM_001310724</td>
      <td>Gemin8</td>
      <td>NM_023122</td>
      <td>Gpm6b</td>
    </tr>
    <tr>
      <th>2167</th>
      <td>chrX</td>
      <td>166241178</td>
      <td>166243587</td>
      <td>8</td>
      <td>18</td>
      <td>1.257844</td>
      <td>7.050309e-06</td>
      <td>NM_023122</td>
      <td>Gpm6b</td>
      <td>NM_001310724</td>
      <td>Gemin8</td>
    </tr>
    <tr>
      <th>2168</th>
      <td>chrX</td>
      <td>166345453</td>
      <td>166350005</td>
      <td>11</td>
      <td>35</td>
      <td>1.507674</td>
      <td>7.204418e-08</td>
      <td>NM_023122</td>
      <td>Gpm6b</td>
      <td>NM_177429</td>
      <td>Ofd1</td>
    </tr>
    <tr>
      <th>2169</th>
      <td>chrX</td>
      <td>168117814</td>
      <td>168120255</td>
      <td>7</td>
      <td>21</td>
      <td>1.162110</td>
      <td>2.957329e-05</td>
      <td>NM_001290427</td>
      <td>Frmpd4</td>
      <td>NM_001290428</td>
      <td>Frmpd4</td>
    </tr>
    <tr>
      <th>2170</th>
      <td>chrX</td>
      <td>169842873</td>
      <td>169845831</td>
      <td>9</td>
      <td>28</td>
      <td>1.314399</td>
      <td>1.292113e-06</td>
      <td>NM_010797</td>
      <td>Mid1</td>
      <td>NM_001290506</td>
      <td>Mid1</td>
    </tr>
  </tbody>
</table>
<p>2171 rows × 11 columns</p>
</div></div>
</div>
<p>Then, we read the barocede file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>barcodes = cc.datasets.mousecortex_data(data=&quot;barcodes&quot;)
barcodes
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AAACCTGAGAACTCGG-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AAACCTGAGCAATCTC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAACCTGAGCCGTCGT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAACCTGAGTAGCGGT-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AAACCTGAGTGGAGTC-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>35945</th>
      <td>TTTGTCAAGTCCCACG-1</td>
    </tr>
    <tr>
      <th>35946</th>
      <td>TTTGTCACAGCGTCCA-1</td>
    </tr>
    <tr>
      <th>35947</th>
      <td>TTTGTCACATTTCACT-1</td>
    </tr>
    <tr>
      <th>35948</th>
      <td>TTTGTCAGTCGCATCG-1</td>
    </tr>
    <tr>
      <th>35949</th>
      <td>TTTGTCATCTTTACAC-1</td>
    </tr>
  </tbody>
</table>
<p>35950 rows × 1 columns</p>
</div></div>
</div>
<p>Use ccf data, peak data and barcode data to make a cell by peak <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">Anndata</a> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf = cc.pp.makeAnndata(ccf_data, peak_annotation, barcodes)
adata_ccf
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 20/20 [00:00&lt;00:00, 53.91it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 35950 × 2171
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;
</pre></div></div>
</div>
<p>Calling cards data is sequenced along with RNA-seq data. For the following steps, we are going to read scRNA-seq data and analyze them together. <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a> is recommended to load and analyze scRNA-seq data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = cc.datasets.mousecortex_data(data=&quot;RNA&quot;)
adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 30300 × 2638
    obs: &#39;batch&#39;, &#39;n_genes&#39;, &#39;total_counts&#39;, &#39;cluster&#39;
    var: &#39;n_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;
</pre></div></div>
</div>
<p>In the scRNA-seq analysis, many cells are filtered because of low quility. We need to make the cells in ccf anndata to be the exactly same as RNA-seq anndata.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf = adata_ccf[adata.obs.index]
adata_ccf,adata_ccf.X,adata_ccf.obs.index==adata_ccf.obs.index
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(View of AnnData object with n_obs × n_vars = 30300 × 2171
     var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;,
 &lt;30300x2171 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
        with 19818 stored elements in Compressed Sparse Column format&gt;,
 array([ True,  True,  True, ...,  True,  True,  True]))
</pre></div></div>
</div>
<p>Many results for RNA could be used in calling cards analysis. Here, we borrow clustering information from scRNA-seq data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf.obs[[&#39;cluster&#39;]] = adata.obs[[&#39;cluster&#39;]]
adata_ccf.obs
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_566777/2786046817.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
  adata_ccf.obs[[&#39;cluster&#39;]] = adata.obs[[&#39;cluster&#39;]]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cluster</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCTGAGAACTCGG-1</th>
      <td>Neuron_Excit_L5_Mixed</td>
    </tr>
    <tr>
      <th>AAACCTGAGCCGTCGT-1</th>
      <td>Astrocyte</td>
    </tr>
    <tr>
      <th>AAACCTGAGTAGCGGT-1</th>
      <td>Astrocyte</td>
    </tr>
    <tr>
      <th>AAACCTGAGTGGAGTC-1</th>
      <td>Neuron_Inhibitory</td>
    </tr>
    <tr>
      <th>AAACCTGCAGCATGAG-1</th>
      <td>Astrocyte</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>TTTGTCAAGTCCCACG-1</th>
      <td>Neuron_Granule_DG</td>
    </tr>
    <tr>
      <th>TTTGTCACAGCGTCCA-1</th>
      <td>Neuron_Cajal-Retzius</td>
    </tr>
    <tr>
      <th>TTTGTCACATTTCACT-1</th>
      <td>Neuron_Granule_DG</td>
    </tr>
    <tr>
      <th>TTTGTCAGTCGCATCG-1</th>
      <td>Neuron_Granule_DG</td>
    </tr>
    <tr>
      <th>TTTGTCATCTTTACAC-1</th>
      <td>Neuron_Granule_DG</td>
    </tr>
  </tbody>
</table>
<p>30300 rows × 1 columns</p>
</div></div>
</div>
<p>Although one peak should have many insertions, but there is a chance that all the cells from the peak was filtered by the RNA preprocesssing. In this case, we advise to filter peaks by the minimum number of cells.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pp.filter_peaks(adata_ccf, min_counts=1)
</pre></div>
</div>
</div>
<p>Differential peak analysis shows significant bindings for each cluster. In this example, we use <em>binomial test</em> to find out.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.tl.rank_peak_groups(adata_ccf,&quot;cluster&quot;,method = &#39;binomtest&#39;,key_added = &#39;binomtest&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 18/18 [02:44&lt;00:00,  9.11s/it]
</pre></div></div>
</div>
<p>Plot the results for differential peak analysis.</p>
<p>Currently, the peaks are ranked by pvalues. It could also be ranked by logfoldchanges by the following codes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_ccf</span><span class="p">,</span><span class="s2">&quot;cluster&quot;</span><span class="p">,</span><span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">,</span><span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_ccf</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">,</span><span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.rank_peak_groups(adata_ccf, key = &#39;binomtest&#39;, save = True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_30_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_30_0.png" />
</div>
</div>
<p>Then, we take a look at the genome for the highly differentiated peaks. The colored ones are the insertions for specific cluster and the grey ones are the total insertions for the whole dataset. We could see that most of insertions are from Astrocyte in the following peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.draw_area(&quot;chr3&quot;,34638588,34656047,400000,peak_data,ccf_data,&quot;mm10&quot;,adata = adata_ccf,
                name = &#39;Astrocyte&#39;,key =&#39;cluster&#39;,figsize = (30,7),peak_line = 4,color = &quot;blue&quot;, title = &quot;chr3_34638588_34656047&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_32_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_32_0.png" />
</div>
</div>
<p>If the peak and the related refernce genes is also differential, we then see whether there is a relationship between the binding and gene experssion.</p>
<p>Let’s do the differential expression analysis for scRNA-seq data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.tl.rank_genes_groups(adata,&#39;cluster&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Default of the method has been changed to &#39;t-test&#39; from &#39;t-test_overestim_var&#39;
</pre></div></div>
</div>
<p>Find out the co-differential peak-gene pairs. It would look into all the differential peaks in each cluster and then see if the annotated genes are also significant expressed in ths cluster. We could set the pvalue and score/log foldchange cutoff easily.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>result = cc.tl.pair_peak_gene_sc(adata_ccf,adata,peak_annotation, pvalue_cutoff_ccf = 0.001, pvalue_cutoff_rna = 0.001, lfc_cutoff = 3, score_cutoff = 3)
result
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cluster</th>
      <th>Peak</th>
      <th>Logfoldchanges</th>
      <th>Pvalue_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Astrocyte</td>
      <td>chr16_43501178_43518253</td>
      <td>3.077262</td>
      <td>1.462410e-16</td>
      <td>Zbtb20</td>
      <td>79.234276</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Astrocyte</td>
      <td>chr8_64645834_64659215</td>
      <td>4.623955</td>
      <td>3.114365e-14</td>
      <td>Msmo1</td>
      <td>61.964111</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Astrocyte</td>
      <td>chr3_141928409_141939733</td>
      <td>4.876938</td>
      <td>3.786296e-14</td>
      <td>Bmpr1b</td>
      <td>24.131538</td>
      <td>2.475603e-121</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>Nfia</td>
      <td>45.417362</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>E130114P18Rik</td>
      <td>16.893749</td>
      <td>3.380841e-62</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>122</th>
      <td>Neuron_Granule_DG</td>
      <td>chr13_89320933_89321928</td>
      <td>4.399915</td>
      <td>7.818271e-04</td>
      <td>Hapln1</td>
      <td>-14.765965</td>
      <td>9.061876e-44</td>
    </tr>
    <tr>
      <th>123</th>
      <td>Neuron_Inhibitory</td>
      <td>chr2_34362478_34369886</td>
      <td>3.988165</td>
      <td>1.977978e-04</td>
      <td>Pbx3</td>
      <td>15.902434</td>
      <td>5.024489e-54</td>
    </tr>
    <tr>
      <th>124</th>
      <td>Neuron_Inhibitory</td>
      <td>chr6_15196298_15198126</td>
      <td>10.414956</td>
      <td>3.822503e-04</td>
      <td>Foxp2</td>
      <td>20.337471</td>
      <td>1.781142e-84</td>
    </tr>
    <tr>
      <th>125</th>
      <td>Vascular_meningeal</td>
      <td>chr14_122487516_122490502</td>
      <td>7.469697</td>
      <td>1.026983e-07</td>
      <td>Zic2</td>
      <td>7.556187</td>
      <td>1.062384e-12</td>
    </tr>
    <tr>
      <th>126</th>
      <td>Vascular_meningeal</td>
      <td>chr9_91959431_91963327</td>
      <td>7.047598</td>
      <td>3.246923e-04</td>
      <td>Plscr1</td>
      <td>4.303550</td>
      <td>2.515998e-05</td>
    </tr>
  </tbody>
</table>
<p>127 rows × 7 columns</p>
</div></div>
</div>
<p>Draw the above results to find out potential relationship between TF bindings and gene expression.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.dotplot_sc(adata_ccf,adata,result)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_39_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_39_0.png" />
</div>
</div>
<p>After seeing the dotplot above, grab some peaks to the RNA Anndata object and see the distribution in UMAP plot.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.pca(adata)
sc.pp.neighbors(adata)
sc.tl.umap(adata)
</pre></div>
</div>
</div>
<p>The first plot is the average insertions of peak in each cluster, the second plot is the gene expression Gou3f2 (one of its nearest genes for the peak) and the third plot is the cluster information.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>p = &quot;chr4_22969921_22973019&quot;
adata = cc.pp.adata_insertions(adata_ccf,adata,p,method=&quot;logAvginsertions&quot; ,peak = p)
sc.pl.umap(adata, color=[p,&#39;Pou3f2&#39;,&#39;cluster&#39;])
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/opt/apps/labs/rmlab/software/juanru/miniconda/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_43_1.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_43_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>p = &#39;chr1_42253068_42265192&#39;
adata = cc.pp.adata_insertions(adata_ccf,adata,p,method=&quot;logAvginsertions&quot; ,peak = p)
sc.pl.umap(adata, color=[p,&#39;Pou3f3&#39;,&#39;cluster&#39;])
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/opt/apps/labs/rmlab/software/juanru/miniconda/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_44_1.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_44_1.png" />
</div>
</div>
<p>We could see the potential relationship between bindings and gene expression.</p>
<p>We then plot the scankey plot of the result to see the potential casual result for binding and gene expression. Please open sankey.htm in your folder to see the result.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cc.pl.sankey(result, figsize = [1500, 1500])
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[d3blocks] &gt;INFO&gt; Cleaning edge_properties and config parameters..
[d3blocks] &gt;INFO&gt; Initializing [Sankey]
[d3blocks] &gt;INFO&gt; Convert to Frame.
[d3blocks] &gt;INFO&gt; Node properties are set.
[d3blocks] &gt;INFO&gt; Edge properties are set.
[d3blocks] &gt;INFO&gt; Set [filepath]: sankey.html
[d3blocks] &gt;INFO&gt; Set [figsize]: [1500, 1500]
[d3blocks] &gt;INFO&gt; File already exists and will be overwritten: [sankey.html]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
This function use D3Blocks. Please make sure D3Blocks is installed and result would be saved as html file.
</pre></div></div>
</div>
<p>Save the file if needed.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ccf.write(&quot;Mouse-Cortex_CCF.h5ad&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;Chr&#39; as categorical
... storing &#39;Nearest Refseq1&#39; as categorical
... storing &#39;Gene Name1&#39; as categorical
... storing &#39;Nearest Refseq2&#39; as categorical
... storing &#39;Gene Name2&#39; as categorical
</pre></div></div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="K562HCT116_SP1.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tutorial: K562 HCT116 SP1 single-cell callingcards data.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorials</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script integrity="sha384-vtXRMe3mGCbOeY7l30aIg8H9p3GdeSe4IFlP6G8JMa7o7lXvnz3GFKzPxzJdPfGK" src="../../_static/jquery.js"></script>
    <script integrity="sha384-lSZeSIVKp9myfKbDQ3GkN/KHjUc+mzg17VKDN4Y2kUeBSJioB9QSM639vM9fuY//" src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/scripts/furo.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>