References
----------

.. [Moudgil20] Moudgil *et al.* (2020),
   *Self-Reporting Transposons Enable Simultaneous Readout of Gene Expression and Transcription Factor Binding in Single Cells*,
   `Cell <https://doi.org/10.1016/j.cell.2020.06.037>`__.

.. [Virshup21]  Virshup *et al.* (2021),
   *anndata: Annotated data*,
   `bioRxiv <https://www.biorxiv.org/content/10.1101/2021.12.16.473007v1>`__.
   
.. [Wolf18] Wolf *et al.* (2018),
   *SCANPY: large-scale single-cell gene expression data analysis*,
   `Genome Biology <https://doi.org/10.1186/s13059-017-1382-0>`__.

.. [Zhang08] Zhang *et al.* (2008),
   *Model-based Analysis of ChIP-Seq (MACS)*,
   `Genome Biology <https://doi.org/10.1186/gb-2008-9-9-r137>`__.

.. [Quinlan10] Quinlan *et al.* (2010),
   *BEDTools: a flexible suite of utilities for comparing genomic featuress*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btq033>`__.

.. [Heinz10]  Heinz *et al.* (2010),
   *Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities*,
   `Mol Cell <https://doi.org/10.1016/j.molcel.2010.05.004>`__.

.. [Kfoury21] Kfoury *et al.* (2021),
   *Brd4-bound enhancers drive cell-intrinsic sex differences in glioblastoma*,
   `PNAS <https://doi.org/10.1073/pnas.2017148118>`__.

.. [Cammack19] Cammack *et al.* (2019),
   *A viral toolkit for recording transcription factorâ€“DNA interactions in live mouse tissues*,
   `PNAS <https://doi.org/10.1073/pnas.1918241117>`__.




