<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data." />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pycallingcards.readthedocs.io/en/latest/tutorials/notebooks/Brd4_bulk.html" />
<meta property="og:site_name" content="Pycallingcards" />
<meta property="og:description" content="We will analyze bulk calling cards datasets from Kfoury et al., PNAS. (2021), which can be downloaded from GEO. These experiments were designed to study differences in Brd4 binding between male and..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="https://pycallingcards.readthedocs.io/en/latest/index.html/_images/social_previews/summary_tutorials_notebooks_Brd4_bulk_b5ef81fd.png" />
<meta property="og:image:alt" content="We will analyze bulk calling cards datasets from Kfoury et al., PNAS. (2021), which can be downloaded from GEO. These experiments were designed to study diff..." />
<meta name="description" content="We will analyze bulk calling cards datasets from Kfoury et al., PNAS. (2021), which can be downloaded from GEO. These experiments were designed to study differences in Brd4 binding between male and..." />
<meta name="twitter:card" content="summary_large_image" />
<link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Tutorial: SP1 bindings in Cre-driver mouse lines." href="SP1_bulk.html" /><link rel="prev" title="Tutorial: K562 HCT116 SP1 single-cell calling cards data." href="K562HCT116_SP1.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.08.06 -->
        <title>Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data. - pycallingcards-tools</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pycallingcards-tools</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Mouse_cortex_Example.html">Tutorial: Analyzing single cell calling cards collected from the mouse cortex</a></li>
<li class="toctree-l2"><a class="reference internal" href="K562HCT116_SP1.html">Tutorial: K562 HCT116 SP1 single-cell calling cards data.</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="SP1_bulk.html">Tutorial: SP1 bindings in Cre-driver mouse lines.</a></li>
<li class="toctree-l2"><a class="reference internal" href="yeast.html">Tutorial: Yeast TYE7 bulk calling cards data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Datasets</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mousecortex_data.html">pycallingcards.datasets.mousecortex_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_K562HCT116_data.html">pycallingcards.datasets.SP1_K562HCT116_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mouse_brd4_data.html">pycallingcards.datasets.mouse_brd4_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_Cre_data.html">pycallingcards.datasets.SP1_Cre_data</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/reading.html">Reading</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Reading</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_qbed.html">pycallingcards.reading.read_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_rna.html">pycallingcards.reading.read_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_barbodes.html">pycallingcards.reading.read_barbodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.combine_qbed.html">pycallingcards.reading.combine_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_qbed.html">pycallingcards.reading.save_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_bed.html">pycallingcards.reading.save_bed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/preprocessing.html">Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Preprocessing</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.clean_qbed.html">pycallingcards.preprocessing.clean_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.call_peaks.html">pycallingcards.preprocessing.call_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.combine_peaks.html">pycallingcards.preprocessing.combine_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.separate_peaks.html">pycallingcards.preprocessing.separate_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.annotation.html">pycallingcards.preprocessing.annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.make_Anndata.html">pycallingcards.preprocessing.make_Anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.filter_peaks.html">pycallingcards.preprocessing.filter_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.adata_insertions.html">pycallingcards.preprocessing.adata_insertions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.down_sample.html">pycallingcards.preprocessing.down_sample</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/tools.html">Tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of Tools</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.call_motif.html">pycallingcards.tools.call_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.compare_motif.html">pycallingcards.tools.compare_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups.html">pycallingcards.tools.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups_mu.html">pycallingcards.tools.rank_peak_groups_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups_df.html">pycallingcards.tools.rank_peak_groups_df</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bygroup.html">pycallingcards.tools.diff2group_bygroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bysample.html">pycallingcards.tools.diff2group_bysample</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.DE_pvalue.html">pycallingcards.tools.DE_pvalue</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_bulk.html">pycallingcards.tools.pair_peak_gene_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc.html">pycallingcards.tools.pair_peak_gene_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc_mu.html">pycallingcards.tools.pair_peak_gene_sc_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.liftover.html">pycallingcards.tools.liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.find_location.html">pycallingcards.tools.find_location</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.result_mapping.html">pycallingcards.tools.result_mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS.html">pycallingcards.tools.GWAS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS_adata_bulk.html">pycallingcards.tools.GWAS_adata_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS_adata_sc.html">pycallingcards.tools.GWAS_adata_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.footprint.html">pycallingcards.tools.footprint</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.find_fastq.html">pycallingcards.tools.find_fastq</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of Plotting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area.html">pycallingcards.plotting.draw_area</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area_mu.html">pycallingcards.plotting.draw_area_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.whole_peaks.html">pycallingcards.plotting.whole_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.calculate_signal.html">pycallingcards.plotting.calculate_signal</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_plot.html">pycallingcards.plotting.signal_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_heatmap.html">pycallingcards.plotting.signal_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.rank_peak_groups.html">pycallingcards.plotting.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.plot_matched.html">pycallingcards.plotting.plot_matched</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.volcano_plot.html">pycallingcards.plotting.volcano_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.heatmap.html">pycallingcards.plotting.heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_bulk.html">pycallingcards.plotting.dotplot_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc.html">pycallingcards.plotting.dotplot_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc_mu.html">pycallingcards.plotting.dotplot_sc_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.WashU_browser_url.html">pycallingcards.plotting.WashU_browser_url</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.GWAS_adata_sc.html">pycallingcards.plotting.GWAS_adata_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.GWAS_adata_bulk.html">pycallingcards.plotting.GWAS_adata_bulk</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../../_sources/tutorials/notebooks/Brd4_bulk.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="Tutorial:-Brd4-bindings-for-Male/Female-mouse-bulk-calling-cards-data.">
<h1>Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data.<a class="headerlink" href="#Tutorial:-Brd4-bindings-for-Male/Female-mouse-bulk-calling-cards-data." title="Permalink to this heading">¶</a></h1>
<p>We will analyze bulk calling cards datasets from <a class="reference external" href="https://www.pnas.org/doi/10.1073/pnas.2017148118">Kfoury et al., PNAS. (2021)</a>, which can be downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE156821">GEO</a>. These experiments were designed to study differences in Brd4 binding between male and female cells in a genetically engineered cellular model of glioblastoma (GBM). These nearly-isogenic cells were derived from murine neocortical postnatal day 1 (p1) astrocytes
(male and female GBM astrocytes) and were engineered to contain a combined loss of neurofibromin (NF1) and p53 function. These cells were previously shown to display sex differences in important GBM phenotypes in vivo, including proliferation, clonogenic stem-like frequency, in vivo tumorigenesis, cell cycle regulation, gene expression, and chemotherapy response, that mimic those observed in GBM patients. For more details see [Kfoury et al., PNAS. (2021)].</p>
<p>In this tutorial, we will call peaks, annotate these peaks, perform differential peak analysis, and pair differentially bound genomic regions with nearby genes that are also differentially expressed between the sexes. There are 668525 insertions in female qbed data and 613728 insertions in male qbed data. For bulk RNA-seq data, there are 6 samples in total: 3 female samples and 3 male samples. It creates a group by peak anndata for calling cards data. If you want to create a replicate by peak
anndata only, please check <a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards_data/blob/main/Brd4_bulk-replicate.ipynb">Github</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pycallingcards</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">150</span>
</pre></div>
</div>
</div>
<p>We start by reading qbed data. In these data, each row provides information about one Brd4-directed insertion organized into columns to indicate the chromosome, start point, end point, number of reads supporting the insertion, and the orientation and barcode for each insertion. For example, in the file displayed below, the first row tells us that the first insertion occured into a TTAA on Chromosome 1, whose coordinates start location 3478112 and end at 3478115. There was 1 read supporting this
insertion and the transposon is orientated so that the plus strand of the transposon is on the plus strand of the genome. The transposon barcode is TTTGTCCAA. We add another column indecating its respected group.</p>
<p>Use <code class="docutils literal notranslate"><span class="pre">cc.rd.read_qbed(filename)</span></code> to load your own qbed data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Female_Brd4</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">mouse_brd4_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;Female_Brd4&quot;</span><span class="p">)</span>
<span class="n">Female_Brd4</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Female_Brd4&quot;</span>
<span class="n">Female_Brd4</span>
<span class="c1">#load female CC data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3478111</td>
      <td>3478115</td>
      <td>1</td>
      <td>+</td>
      <td>TTTGTCCAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3481743</td>
      <td>3481747</td>
      <td>1</td>
      <td>+</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3481957</td>
      <td>3481961</td>
      <td>1</td>
      <td>-</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3493161</td>
      <td>3493165</td>
      <td>16</td>
      <td>+</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3493702</td>
      <td>3493706</td>
      <td>1</td>
      <td>+</td>
      <td>CGTTACACA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1290214</th>
      <td>chrY</td>
      <td>90802841</td>
      <td>90802845</td>
      <td>1</td>
      <td>-</td>
      <td>CGTTACACA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>1290215</th>
      <td>chrY</td>
      <td>90803579</td>
      <td>90803583</td>
      <td>1</td>
      <td>-</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>1290216</th>
      <td>chrY</td>
      <td>90803579</td>
      <td>90803583</td>
      <td>1</td>
      <td>-</td>
      <td>TGTCGTGCA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>1290217</th>
      <td>chrY</td>
      <td>90804429</td>
      <td>90804433</td>
      <td>1</td>
      <td>-</td>
      <td>TATACTCTA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>1290218</th>
      <td>chrY</td>
      <td>90805130</td>
      <td>90805134</td>
      <td>1</td>
      <td>-</td>
      <td>GACGGCTCC</td>
      <td>Female_Brd4</td>
    </tr>
  </tbody>
</table>
<p>1290219 rows × 7 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Male_Brd4</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">mouse_brd4_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;Male_Brd4&quot;</span><span class="p">)</span>
<span class="n">Male_Brd4</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Male_Brd4&quot;</span>
<span class="n">Male_Brd4</span>
<span class="c1">#load male CC data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3493161</td>
      <td>3493165</td>
      <td>1</td>
      <td>-</td>
      <td>GAGGTACAG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3493275</td>
      <td>3493279</td>
      <td>1</td>
      <td>+</td>
      <td>TACTTTCCG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3493341</td>
      <td>3493345</td>
      <td>2</td>
      <td>+</td>
      <td>AGCACAGTG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3493341</td>
      <td>3493345</td>
      <td>10</td>
      <td>+</td>
      <td>TGTCCATTG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3493341</td>
      <td>3493345</td>
      <td>8</td>
      <td>+</td>
      <td>ATGAAAGCA</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1185697</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>13</td>
      <td>-</td>
      <td>TACTTTCCG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>1185698</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>9</td>
      <td>-</td>
      <td>CACGGCGGA</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>1185699</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>17</td>
      <td>-</td>
      <td>GATCTTATG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>1185700</th>
      <td>chrY</td>
      <td>90803774</td>
      <td>90803778</td>
      <td>4</td>
      <td>-</td>
      <td>ATGAAAGCA</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>1185701</th>
      <td>chrY</td>
      <td>90806562</td>
      <td>90806566</td>
      <td>1</td>
      <td>-</td>
      <td>AGCACAGTG</td>
      <td>Male_Brd4</td>
    </tr>
  </tbody>
</table>
<p>1185702 rows × 7 columns</p>
</div></div>
</div>
<p>Since we are interested in differential peak calling, we now combine the two datasets to call peaks on the joint dataset:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Brd4</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">rd</span><span class="o">.</span><span class="n">combine_qbed</span><span class="p">([</span><span class="n">Female_Brd4</span><span class="p">,</span> <span class="n">Male_Brd4</span><span class="p">])</span>
<span class="n">Brd4</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3478111</td>
      <td>3478115</td>
      <td>1</td>
      <td>+</td>
      <td>TTTGTCCAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3481743</td>
      <td>3481747</td>
      <td>1</td>
      <td>+</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3481957</td>
      <td>3481961</td>
      <td>1</td>
      <td>-</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3493161</td>
      <td>3493165</td>
      <td>16</td>
      <td>+</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3493161</td>
      <td>3493165</td>
      <td>1</td>
      <td>-</td>
      <td>GAGGTACAG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2475916</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>17</td>
      <td>-</td>
      <td>GATCTTATG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2475917</th>
      <td>chrY</td>
      <td>90803774</td>
      <td>90803778</td>
      <td>4</td>
      <td>-</td>
      <td>ATGAAAGCA</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2475918</th>
      <td>chrY</td>
      <td>90804429</td>
      <td>90804433</td>
      <td>1</td>
      <td>-</td>
      <td>TATACTCTA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>2475919</th>
      <td>chrY</td>
      <td>90805130</td>
      <td>90805134</td>
      <td>1</td>
      <td>-</td>
      <td>GACGGCTCC</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>2475920</th>
      <td>chrY</td>
      <td>90806562</td>
      <td>90806566</td>
      <td>1</td>
      <td>-</td>
      <td>AGCACAGTG</td>
      <td>Male_Brd4</td>
    </tr>
  </tbody>
</table>
<p>2475921 rows × 7 columns</p>
</div></div>
</div>
<p>Optional: filter the insertions with reads greater than 2. Because we use strict quality control on read insertions, even insertions supported by 1 read are likely to be bona fide.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Brd4</span> <span class="o">=</span> <span class="n">Brd4</span><span class="p">[</span><span class="n">Brd4</span><span class="p">[</span><span class="s1">&#39;Reads&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">]</span>
<span class="n">Brd4</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
      <th>group</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3493161</td>
      <td>3493165</td>
      <td>16</td>
      <td>+</td>
      <td>TATGTACAA</td>
      <td>Female_Brd4</td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr1</td>
      <td>3493341</td>
      <td>3493345</td>
      <td>10</td>
      <td>+</td>
      <td>TGTCCATTG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr1</td>
      <td>3493341</td>
      <td>3493345</td>
      <td>8</td>
      <td>+</td>
      <td>ATGAAAGCA</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr1</td>
      <td>3493643</td>
      <td>3493647</td>
      <td>3</td>
      <td>+</td>
      <td>TCTACTGCC</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr1</td>
      <td>3514553</td>
      <td>3514557</td>
      <td>7</td>
      <td>+</td>
      <td>TACTTTCCG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2475905</th>
      <td>chrY</td>
      <td>42512546</td>
      <td>42512550</td>
      <td>3</td>
      <td>-</td>
      <td>ACGCGCTGT</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2475914</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>13</td>
      <td>-</td>
      <td>TACTTTCCG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2475915</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>9</td>
      <td>-</td>
      <td>CACGGCGGA</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2475916</th>
      <td>chrY</td>
      <td>90803733</td>
      <td>90803737</td>
      <td>17</td>
      <td>-</td>
      <td>GATCTTATG</td>
      <td>Male_Brd4</td>
    </tr>
    <tr>
      <th>2475917</th>
      <td>chrY</td>
      <td>90803774</td>
      <td>90803778</td>
      <td>4</td>
      <td>-</td>
      <td>ATGAAAGCA</td>
      <td>Male_Brd4</td>
    </tr>
  </tbody>
</table>
<p>1229664 rows × 7 columns</p>
</div></div>
</div>
<p>Because insertions are discrete, we now need to call peaks to deduce potential binding sites. Three different methods (<em>CCcaller, MACCs, Blockify</em>) are available along with three different species (<em>hg38, mm10, sacCer3</em>).</p>
<p>In this setting, we use <strong>CCcaller</strong> in <em>mouse</em>(‘mm10’) data. <em>maxbetween</em> is the most important parameter for <em>CCcaller</em>. It controls the maximum distance between two nearby insertions, or in another words, the minimum distance between two peaks. We recommend using a value between 800-1200 to set <em>maxbetween</em>. <em>pvalue_cutoff</em> is also an important parameter and a number 0.0001-0.01 is strongly advised. The setting of <em>pseudocounts</em> is largely influenced by library size. For the first time of
trial, it can be adjusted to <span class="math notranslate nohighlight">\(10^{6}-10^{-5} \times\)</span> the number of insertions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_data</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">call_peaks</span><span class="p">(</span><span class="n">Brd4</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;CCcaller&quot;</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">pvalue_cutoff</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span>
                             <span class="n">maxbetween</span> <span class="o">=</span> <span class="mi">1100</span><span class="p">,</span> <span class="n">lam_win_size</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span> <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
                             <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="s2">&quot;peak.bed&quot;</span><span class="p">)</span>
<span class="n">peak_data</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the CCcaller method without background, [expdata, reference, pvalue_cutoff, lam_win_size, pseudocounts, minlen, extend, maxbetween, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 21/21 [00:21&lt;00:00,  1.00s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
      <th>pvalue_adj</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>4196775</td>
      <td>4198714</td>
      <td>16</td>
      <td>12</td>
      <td>0.142389</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>4785068</td>
      <td>4786551</td>
      <td>17</td>
      <td>15</td>
      <td>0.572576</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>4806673</td>
      <td>4807730</td>
      <td>21</td>
      <td>12</td>
      <td>0.477937</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>4856928</td>
      <td>4862019</td>
      <td>104</td>
      <td>44</td>
      <td>1.490080</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>4912814</td>
      <td>4916593</td>
      <td>7</td>
      <td>22</td>
      <td>0.791756</td>
      <td>1.900897e-06</td>
      <td>1.024407e-04</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>24546</th>
      <td>chrX</td>
      <td>169936677</td>
      <td>169937745</td>
      <td>7</td>
      <td>8</td>
      <td>1.189526</td>
      <td>3.479200e-05</td>
      <td>1.673495e-03</td>
    </tr>
    <tr>
      <th>24547</th>
      <td>chrY</td>
      <td>897430</td>
      <td>899700</td>
      <td>11</td>
      <td>24</td>
      <td>0.528165</td>
      <td>6.046275e-13</td>
      <td>5.162843e-11</td>
    </tr>
    <tr>
      <th>24548</th>
      <td>chrY</td>
      <td>1009018</td>
      <td>1011799</td>
      <td>27</td>
      <td>25</td>
      <td>0.489073</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>24549</th>
      <td>chrY</td>
      <td>1243715</td>
      <td>1246316</td>
      <td>24</td>
      <td>29</td>
      <td>0.504932</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>24550</th>
      <td>chrY</td>
      <td>1282482</td>
      <td>1287504</td>
      <td>19</td>
      <td>60</td>
      <td>0.945958</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
    </tr>
  </tbody>
</table>
<p>24551 rows × 8 columns</p>
</div></div>
</div>
<p>In order to tune parameters for peak calling, we advise looking at the data and evaluating the validity of the called peaks. The default settings are what we recommend, but for some TFs adjacent peaks may be merged that should not be, or, alternatively, peaks that should be joined may be called separately.</p>
<p>Below, we plot the combined calling card data in males and females for a region in chromosome 1. The top track displays insertion locations and their read counts. Each dot is an insertion and the height is log(reads+1). The middle track plots the insertion density. The thrid track represents the reference genes and peaks. Finally, the last track represents peak calls. Below you can see that regions with high densities of insertions are accurately called as Brd4 binding sites.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">4856929</span><span class="p">,</span> <span class="mi">4863861</span><span class="p">,</span> <span class="mi">100000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">Brd4</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">font_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">example_length</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_13_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_13_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr2&quot;</span><span class="p">,</span> <span class="mi">6136575</span><span class="p">,</span> <span class="mi">6144043</span><span class="p">,</span> <span class="mi">75000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">Brd4</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">font_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">example_length</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_14_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_14_0.png" />
</div>
</div>
<p>We can also visualize our data directly in the <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a>. This can be useful for overlaying your data with other published datasets. Notice that this link only valid for 24hrs, so please rerun it if you want to use it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">qbed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Brd4&quot;</span><span class="p">:</span> <span class="n">Brd4</span><span class="p">}</span>
<span class="n">bed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;peak&quot;</span><span class="p">:</span> <span class="n">peak_data</span><span class="p">}</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">WashU_browser_url</span><span class="p">(</span><span class="n">qbed</span> <span class="o">=</span> <span class="n">qbed</span><span class="p">,</span> <span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span><span class="p">,</span> <span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;mm10&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=mm10&amp;hub=https://companion.epigenomegateway.org//task/544b112245ab68df385f75da9d4bfd8b/output//datahub.json
</pre></div></div>
</div>
<p>Pycallingcards can be used to visual peak locations acorss the genome to see if the distribution of peaks is unbiased and all chromosomes are represented.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">whole_peaks</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_18_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_18_0.png" />
</div>
</div>
<p>For differential peak calling, we recommend first combining the data and calling peaks together (and then splitting the data and looking for enrichments under the jointly called peaks). However, it is often of interest to call peaks in each sample to analyze them separately or to then merge these peak calls using <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>. Below is the code to do this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pybedtools</span>
<span class="n">peak_data1</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">call_peaks</span><span class="p">(</span><span class="n">Female_Brd4</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;CCcaller&quot;</span><span class="p">,</span>
                              <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">pvalue_cutoffbg</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
                              <span class="n">maxbetween</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">pvalue_cutoffTTAA</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
                              <span class="n">lam_win_size</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span>  <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">peak_data2</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">call_peaks</span><span class="p">(</span><span class="n">Male_Brd4</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;CCcaller&quot;</span><span class="p">,</span>
                              <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span>  <span class="n">pvalue_cutoffbg</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
                              <span class="n">maxbetween</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">pvalue_cutoffTTAA</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
                              <span class="n">lam_win_size</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span> <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">rd</span><span class="o">.</span><span class="n">combine_qbed</span><span class="p">([</span><span class="n">peak_data1</span><span class="p">,</span> <span class="n">peak_data2</span><span class="p">])</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="o">.</span><span class="n">from_dataframe</span><span class="p">(</span><span class="n">peak</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">peak_data</span> <span class="o">=</span> <span class="n">peak</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chrom&quot;</span><span class="p">:</span><span class="s2">&quot;Chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span><span class="s2">&quot;End&quot;</span><span class="p">})</span>
</pre></div>
</div>
<p>In the next step, we annote each peak to annotate using <a class="reference external" href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>. Make sure they are all previously installed before using.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_annotation</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">annotation</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">)</span>
<span class="n">peak_annotation</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">combine_annotation</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">peak_annotation</span><span class="p">)</span>
<span class="n">peak_annotation</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In the bedtools method, we would use bedtools in the default path. Set bedtools path by &#39;bedtools_path&#39; if needed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
      <th>pvalue_adj</th>
      <th>Nearest Refseq1</th>
      <th>Gene Name1</th>
      <th>Direction1</th>
      <th>Distance1</th>
      <th>Nearest Refseq2</th>
      <th>Gene Name2</th>
      <th>Direction2</th>
      <th>Distance2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>4196775</td>
      <td>4198714</td>
      <td>16</td>
      <td>12</td>
      <td>0.142389</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NM_001195662</td>
      <td>Rp1</td>
      <td>-</td>
      <td>92132</td>
      <td>NM_011283</td>
      <td>Rp1</td>
      <td>-</td>
      <td>144793</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>4785068</td>
      <td>4786551</td>
      <td>17</td>
      <td>15</td>
      <td>0.572576</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NR_033530</td>
      <td>Mrpl15</td>
      <td>-</td>
      <td>0</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
      <td>+</td>
      <td>21342</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>4806673</td>
      <td>4807730</td>
      <td>21</td>
      <td>12</td>
      <td>0.477937</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
      <td>+</td>
      <td>163</td>
      <td>NR_033530</td>
      <td>Mrpl15</td>
      <td>-</td>
      <td>-20948</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>4856928</td>
      <td>4862019</td>
      <td>104</td>
      <td>44</td>
      <td>1.490080</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NM_011541</td>
      <td>Tcea1</td>
      <td>+</td>
      <td>0</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
      <td>+</td>
      <td>-10194</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>4912814</td>
      <td>4916593</td>
      <td>7</td>
      <td>22</td>
      <td>0.791756</td>
      <td>1.900897e-06</td>
      <td>1.024407e-04</td>
      <td>NM_001290372</td>
      <td>Rgs20</td>
      <td>-</td>
      <td>0</td>
      <td>NM_011541</td>
      <td>Tcea1</td>
      <td>+</td>
      <td>-14906</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>24546</th>
      <td>chrX</td>
      <td>169936677</td>
      <td>169937745</td>
      <td>7</td>
      <td>8</td>
      <td>1.189526</td>
      <td>3.479200e-05</td>
      <td>1.673495e-03</td>
      <td>NM_010797</td>
      <td>Mid1</td>
      <td>+</td>
      <td>0</td>
      <td>NR_029457</td>
      <td>G530011O06Rik</td>
      <td>-</td>
      <td>37298</td>
    </tr>
    <tr>
      <th>24547</th>
      <td>chrY</td>
      <td>897430</td>
      <td>899700</td>
      <td>11</td>
      <td>24</td>
      <td>0.528165</td>
      <td>6.046275e-13</td>
      <td>5.162843e-11</td>
      <td>NM_011419</td>
      <td>Kdm5d</td>
      <td>+</td>
      <td>0</td>
      <td>NM_011667</td>
      <td>Uba1y</td>
      <td>+</td>
      <td>-53207</td>
    </tr>
    <tr>
      <th>24548</th>
      <td>chrY</td>
      <td>1009018</td>
      <td>1011799</td>
      <td>27</td>
      <td>25</td>
      <td>0.489073</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NM_012011</td>
      <td>Eif2s3y</td>
      <td>+</td>
      <td>0</td>
      <td>NR_027507</td>
      <td>Tspy-ps</td>
      <td>-</td>
      <td>43965</td>
    </tr>
    <tr>
      <th>24549</th>
      <td>chrY</td>
      <td>1243715</td>
      <td>1246316</td>
      <td>24</td>
      <td>29</td>
      <td>0.504932</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NM_009484</td>
      <td>Uty</td>
      <td>-</td>
      <td>0</td>
      <td>NM_012008</td>
      <td>Ddx3y</td>
      <td>-</td>
      <td>14399</td>
    </tr>
    <tr>
      <th>24550</th>
      <td>chrY</td>
      <td>1282482</td>
      <td>1287504</td>
      <td>19</td>
      <td>60</td>
      <td>0.945958</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>NM_012008</td>
      <td>Ddx3y</td>
      <td>-</td>
      <td>0</td>
      <td>NM_148943</td>
      <td>Usp9y</td>
      <td>-</td>
      <td>11457</td>
    </tr>
  </tbody>
</table>
<p>24551 rows × 16 columns</p>
</div></div>
</div>
<p>Use qbed data, peak data, and barcode data to make a group by peak <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">Anndata</a> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">make_Anndata</span><span class="p">(</span><span class="n">Brd4</span><span class="p">,</span> <span class="n">peak_annotation</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Female_Brd4&quot;</span><span class="p">,</span> <span class="s2">&quot;Male_Brd4&quot;</span><span class="p">],</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;group&#39;</span><span class="p">)</span>
<span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 21/21 [00:07&lt;00:00,  2.64it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2 × 24551
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;pvalue_adj&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
    <tr>
      <th>Index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Female_Brd4</th>
    </tr>
    <tr>
      <th>Male_Brd4</th>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2 × 24551
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;pvalue_adj&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">liftover</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">)</span>
<span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 24551/24551 [04:44&lt;00:00, 86.27it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2 × 24551
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;pvalue_adj&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;, &#39;Chr_liftover&#39;, &#39;Start_liftover&#39;, &#39;End_liftover&#39;
</pre></div></div>
</div>
<p>Differential peak analysis will find out the significant binding for each group. In this example, we use the <em>Fisher’s exact test</em> to find out.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2/2 [01:43&lt;00:00, 51.87s/it]
</pre></div></div>
</div>
<p>Plot the results for differential peak analysis.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span>
                       <span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span><span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span> <span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;pvalues&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_30_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_30_0.png" />
</div>
</div>
<p>We can use Pycallingcards to look at the insertion densities at peaks that are differentially bound. In the plot below, we first plot female peaks as blue points and male as grey points, and then we plot male peaks as red and female peaks as grey. From the plots below we can see there is much more Brd4 binding at this locus in females than in males.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr9&quot;</span><span class="p">,</span> <span class="mi">91366534</span><span class="p">,</span> <span class="mi">91425373</span><span class="p">,</span> <span class="mi">50000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">Brd4</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Female_Brd4&quot;</span><span class="p">,</span><span class="n">key</span> <span class="o">=</span> <span class="s2">&quot;Index&quot;</span><span class="p">,</span><span class="n">insertionkey</span> <span class="o">=</span> <span class="s2">&quot;group&quot;</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;Female Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;Female Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">12</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Female&quot;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr9&quot;</span><span class="p">,</span> <span class="mi">91366534</span><span class="p">,</span> <span class="mi">91425373</span><span class="p">,</span> <span class="mi">50000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">Brd4</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span>  <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Male_Brd4&quot;</span><span class="p">,</span><span class="n">key</span> <span class="o">=</span> <span class="s2">&quot;Index&quot;</span><span class="p">,</span> <span class="n">insertionkey</span> <span class="o">=</span> <span class="s2">&quot;group&quot;</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;Male Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;Male Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">12</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Male&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_32_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_32_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_32_1.png" src="../../_images/tutorials_notebooks_Brd4_bulk_32_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr12&quot;</span><span class="p">,</span> <span class="mi">56516453</span><span class="p">,</span> <span class="mi">56538107</span><span class="p">,</span> <span class="mi">50000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">Brd4</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Female_Brd4&quot;</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s2">&quot;Index&quot;</span><span class="p">,</span><span class="n">insertionkey</span> <span class="o">=</span> <span class="s2">&quot;group&quot;</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;Female Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;Female Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Female&quot;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr12&quot;</span><span class="p">,</span> <span class="mi">56516453</span><span class="p">,</span> <span class="mi">56538107</span><span class="p">,</span> <span class="mi">50000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">Brd4</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Male_Brd4&quot;</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s2">&quot;Index&quot;</span><span class="p">,</span> <span class="n">insertionkey</span> <span class="o">=</span> <span class="s2">&quot;group&quot;</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;Male Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;Male Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Male&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_33_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_33_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_33_1.png" src="../../_images/tutorials_notebooks_Brd4_bulk_33_1.png" />
</div>
</div>
<p>We can now plot a volcano plot to visualize the distribution of differentially bound peaks. labelright and labellest are the location points of the labels.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">volcano_plot</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">pvalue_name</span> <span class="o">=</span> <span class="s1">&#39;pvalues_adj&#39;</span><span class="p">,</span> <span class="n">pvalue_cutoff</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">lfc_cutoff</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
                   <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">labelright</span> <span class="o">=</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">250</span><span class="p">),</span> <span class="n">labelleft</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">9</span><span class="p">,</span><span class="mi">250</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/Desktop/pycallingcards/pycallingcards/pycallingcards/plotting/_volcano.py:70: RuntimeWarning: divide by zero encountered in log10
  pva = -np.log10(np.array(adata_cc.uns[name][pvalue_name].tolist())[:, 1])
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_35_1.png" src="../../_images/tutorials_notebooks_Brd4_bulk_35_1.png" />
</div>
</div>
<p>We can now use <a class="reference external" href="http://homer.ucsd.edu/homer/motif/">HOMER</a> to call motifs for both female and male bound peaks separately to find out potential motif binding difference.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">call_motif</span><span class="p">(</span><span class="n">peaks_frame</span><span class="o">=</span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups_df</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span> <span class="n">pval_cutoff</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">logfc_min</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">group</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Female_Brd4&#39;</span><span class="p">])[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                 <span class="n">reference</span> <span class="o">=</span><span class="s2">&quot;mm10&quot;</span><span class="p">,</span><span class="n">save_homer</span> <span class="o">=</span> <span class="s2">&quot;Homer/GBM_female&quot;</span><span class="p">,</span>
                 <span class="n">homer_path</span> <span class="o">=</span> <span class="s2">&quot;/home/juanru/miniconda3/bin/&quot;</span><span class="p">,</span> <span class="n">num_cores</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">call_motif</span><span class="p">(</span><span class="n">peaks_frame</span><span class="o">=</span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups_df</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span> <span class="n">pval_cutoff</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">logfc_min</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">group</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Male_Brd4&#39;</span><span class="p">])[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                 <span class="n">reference</span> <span class="o">=</span><span class="s2">&quot;mm10&quot;</span><span class="p">,</span><span class="n">save_homer</span> <span class="o">=</span> <span class="s2">&quot;Homer/GBM_male&quot;</span><span class="p">,</span>
                 <span class="n">homer_path</span> <span class="o">=</span> <span class="s2">&quot;/home/juanru/miniconda3/bin/&quot;</span><span class="p">,</span> <span class="n">num_cores</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
There is no save_name, it will save to temp_Homer_trial.bed and then delete.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

        Position file = temp_Homer_trial.bed
        Genome = mm10
        Output Directory = Homer/GBM_female
        Fragment size set to 1000
        Using 12 CPUs
        Will not run homer for de novo motifs
        Found mset for &#34;mouse&#34;, will check against vertebrates motifs
        Peak/BED file conversion summary:
                BED/Header formatted lines: 1009
                peakfile formatted lines: 0

        Peak File Statistics:
                Total Peaks: 1009
                Redundant Peak IDs: 0
                Peaks lacking information: 0 (need at least 5 columns per peak)
                Peaks with misformatted coordinates: 0 (should be integer)
                Peaks with misformatted strand: 0 (should be either +/- or 0/1)

        Peak file looks good!

        Background files for 1000 bp fragments found.
        Custom genome sequence directory: /home/juanru/miniconda3/share/homer/.//data/genomes/mm10//

        Extracting sequences from directory: /home/juanru/miniconda3/share/homer/.//data/genomes/mm10//
        Extracting 96 sequences from chr1
        Extracting 93 sequences from chr2
        Extracting 62 sequences from chr3
        Extracting 39 sequences from chr4
        Extracting 67 sequences from chr5
        Extracting 66 sequences from chr6
        Extracting 62 sequences from chr7
        Extracting 45 sequences from chr8
        Extracting 59 sequences from chr9
        Extracting 50 sequences from chr10
        Extracting 41 sequences from chr11
        Extracting 41 sequences from chr12
        Extracting 46 sequences from chr13
        Extracting 25 sequences from chr14
        Extracting 26 sequences from chr15
        Extracting 40 sequences from chr16
        Extracting 40 sequences from chr17
        Extracting 32 sequences from chr18
        Extracting 32 sequences from chr19
        Extracting 47 sequences from chrX

        Not removing redundant sequences


        Sequences processed:
                Auto detected maximum sequence length of 1001 bp
                1009 total

        Frequency Bins: 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.6 0.7 0.8
        Freq    Bin     Count
        0.3     2       14
        0.35    3       101
        0.4     4       317
        0.45    5       309
        0.5     6       201
        0.6     7       66
        0.7     8       1

        Total sequences set to 50000

        Choosing background that matches in CpG/GC content...
        Bin     # Targets       # Background    Background Weight
        2       14      680     1.000
        3       101     4904    1.000
        4       317     15392   1.000
        5       309     15003   1.000
        6       201     9759    1.000
        7       66      3205    1.000
        8       1       49      0.991
        Assembling sequence file...
        Normalizing lower order oligos using homer2

        Reading input files...
        50001 total sequences read
        Autonormalization: 1-mers (4 total)
                A       29.26%  29.21%  1.001
                C       20.74%  20.79%  0.998
                G       20.74%  20.79%  0.998
                T       29.26%  29.21%  1.001
        Autonormalization: 2-mers (16 total)
                AA      9.53%   9.22%   1.034
                CA      7.34%   7.44%   0.987
                GA      6.15%   6.15%   1.001
                TA      6.23%   6.41%   0.972
                AC      5.27%   5.38%   0.980
                CC      5.08%   5.17%   0.983
                GC      4.24%   4.08%   1.037
                TC      6.15%   6.15%   1.001
                AG      7.36%   7.34%   1.003
                CG      0.96%   0.84%   1.148
                GG      5.08%   5.17%   0.983
                TG      7.34%   7.44%   0.987
                AT      7.09%   7.28%   0.974
                CT      7.36%   7.34%   1.003
                GT      5.27%   5.38%   0.980
                TT      9.53%   9.22%   1.034
        Autonormalization: 3-mers (64 total)
        Normalization weights can be found in file: Homer/GBM_female/seq.autonorm.tsv
        Converging on autonormalization solution:
        ...............................................................................
        Final normalization:    Autonormalization: 1-mers (4 total)
                A       29.26%  29.24%  1.001
                C       20.74%  20.76%  0.999
                G       20.74%  20.76%  0.999
                T       29.26%  29.24%  1.001
        Autonormalization: 2-mers (16 total)
                AA      9.53%   9.41%   1.013
                CA      7.34%   7.35%   0.998
                GA      6.15%   6.14%   1.001
                TA      6.23%   6.33%   0.985
                AC      5.27%   5.32%   0.991
                CC      5.08%   5.14%   0.989
                GC      4.24%   4.16%   1.018
                TC      6.15%   6.14%   1.001
                AG      7.36%   7.34%   1.002
                CG      0.96%   0.93%   1.033
                GG      5.08%   5.14%   0.989
                TG      7.34%   7.35%   0.998
                AT      7.09%   7.16%   0.990
                CT      7.36%   7.34%   1.002
                GT      5.27%   5.32%   0.991
                TT      9.53%   9.41%   1.013
        Autonormalization: 3-mers (64 total)
        Finished preparing sequence/group files

        ----------------------------------------------------------
        Known motif enrichment

        Reading input files...
        50001 total sequences read
        440 motifs loaded
        Cache length = 11180
        Using binomial scoring
        Checking enrichment of 440 motif(s)
        |0%                                    50%                                  100%|
        =================================================================================
        Preparing HTML output with sequence logos...
                1 of 440 (1e-18) Chop(bZIP)/MEF-Chop-ChIP-Seq(GSE35681)/Homer
                2 of 440 (1e-17) Atf4(bZIP)/MEF-Atf4-ChIP-Seq(GSE35681)/Homer
                3 of 440 (1e-17) ISRE(IRF)/ThioMac-LPS-Expression(GSE23622)/Homer
                4 of 440 (1e-15) Atf3(bZIP)/GBM-ATF3-ChIP-Seq(GSE33912)/Homer
                5 of 440 (1e-15) Fos(bZIP)/TSC-Fos-ChIP-Seq(GSE110950)/Homer
                6 of 440 (1e-15) AP-1(bZIP)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer
                7 of 440 (1e-15) JunB(bZIP)/DendriticCells-Junb-ChIP-Seq(GSE36099)/Homer
                8 of 440 (1e-14) Fra1(bZIP)/BT549-Fra1-ChIP-Seq(GSE46166)/Homer
                9 of 440 (1e-13) BATF(bZIP)/Th17-BATF-ChIP-Seq(GSE39756)/Homer
                10 of 440 (1e-13) IRF2(IRF)/Erythroblas-IRF2-ChIP-Seq(GSE36985)/Homer
                11 of 440 (1e-12) IRF3(IRF)/BMDM-Irf3-ChIP-Seq(GSE67343)/Homer
                12 of 440 (1e-12) IRF8(IRF)/BMDM-IRF8-ChIP-Seq(GSE77884)/Homer
                13 of 440 (1e-11) Fra2(bZIP)/Striatum-Fra2-ChIP-Seq(GSE43429)/Homer
                14 of 440 (1e-11) Fosl2(bZIP)/3T3L1-Fosl2-ChIP-Seq(GSE56872)/Homer
                15 of 440 (1e-10) Jun-AP1(bZIP)/K562-cJun-ChIP-Seq(GSE31477)/Homer
                16 of 440 (1e-9) IRF1(IRF)/PBMC-IRF1-ChIP-Seq(GSE43036)/Homer
                17 of 440 (1e-7) CEBP:AP1(bZIP)/ThioMac-CEBPb-ChIP-Seq(GSE21512)/Homer
                18 of 440 (1e-7) Rbpj1(?)/Panc1-Rbpj1-ChIP-Seq(GSE47459)/Homer
                19 of 440 (1e-7) TEAD1(TEAD)/HepG2-TEAD1-ChIP-Seq(Encode)/Homer
                20 of 440 (1e-6) PU.1:IRF8(ETS:IRF)/pDC-Irf8-ChIP-Seq(GSE66899)/Homer
                21 of 440 (1e-6) GABPA(ETS)/Jurkat-GABPa-ChIP-Seq(GSE17954)/Homer
                22 of 440 (1e-6) Zic3(Zf)/mES-Zic3-ChIP-Seq(GSE37889)/Homer
                23 of 440 (1e-5) Unknown-ESC-element(?)/mES-Nanog-ChIP-Seq(GSE11724)/Homer
                24 of 440 (1e-4) ETS1(ETS)/Jurkat-ETS1-ChIP-Seq(GSE17954)/Homer
                25 of 440 (1e-4) NF1-halfsite(CTF)/LNCaP-NF1-ChIP-Seq(Unpublished)/Homer
                26 of 440 (1e-4) Mef2c(MADS)/GM12878-Mef2c-ChIP-Seq(GSE32465)/Homer
                27 of 440 (1e-4) TEAD(TEA)/Fibroblast-PU.1-ChIP-Seq(Unpublished)/Homer
                28 of 440 (1e-4) Isl1(Homeobox)/Neuron-Isl1-ChIP-Seq(GSE31456)/Homer
                29 of 440 (1e-4) STAT5(Stat)/mCD4+-Stat5-ChIP-Seq(GSE12346)/Homer
                30 of 440 (1e-4) TEAD3(TEA)/HepG2-TEAD3-ChIP-Seq(Encode)/Homer
                31 of 440 (1e-4) CEBP:CEBP(bZIP)/MEF-Chop-ChIP-Seq(GSE35681)/Homer
                32 of 440 (1e-4) TEAD4(TEA)/Tropoblast-Tead4-ChIP-Seq(GSE37350)/Homer
                33 of 440 (1e-4) Pitx1:Ebox(Homeobox,bHLH)/Hindlimb-Pitx1-ChIP-Seq(GSE41591)/Homer
                34 of 440 (1e-4) PAX6(Paired,Homeobox)/Forebrain-Pax6-ChIP-Seq(GSE66961)/Homer
                35 of 440 (1e-4) Smad4(MAD)/ESC-SMAD4-ChIP-Seq(GSE29422)/Homer
                36 of 440 (1e-4) Dlx3(Homeobox)/Kerainocytes-Dlx3-ChIP-Seq(GSE89884)/Homer
                37 of 440 (1e-4) Fli1(ETS)/CD8-FLI-ChIP-Seq(GSE20898)/Homer
                38 of 440 (1e-4) CREB5(bZIP)/LNCaP-CREB5.V5-ChIP-Seq(GSE137775)/Homer
                39 of 440 (1e-3) EWS:FLI1-fusion(ETS)/SK_N_MC-EWS:FLI1-ChIP-Seq(SRA014231)/Homer
                40 of 440 (1e-3) Bach2(bZIP)/OCILy7-Bach2-ChIP-Seq(GSE44420)/Homer
                41 of 440 (1e-3) STAT4(Stat)/CD4-Stat4-ChIP-Seq(GSE22104)/Homer
                42 of 440 (1e-3) IRF4(IRF)/GM12878-IRF4-ChIP-Seq(GSE32465)/Homer
                43 of 440 (1e-3) Etv2(ETS)/ES-ER71-ChIP-Seq(GSE59402)/Homer
                44 of 440 (1e-3) HLF(bZIP)/HSC-HLF.Flag-ChIP-Seq(GSE69817)/Homer
                45 of 440 (1e-3) c-Jun-CRE(bZIP)/K562-cJun-ChIP-Seq(GSE31477)/Homer
                46 of 440 (1e-3) AMYB(HTH)/Testes-AMYB-ChIP-Seq(GSE44588)/Homer
                47 of 440 (1e-3) EWS:ERG-fusion(ETS)/CADO_ES1-EWS:ERG-ChIP-Seq(SRA014231)/Homer
                48 of 440 (1e-3) TEAD2(TEA)/Py2T-Tead2-ChIP-Seq(GSE55709)/Homer
                49 of 440 (1e-3) NFIL3(bZIP)/HepG2-NFIL3-ChIP-Seq(Encode)/Homer
                50 of 440 (1e-3) ETV1(ETS)/GIST48-ETV1-ChIP-Seq(GSE22441)/Homer
                51 of 440 (1e-3) Zic(Zf)/Cerebellum-ZIC1.2-ChIP-Seq(GSE60731)/Homer
                52 of 440 (1e-3) Mef2a(MADS)/HL1-Mef2a.biotin-ChIP-Seq(GSE21529)/Homer
                53 of 440 (1e-3) Elf4(ETS)/BMDM-Elf4-ChIP-Seq(GSE88699)/Homer
                54 of 440 (1e-3) CArG(MADS)/PUER-Srf-ChIP-Seq(Sullivan_et_al.)/Homer
                55 of 440 (1e-3) Elk1(ETS)/Hela-Elk1-ChIP-Seq(GSE31477)/Homer
                56 of 440 (1e-2) MafA(bZIP)/Islet-MafA-ChIP-Seq(GSE30298)/Homer
                57 of 440 (1e-2) MYB(HTH)/ERMYB-Myb-ChIPSeq(GSE22095)/Homer
                58 of 440 (1e-2) PU.1-IRF(ETS:IRF)/Bcell-PU.1-ChIP-Seq(GSE21512)/Homer
                59 of 440 (1e-2) EBF1(EBF)/Near-E2A-ChIP-Seq(GSE21512)/Homer
                60 of 440 (1e-2) DLX5(Homeobox)/BasalGanglia-Dlx5-ChIP-seq(GSE124936)/Homer
                61 of 440 (1e-2) Zic2(Zf)/ESC-Zic2-ChIP-Seq(SRP197560)/Homer
                62 of 440 (1e-2) DLX1(Homeobox)/BasalGanglia-Dlx1-ChIP-seq(GSE124936)/Homer
                63 of 440 (1e-2) Six2(Homeobox)/NephronProgenitor-Six2-ChIP-Seq(GSE39837)/Homer
                64 of 440 (1e-2) STAT6(Stat)/Macrophage-Stat6-ChIP-Seq(GSE38377)/Homer
                65 of 440 (1e-2) NFAT:AP1(RHD,bZIP)/Jurkat-NFATC1-ChIP-Seq(Jolma_et_al.)/Homer
                66 of 440 (1e-2) Fox:Ebox(Forkhead,bHLH)/Panc1-Foxa2-ChIP-Seq(GSE47459)/Homer
                67 of 440 (1e-2) Stat3(Stat)/mES-Stat3-ChIP-Seq(GSE11431)/Homer
                68 of 440 (1e-2) NFkB-p65-Rel(RHD)/ThioMac-LPS-Expression(GSE23622)/Homer
                69 of 440 (1e-2) EHF(ETS)/LoVo-EHF-ChIP-Seq(GSE49402)/Homer
                70 of 440 (1e-2) Hoxd11(Homeobox)/ChickenMSG-Hoxd11.Flag-ChIP-Seq(GSE86088)/Homer
                71 of 440 (1e-2) IRF:BATF(IRF:bZIP)/pDC-Irf8-ChIP-Seq(GSE66899)/Homer
                72 of 440 (1e-2) ETV4(ETS)/HepG2-ETV4-ChIP-Seq(ENCODE)/Homer
                73 of 440 (1e-2) Smad2(MAD)/ES-SMAD2-ChIP-Seq(GSE29422)/Homer
                74 of 440 (1e-2) STAT6(Stat)/CD4-Stat6-ChIP-Seq(GSE22104)/Homer
                75 of 440 (1e-2) Prop1(Homeobox)/GHFT1-PROP1.biotin-ChIP-Seq(GSE77302)/Homer
                76 of 440 (1e-2) Brn2(POU,Homeobox)/NPC-Brn2-ChIP-Seq(GSE35496)/Homer
                77 of 440 (1e-2) NFkB-p65(RHD)/GM12787-p65-ChIP-Seq(GSE19485)/Homer
                78 of 440 (1e-2) ETS:E-box(ETS,bHLH)/HPC7-Scl-ChIP-Seq(GSE22178)/Homer
                79 of 440 (1e-2) PU.1(ETS)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer
                80 of 440 (1e-2) Nkx6.1(Homeobox)/Islet-Nkx6.1-ChIP-Seq(GSE40975)/Homer
                81 of 440 (1e-2) Mef2d(MADS)/Retina-Mef2d-ChIP-Seq(GSE61391)/Homer
        Skipping...
        Job finished - if results look good, please send beer to ..

        Cleaning up tmp files...


        Position file = temp_Homer_trial.bed
        Genome = mm10
        Output Directory = Homer/GBM_male
        Fragment size set to 1000
        Using 12 CPUs
        Will not run homer for de novo motifs
        Found mset for &#34;mouse&#34;, will check against vertebrates motifs
        Peak/BED file conversion summary:
                BED/Header formatted lines: 585
                peakfile formatted lines: 0

        Peak File Statistics:
                Total Peaks: 585
                Redundant Peak IDs: 0
                Peaks lacking information: 0 (need at least 5 columns per peak)
                Peaks with misformatted coordinates: 0 (should be integer)
                Peaks with misformatted strand: 0 (should be either +/- or 0/1)

        Peak file looks good!

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finished!
There is no save_name, it will save to temp_Homer_trial.bed and then delete.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
        Background files for 1000 bp fragments found.
        Custom genome sequence directory: /home/juanru/miniconda3/share/homer/.//data/genomes/mm10//

        Extracting sequences from directory: /home/juanru/miniconda3/share/homer/.//data/genomes/mm10//
        Extracting 36 sequences from chr1
        Extracting 40 sequences from chr2
        Extracting 33 sequences from chr3
        Extracting 20 sequences from chr4
        Extracting 28 sequences from chr5
        Extracting 35 sequences from chr6
        Extracting 18 sequences from chr7
        Extracting 25 sequences from chr8
        Extracting 21 sequences from chr9
        Extracting 52 sequences from chr10
        Extracting 39 sequences from chr11
        Extracting 26 sequences from chr12
        Extracting 23 sequences from chr13
        Extracting 36 sequences from chr14
        Extracting 43 sequences from chr15
        Extracting 18 sequences from chr16
        Extracting 17 sequences from chr17
        Extracting 28 sequences from chr18
        Extracting 22 sequences from chr19
        Extracting 21 sequences from chrX
        Extracting 4 sequences from chrY

        Not removing redundant sequences


        Sequences processed:
                Auto detected maximum sequence length of 1001 bp
                585 total

        Frequency Bins: 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.6 0.7 0.8
        Freq    Bin     Count
        0.3     2       2
        0.35    3       41
        0.4     4       134
        0.45    5       215
        0.5     6       133
        0.6     7       56
        0.7     8       4

        Total sequences set to 50000

        Choosing background that matches in CpG/GC content...
        Bin     # Targets       # Background    Background Weight
        2       2       169     1.000
        3       41      3463    1.000
        4       134     11319   1.000
        5       215     18161   1.000
        6       133     11235   1.000
        7       56      4730    1.000
        8       4       338     1.000
        Assembling sequence file...
        Normalizing lower order oligos using homer2

        Reading input files...
        50000 total sequences read
        Autonormalization: 1-mers (4 total)
                A       28.57%  28.51%  1.002
                C       21.43%  21.49%  0.997
                G       21.43%  21.49%  0.997
                T       28.57%  28.51%  1.002
        Autonormalization: 2-mers (16 total)
                AA      9.11%   8.75%   1.041
                CA      7.48%   7.54%   0.993
                GA      6.12%   6.21%   0.985
                TA      5.87%   6.02%   0.974
                AC      5.39%   5.41%   0.995
                CC      5.41%   5.51%   0.982
                GC      4.51%   4.35%   1.035
                TC      6.12%   6.21%   0.985
                AG      7.40%   7.49%   0.988
                CG      1.13%   0.95%   1.196
                GG      5.41%   5.51%   0.982
                TG      7.48%   7.54%   0.993
                AT      6.67%   6.86%   0.973
                CT      7.40%   7.49%   0.988
                GT      5.39%   5.41%   0.995
                TT      9.11%   8.75%   1.041
        Autonormalization: 3-mers (64 total)
        Normalization weights can be found in file: Homer/GBM_male/seq.autonorm.tsv
        Converging on autonormalization solution:
        ...............................................................................
        Final normalization:    Autonormalization: 1-mers (4 total)
                A       28.57%  28.55%  1.001
                C       21.43%  21.45%  0.999
                G       21.43%  21.45%  0.999
                T       28.57%  28.55%  1.001
        Autonormalization: 2-mers (16 total)
                AA      9.11%   8.97%   1.016
                CA      7.48%   7.47%   1.001
                GA      6.12%   6.15%   0.995
                TA      5.87%   5.97%   0.983
                AC      5.39%   5.40%   0.998
                CC      5.41%   5.46%   0.992
                GC      4.51%   4.44%   1.014
                TC      6.12%   6.15%   0.995
                AG      7.40%   7.42%   0.997
                CG      1.13%   1.10%   1.033
                GG      5.41%   5.46%   0.992
                TG      7.48%   7.47%   1.001
                AT      6.67%   6.77%   0.986
                CT      7.40%   7.42%   0.997
                GT      5.39%   5.40%   0.998
                TT      9.11%   8.97%   1.016
        Autonormalization: 3-mers (64 total)
        Finished preparing sequence/group files

        ----------------------------------------------------------
        Known motif enrichment

        Reading input files...
        50000 total sequences read
        440 motifs loaded
        Cache length = 11180
        Using binomial scoring
        Checking enrichment of 440 motif(s)
        |0%                                    50%                                  100%|
        =================================================================================
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finished!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

        Preparing HTML output with sequence logos...
                1 of 440 (1e-42) Fosl2(bZIP)/3T3L1-Fosl2-ChIP-Seq(GSE56872)/Homer
                2 of 440 (1e-41) Jun-AP1(bZIP)/K562-cJun-ChIP-Seq(GSE31477)/Homer
                3 of 440 (1e-39) Fos(bZIP)/TSC-Fos-ChIP-Seq(GSE110950)/Homer
                4 of 440 (1e-38) JunB(bZIP)/DendriticCells-Junb-ChIP-Seq(GSE36099)/Homer
                5 of 440 (1e-37) Fra1(bZIP)/BT549-Fra1-ChIP-Seq(GSE46166)/Homer
                6 of 440 (1e-35) Fra2(bZIP)/Striatum-Fra2-ChIP-Seq(GSE43429)/Homer
                7 of 440 (1e-35) BATF(bZIP)/Th17-BATF-ChIP-Seq(GSE39756)/Homer
                8 of 440 (1e-34) Atf3(bZIP)/GBM-ATF3-ChIP-Seq(GSE33912)/Homer
                9 of 440 (1e-34) AP-1(bZIP)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer
                10 of 440 (1e-22) Bach2(bZIP)/OCILy7-Bach2-ChIP-Seq(GSE44420)/Homer
                11 of 440 (1e-10) Nrf2(bZIP)/Lymphoblast-Nrf2-ChIP-Seq(GSE37589)/Homer
                12 of 440 (1e-9) NF-E2(bZIP)/K562-NFE2-ChIP-Seq(GSE31477)/Homer
                13 of 440 (1e-7) Bach1(bZIP)/K562-Bach1-ChIP-Seq(GSE31477)/Homer
                14 of 440 (1e-7) NFE2L2(bZIP)/HepG2-NFE2L2-ChIP-Seq(Encode)/Homer
                15 of 440 (1e-6) MafK(bZIP)/C2C12-MafK-ChIP-Seq(GSE36030)/Homer
                16 of 440 (1e-5) Ets1-distal(ETS)/CD4+-PolII-ChIP-Seq(Barski_et_al.)/Homer
                17 of 440 (1e-4) Isl1(Homeobox)/Neuron-Isl1-ChIP-Seq(GSE31456)/Homer
                18 of 440 (1e-4) Sox3(HMG)/NPC-Sox3-ChIP-Seq(GSE33059)/Homer
                19 of 440 (1e-3) Bapx1(Homeobox)/VertebralCol-Bapx1-ChIP-Seq(GSE36672)/Homer
                20 of 440 (1e-3) Fli1(ETS)/CD8-FLI-ChIP-Seq(GSE20898)/Homer
                21 of 440 (1e-3) MafA(bZIP)/Islet-MafA-ChIP-Seq(GSE30298)/Homer
                22 of 440 (1e-3) Fox:Ebox(Forkhead,bHLH)/Panc1-Foxa2-ChIP-Seq(GSE47459)/Homer
                23 of 440 (1e-3) Nkx2.2(Homeobox)/NPC-Nkx2.2-ChIP-Seq(GSE61673)/Homer
                24 of 440 (1e-3) Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132)/Homer
                25 of 440 (1e-3) LHX9(Homeobox)/Hct116-LHX9.V5-ChIP-Seq(GSE116822)/Homer
                26 of 440 (1e-3) TATA-Box(TBP)/Promoter/Homer
                27 of 440 (1e-3) KLF14(Zf)/HEK293-KLF14.GFP-ChIP-Seq(GSE58341)/Homer
                28 of 440 (1e-2) ZNF669(Zf)/HEK293-ZNF669.GFP-ChIP-Seq(GSE58341)/Homer
                29 of 440 (1e-2) Lhx2(Homeobox)/HFSC-Lhx2-ChIP-Seq(GSE48068)/Homer
                30 of 440 (1e-2) Sox4(HMG)/proB-Sox4-ChIP-Seq(GSE50066)/Homer
                31 of 440 (1e-2) DLX5(Homeobox)/BasalGanglia-Dlx5-ChIP-seq(GSE124936)/Homer
                32 of 440 (1e-2) RUNX-AML(Runt)/CD4+-PolII-ChIP-Seq(Barski_et_al.)/Homer
                33 of 440 (1e-2) Sp5(Zf)/mES-Sp5.Flag-ChIP-Seq(GSE72989)/Homer
                34 of 440 (1e-2) EWS:ERG-fusion(ETS)/CADO_ES1-EWS:ERG-ChIP-Seq(SRA014231)/Homer
                35 of 440 (1e-2) RUNX1(Runt)/Jurkat-RUNX1-ChIP-Seq(GSE29180)/Homer
                36 of 440 (1e-2) Etv2(ETS)/ES-ER71-ChIP-Seq(GSE59402)/Homer
                37 of 440 (1e-2) ETV4(ETS)/HepG2-ETV4-ChIP-Seq(ENCODE)/Homer
                38 of 440 (1e-2) Sox2(HMG)/mES-Sox2-ChIP-Seq(GSE11431)/Homer
                39 of 440 (1e-2) ETS1(ETS)/Jurkat-ETS1-ChIP-Seq(GSE17954)/Homer
                40 of 440 (1e-2) Brn1(POU,Homeobox)/NPC-Brn1-ChIP-Seq(GSE35496)/Homer
                41 of 440 (1e-2) CHR(?)/Hela-CellCycle-Expression/Homer
                42 of 440 (1e-2) Dlx3(Homeobox)/Kerainocytes-Dlx3-ChIP-Seq(GSE89884)/Homer
                43 of 440 (1e-2) EWS:FLI1-fusion(ETS)/SK_N_MC-EWS:FLI1-ChIP-Seq(SRA014231)/Homer
                44 of 440 (1e-2) KLF3(Zf)/MEF-Klf3-ChIP-Seq(GSE44748)/Homer
                45 of 440 (1e-2) STAT5(Stat)/mCD4+-Stat5-ChIP-Seq(GSE12346)/Homer
                46 of 440 (1e-2) Elk4(ETS)/Hela-Elk4-ChIP-Seq(GSE31477)/Homer
                47 of 440 (1e-2) Tbx20(T-box)/Heart-Tbx20-ChIP-Seq(GSE29636)/Homer
                48 of 440 (1e-2) Rfx6(HTH)/Min6b1-Rfx6.HA-ChIP-Seq(GSE62844)/Homer
                49 of 440 (1e-2) ETS:RUNX(ETS,Runt)/Jurkat-RUNX1-ChIP-Seq(GSE17954)/Homer
                50 of 440 (1e-2) AMYB(HTH)/Testes-AMYB-ChIP-Seq(GSE44588)/Homer
                51 of 440 (1e-2) Lhx1(Homeobox)/EmbryoCarcinoma-Lhx1-ChIP-Seq(GSE70957)/Homer
                52 of 440 (1e-2) Lhx3(Homeobox)/Neuron-Lhx3-ChIP-Seq(GSE31456)/Homer
                53 of 440 (1e-2) DLX2(Homeobox)/BasalGanglia-Dlx2-ChIP-seq(GSE124936)/Homer
                54 of 440 (1e-2) Nkx6.1(Homeobox)/Islet-Nkx6.1-ChIP-Seq(GSE40975)/Homer
                55 of 440 (1e-2) OCT4-SOX2-TCF-NANOG(POU,Homeobox,HMG)/mES-Oct4-ChIP-Seq(GSE11431)/Homer
                56 of 440 (1e-2) Sox7(HMG)/ESC-Sox7-ChIP-Seq(GSE133899)/Homer
                57 of 440 (1e-2) Sox6(HMG)/Myotubes-Sox6-ChIP-Seq(GSE32627)/Homer
                58 of 440 (1e-2) DLX1(Homeobox)/BasalGanglia-Dlx1-ChIP-seq(GSE124936)/Homer
                59 of 440 (1e-2) AP-2alpha(AP2)/Hela-AP2alpha-ChIP-Seq(GSE31477)/Homer
                60 of 440 (1e-2) GABPA(ETS)/Jurkat-GABPa-ChIP-Seq(GSE17954)/Homer
                61 of 440 (1e-2) Maz(Zf)/HepG2-Maz-ChIP-Seq(GSE31477)/Homer
        Skipping...
        Job finished - if results look good, please send beer to ..

        Cleaning up tmp files...

</pre></div></div>
</div>
<p>Find motifs under male peaks but not in female peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">compare_motif</span><span class="p">(</span><span class="s2">&quot;Homer/GBM_male&quot;</span><span class="p">,</span> <span class="s2">&quot;Homer/GBM_female&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Motif Name</th>
      <th>Consensus</th>
      <th>P-value</th>
      <th>Log P-value</th>
      <th>q-value (Benjamini)</th>
      <th># of Target Sequences with Motif(of 585)</th>
      <th>% of Target Sequences with Motif</th>
      <th># of Background Sequences with Motif(of 2133)</th>
      <th>% of Background Sequences with Motif</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>Nrf2(bZIP)/Lymphoblast-Nrf2-ChIP-Seq(GSE37589)...</td>
      <td>HTGCTGAGTCAT</td>
      <td>1.000000e-13</td>
      <td>-30.290</td>
      <td>0.0000</td>
      <td>39.0</td>
      <td>6.67%</td>
      <td>33.5</td>
      <td>1.57%</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NF-E2(bZIP)/K562-NFE2-ChIP-Seq(GSE31477)/Homer</td>
      <td>GATGACTCAGCA</td>
      <td>1.000000e-11</td>
      <td>-26.520</td>
      <td>0.0000</td>
      <td>41.0</td>
      <td>7.01%</td>
      <td>41.6</td>
      <td>1.95%</td>
    </tr>
    <tr>
      <th>12</th>
      <td>NFE2L2(bZIP)/HepG2-NFE2L2-ChIP-Seq(Encode)/Homer</td>
      <td>AWWWTGCTGAGTCAT</td>
      <td>1.000000e-11</td>
      <td>-25.480</td>
      <td>0.0000</td>
      <td>48.0</td>
      <td>8.21%</td>
      <td>56.4</td>
      <td>2.65%</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Bach1(bZIP)/K562-Bach1-ChIP-Seq(GSE31477)/Homer</td>
      <td>AWWNTGCTGAGTCAT</td>
      <td>1.000000e-09</td>
      <td>-21.840</td>
      <td>0.0000</td>
      <td>39.0</td>
      <td>6.67%</td>
      <td>44.3</td>
      <td>2.07%</td>
    </tr>
    <tr>
      <th>14</th>
      <td>MafK(bZIP)/C2C12-MafK-ChIP-Seq(GSE36030)/Homer</td>
      <td>GCTGASTCAGCA</td>
      <td>1.000000e-09</td>
      <td>-20.870</td>
      <td>0.0000</td>
      <td>97.0</td>
      <td>16.58%</td>
      <td>186.9</td>
      <td>8.76%</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Ets1-distal(ETS)/CD4+-PolII-ChIP-Seq(Barski_et...</td>
      <td>MACAGGAAGT</td>
      <td>1.000000e-05</td>
      <td>-13.680</td>
      <td>0.0000</td>
      <td>116.0</td>
      <td>19.83%</td>
      <td>273.6</td>
      <td>12.82%</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Sox3(HMG)/NPC-Sox3-ChIP-Seq(GSE33059)/Homer</td>
      <td>CCWTTGTY</td>
      <td>1.000000e-04</td>
      <td>-9.625</td>
      <td>0.0016</td>
      <td>434.0</td>
      <td>74.19%</td>
      <td>1425.7</td>
      <td>66.82%</td>
    </tr>
    <tr>
      <th>18</th>
      <td>Nkx2.2(Homeobox)/NPC-Nkx2.2-ChIP-Seq(GSE61673)...</td>
      <td>BTBRAGTGSN</td>
      <td>1.000000e-04</td>
      <td>-9.612</td>
      <td>0.0016</td>
      <td>419.0</td>
      <td>71.62%</td>
      <td>1367.5</td>
      <td>64.09%</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Bapx1(Homeobox)/VertebralCol-Bapx1-ChIP-Seq(GS...</td>
      <td>TTRAGTGSYK</td>
      <td>1.000000e-04</td>
      <td>-9.553</td>
      <td>0.0016</td>
      <td>467.0</td>
      <td>79.83%</td>
      <td>1556.7</td>
      <td>72.96%</td>
    </tr>
    <tr>
      <th>20</th>
      <td>RUNX1(Runt)/Jurkat-RUNX1-ChIP-Seq(GSE29180)/Homer</td>
      <td>AAACCACARM</td>
      <td>1.000000e-03</td>
      <td>-8.988</td>
      <td>0.0026</td>
      <td>293.0</td>
      <td>50.09%</td>
      <td>906.6</td>
      <td>42.49%</td>
    </tr>
    <tr>
      <th>23</th>
      <td>ZNF669(Zf)/HEK293-ZNF669.GFP-ChIP-Seq(GSE58341...</td>
      <td>GARTGGTCATCGCCC</td>
      <td>1.000000e-03</td>
      <td>-8.215</td>
      <td>0.0050</td>
      <td>24.0</td>
      <td>4.10%</td>
      <td>39.7</td>
      <td>1.86%</td>
    </tr>
    <tr>
      <th>27</th>
      <td>RUNX-AML(Runt)/CD4+-PolII-ChIP-Seq(Barski_et_a...</td>
      <td>GCTGTGGTTW</td>
      <td>1.000000e-03</td>
      <td>-7.530</td>
      <td>0.0086</td>
      <td>212.0</td>
      <td>36.24%</td>
      <td>637.6</td>
      <td>29.89%</td>
    </tr>
    <tr>
      <th>29</th>
      <td>Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132...</td>
      <td>CCWTTGTYYB</td>
      <td>1.000000e-03</td>
      <td>-7.342</td>
      <td>0.0095</td>
      <td>407.0</td>
      <td>69.57%</td>
      <td>1347.5</td>
      <td>63.15%</td>
    </tr>
    <tr>
      <th>30</th>
      <td>Sp1(Zf)/Promoter/Homer</td>
      <td>GGCCCCGCCCCC</td>
      <td>1.000000e-03</td>
      <td>-7.220</td>
      <td>0.0104</td>
      <td>31.0</td>
      <td>5.30%</td>
      <td>60.7</td>
      <td>2.85%</td>
    </tr>
    <tr>
      <th>32</th>
      <td>E2F3(E2F)/MEF-E2F3-ChIP-Seq(GSE71376)/Homer</td>
      <td>BTKGGCGGGAAA</td>
      <td>1.000000e-03</td>
      <td>-7.074</td>
      <td>0.0113</td>
      <td>77.0</td>
      <td>13.16%</td>
      <td>195.8</td>
      <td>9.18%</td>
    </tr>
    <tr>
      <th>37</th>
      <td>Sox7(HMG)/ESC-Sox7-ChIP-Seq(GSE133899)/Homer</td>
      <td>VVRRAACAATGG</td>
      <td>1.000000e-02</td>
      <td>-6.569</td>
      <td>0.0162</td>
      <td>102.0</td>
      <td>17.44%</td>
      <td>278.9</td>
      <td>13.07%</td>
    </tr>
    <tr>
      <th>40</th>
      <td>Brn1(POU,Homeobox)/NPC-Brn1-ChIP-Seq(GSE35496)...</td>
      <td>TATGCWAATBAV</td>
      <td>1.000000e-02</td>
      <td>-6.345</td>
      <td>0.0188</td>
      <td>134.0</td>
      <td>22.91%</td>
      <td>385.4</td>
      <td>18.06%</td>
    </tr>
    <tr>
      <th>42</th>
      <td>Sox4(HMG)/proB-Sox4-ChIP-Seq(GSE50066)/Homer</td>
      <td>YCTTTGTTCC</td>
      <td>1.000000e-02</td>
      <td>-6.213</td>
      <td>0.0205</td>
      <td>255.0</td>
      <td>43.59%</td>
      <td>804.0</td>
      <td>37.68%</td>
    </tr>
    <tr>
      <th>43</th>
      <td>ZNF317(Zf)/HEK293-ZNF317.GFP-ChIP-Seq(GSE58341...</td>
      <td>GTCWGCTGTYYCTCT</td>
      <td>1.000000e-02</td>
      <td>-6.112</td>
      <td>0.0222</td>
      <td>29.0</td>
      <td>4.96%</td>
      <td>59.0</td>
      <td>2.77%</td>
    </tr>
    <tr>
      <th>45</th>
      <td>OCT4-SOX2-TCF-NANOG(POU,Homeobox,HMG)/mES-Oct4...</td>
      <td>ATTTGCATAACAATG</td>
      <td>1.000000e-02</td>
      <td>-5.636</td>
      <td>0.0341</td>
      <td>96.0</td>
      <td>16.41%</td>
      <td>267.1</td>
      <td>12.52%</td>
    </tr>
    <tr>
      <th>46</th>
      <td>Isl1(Homeobox)/Neuron-Isl1-ChIP-Seq(GSE31456)/...</td>
      <td>CTAATKGV</td>
      <td>1.000000e-02</td>
      <td>-5.578</td>
      <td>0.0354</td>
      <td>467.0</td>
      <td>79.83%</td>
      <td>1601.5</td>
      <td>75.06%</td>
    </tr>
    <tr>
      <th>48</th>
      <td>Sox2(HMG)/mES-Sox2-ChIP-Seq(GSE11431)/Homer</td>
      <td>BCCATTGTTC</td>
      <td>1.000000e-02</td>
      <td>-5.408</td>
      <td>0.0402</td>
      <td>272.0</td>
      <td>46.50%</td>
      <td>876.6</td>
      <td>41.09%</td>
    </tr>
    <tr>
      <th>49</th>
      <td>Egr1(Zf)/K562-Egr1-ChIP-Seq(GSE32465)/Homer</td>
      <td>TGCGTGGGYG</td>
      <td>1.000000e-02</td>
      <td>-5.339</td>
      <td>0.0422</td>
      <td>137.0</td>
      <td>23.42%</td>
      <td>406.2</td>
      <td>19.04%</td>
    </tr>
    <tr>
      <th>50</th>
      <td>Tbx20(T-box)/Heart-Tbx20-ChIP-Seq(GSE29636)/Homer</td>
      <td>GGTGYTGACAGS</td>
      <td>1.000000e-02</td>
      <td>-5.310</td>
      <td>0.0426</td>
      <td>67.0</td>
      <td>11.45%</td>
      <td>178.0</td>
      <td>8.34%</td>
    </tr>
    <tr>
      <th>51</th>
      <td>Nkx3.1(Homeobox)/LNCaP-Nkx3.1-ChIP-Seq(GSE2826...</td>
      <td>AAGCACTTAA</td>
      <td>1.000000e-02</td>
      <td>-5.277</td>
      <td>0.0432</td>
      <td>479.0</td>
      <td>81.88%</td>
      <td>1652.0</td>
      <td>77.43%</td>
    </tr>
    <tr>
      <th>52</th>
      <td>E2F6(E2F)/Hela-E2F6-ChIP-Seq(GSE31477)/Homer</td>
      <td>GGCGGGAARN</td>
      <td>1.000000e-02</td>
      <td>-5.256</td>
      <td>0.0433</td>
      <td>64.0</td>
      <td>10.94%</td>
      <td>168.7</td>
      <td>7.91%</td>
    </tr>
    <tr>
      <th>53</th>
      <td>TATA-Box(TBP)/Promoter/Homer</td>
      <td>CCTTTTAWAGSC</td>
      <td>1.000000e-02</td>
      <td>-5.201</td>
      <td>0.0449</td>
      <td>382.0</td>
      <td>65.30%</td>
      <td>1282.4</td>
      <td>60.10%</td>
    </tr>
    <tr>
      <th>55</th>
      <td>Klf9(Zf)/GBM-Klf9-ChIP-Seq(GSE62211)/Homer</td>
      <td>GCCACRCCCACY</td>
      <td>1.000000e-02</td>
      <td>-5.137</td>
      <td>0.0462</td>
      <td>85.0</td>
      <td>14.53%</td>
      <td>236.3</td>
      <td>11.08%</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Find motifs in female but not in male.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">compare_motif</span><span class="p">(</span><span class="s2">&quot;Homer/GBM_female&quot;</span><span class="p">,</span> <span class="s2">&quot;Homer/GBM_male&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Motif Name</th>
      <th>Consensus</th>
      <th>P-value</th>
      <th>Log P-value</th>
      <th>q-value (Benjamini)</th>
      <th># of Target Sequences with Motif(of 1009)</th>
      <th>% of Target Sequences with Motif</th>
      <th># of Background Sequences with Motif(of 2327)</th>
      <th>% of Background Sequences with Motif</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Atf4(bZIP)/MEF-Atf4-ChIP-Seq(GSE35681)/Homer</td>
      <td>MTGATGCAAT</td>
      <td>1.000000e-14</td>
      <td>-33.810</td>
      <td>0.0000</td>
      <td>228.0</td>
      <td>22.60%</td>
      <td>313.9</td>
      <td>13.49%</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Chop(bZIP)/MEF-Chop-ChIP-Seq(GSE35681)/Homer</td>
      <td>ATTGCATCAT</td>
      <td>1.000000e-14</td>
      <td>-32.890</td>
      <td>0.0000</td>
      <td>197.0</td>
      <td>19.52%</td>
      <td>259.5</td>
      <td>11.15%</td>
    </tr>
    <tr>
      <th>2</th>
      <td>IRF8(IRF)/BMDM-IRF8-ChIP-Seq(GSE77884)/Homer</td>
      <td>GRAASTGAAAST</td>
      <td>1.000000e-12</td>
      <td>-28.550</td>
      <td>0.0000</td>
      <td>255.0</td>
      <td>25.27%</td>
      <td>381.5</td>
      <td>16.40%</td>
    </tr>
    <tr>
      <th>5</th>
      <td>EBF1(EBF)/Near-E2A-ChIP-Seq(GSE21512)/Homer</td>
      <td>GTCCCCWGGGGA</td>
      <td>1.000000e-11</td>
      <td>-27.250</td>
      <td>0.0000</td>
      <td>310.0</td>
      <td>30.72%</td>
      <td>495.0</td>
      <td>21.27%</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Zic3(Zf)/mES-Zic3-ChIP-Seq(GSE37889)/Homer</td>
      <td>GGCCYCCTGCTGDGH</td>
      <td>1.000000e-11</td>
      <td>-25.440</td>
      <td>0.0000</td>
      <td>189.0</td>
      <td>18.73%</td>
      <td>266.5</td>
      <td>11.45%</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>122</th>
      <td>SPDEF(ETS)/VCaP-SPDEF-ChIP-Seq(SRA014231)/Homer</td>
      <td>ASWTCCTGBT</td>
      <td>1.000000e-02</td>
      <td>-4.786</td>
      <td>0.0299</td>
      <td>420.0</td>
      <td>41.63%</td>
      <td>882.4</td>
      <td>37.92%</td>
    </tr>
    <tr>
      <th>123</th>
      <td>Srebp2(bHLH)/HepG2-Srebp2-ChIP-Seq(GSE31477)/H...</td>
      <td>CGGTCACSCCAC</td>
      <td>1.000000e-02</td>
      <td>-4.760</td>
      <td>0.0304</td>
      <td>45.0</td>
      <td>4.46%</td>
      <td>71.8</td>
      <td>3.08%</td>
    </tr>
    <tr>
      <th>124</th>
      <td>ZNF519(Zf)/HEK293-ZNF519.GFP-ChIP-Seq(GSE58341...</td>
      <td>GAGSCCGAGC</td>
      <td>1.000000e-02</td>
      <td>-4.720</td>
      <td>0.0314</td>
      <td>37.0</td>
      <td>3.67%</td>
      <td>56.1</td>
      <td>2.41%</td>
    </tr>
    <tr>
      <th>125</th>
      <td>AP-2alpha(AP2)/Hela-AP2alpha-ChIP-Seq(GSE31477...</td>
      <td>ATGCCCTGAGGC</td>
      <td>1.000000e-02</td>
      <td>-4.695</td>
      <td>0.0319</td>
      <td>167.0</td>
      <td>16.55%</td>
      <td>323.9</td>
      <td>13.92%</td>
    </tr>
    <tr>
      <th>126</th>
      <td>Atf7(bZIP)/3T3L1-Atf7-ChIP-Seq(GSE56872)/Homer</td>
      <td>NGRTGACGTCAY</td>
      <td>1.000000e-02</td>
      <td>-4.676</td>
      <td>0.0323</td>
      <td>192.0</td>
      <td>19.03%</td>
      <td>377.5</td>
      <td>16.22%</td>
    </tr>
  </tbody>
</table>
<p>98 rows × 9 columns</p>
</div></div>
</div>
<p>Next, we want to identify differentially bound peaks whose nearby genes are differentially expressed between males and females. To do so, we must first read in the bulk RNA-seq data. Note that this RNA-seq data is normalized by <a class="reference external" href="https://wiki.arrayserver.com/wiki/index.php?title=RPKM">RPKM</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna</span> <span class="o">=</span>  <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">mouse_brd4_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span><span class="p">)</span>
<span class="n">rna</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample.f6_dmso_3</th>
      <th>sample.f6_dmso_2</th>
      <th>sample.f6_dmso_1</th>
      <th>sample.m6_dmso_3</th>
      <th>sample.m6_dmso_2</th>
      <th>sample.m6_dmso_1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>mt-Tf</th>
      <td>0.000000</td>
      <td>0.639272</td>
      <td>0.000000</td>
      <td>1.156308</td>
      <td>0.000000</td>
      <td>0.687531</td>
    </tr>
    <tr>
      <th>mt-Rnr1</th>
      <td>127.907281</td>
      <td>184.396768</td>
      <td>172.544723</td>
      <td>130.417054</td>
      <td>192.566581</td>
      <td>114.212201</td>
    </tr>
    <tr>
      <th>mt-Tv</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.761150</td>
      <td>0.000000</td>
      <td>0.510777</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>mt-Rnr2</th>
      <td>247.267306</td>
      <td>527.855824</td>
      <td>324.345998</td>
      <td>675.031187</td>
      <td>684.376279</td>
      <td>418.345603</td>
    </tr>
    <tr>
      <th>mt-Tl1</th>
      <td>44.418710</td>
      <td>117.660098</td>
      <td>70.481211</td>
      <td>167.217607</td>
      <td>115.129081</td>
      <td>127.789076</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Gm16897</th>
      <td>0.281553</td>
      <td>0.349020</td>
      <td>0.463442</td>
      <td>0.078913</td>
      <td>0.070742</td>
      <td>0.112610</td>
    </tr>
    <tr>
      <th>A330023F24Rik</th>
      <td>0.118678</td>
      <td>0.055749</td>
      <td>0.058441</td>
      <td>0.029411</td>
      <td>0.018833</td>
      <td>0.049965</td>
    </tr>
    <tr>
      <th>Cd46</th>
      <td>0.127930</td>
      <td>0.095151</td>
      <td>0.075997</td>
      <td>0.061467</td>
      <td>0.055103</td>
      <td>0.087715</td>
    </tr>
    <tr>
      <th>Cr1l</th>
      <td>9.993119</td>
      <td>14.027801</td>
      <td>13.717564</td>
      <td>13.586101</td>
      <td>13.240219</td>
      <td>11.089902</td>
    </tr>
    <tr>
      <th>Cr2</th>
      <td>0.018832</td>
      <td>0.056028</td>
      <td>0.019578</td>
      <td>0.012668</td>
      <td>0.017034</td>
      <td>0.015064</td>
    </tr>
  </tbody>
</table>
<p>21430 rows × 6 columns</p>
</div></div>
</div>
<p>We now plot the heatmap for calling cards data and gene expression data.</p>
<p>The top lines are the relative Brd4 binding for each of the two groups. The bottom lines are the gene expressions of the top two nearest genes in the 3 RNA-seq replicates for males and females; each gene shown separately.</p>
<p>We can see a relationship between Brd4 binding and the expression of nearby genes. Genes that are strongly bound in females tend to be expressed in females and expressed at lower levels in males and vice versa. The effect is much more muted for the 2nd nearest gene to the peak.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span><span class="n">rna</span><span class="p">,</span> <span class="n">rnalabels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Female1&quot;</span><span class="p">,</span> <span class="s2">&quot;Female2&quot;</span><span class="p">,</span> <span class="s2">&quot;Female3&quot;</span><span class="p">,</span><span class="s2">&quot;Male1&quot;</span><span class="p">,</span> <span class="s2">&quot;Male2&quot;</span><span class="p">,</span> <span class="s2">&quot;Male3&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Please make sure that the samples in adata_cc and rna are in the same order.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/Desktop/pycallingcards/pycallingcards/pycallingcards/plotting/_heatmap_ccrna.py:193: RuntimeWarning: invalid value encountered in true_divide
  data[groupnumber:secondnum, :]
/home/juanru/Desktop/pycallingcards/pycallingcards/pycallingcards/plotting/_heatmap_ccrna.py:195: RuntimeWarning: invalid value encountered in true_divide
  data[secondnum:, :]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_45_2.png" src="../../_images/tutorials_notebooks_Brd4_bulk_45_2.png" />
</div>
</div>
<p>Find out the co-differential peak-gene pairs. Look into all the differential peaks and then see if the annotated genes are also significantly expressed. We can set the pvalue and score/log foldchange cutoff easily.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pair_peak_gene_bulk</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="s2">&quot;https://github.com/The-Mitra-Lab/pycallingcards_data/releases/download/data/deseq_MF.csv&quot;</span><span class="p">,</span>
                                   <span class="n">name_cc</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">,</span> <span class="n">pvalue_cutoff_cc</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span> <span class="n">pvalue_cutoff_rna</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
                                   <span class="n">lfc_cutoff_rna</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">lfc_cutoff_cc</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>
<span class="n">result</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Peak</th>
      <th>logfoldchanges_peak</th>
      <th>Pvalue_peak</th>
      <th>Pvalue_adj_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
      <th>Pvalue_adj_gene</th>
      <th>Distance_peak_to_gene</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1_174659622_174662521</td>
      <td>5.652435</td>
      <td>1.530047e-08</td>
      <td>4.586591e-07</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
      <td>1.252748e-65</td>
      <td>171264</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1_174917639_174921559</td>
      <td>6.208699</td>
      <td>2.711848e-12</td>
      <td>1.562878e-10</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
      <td>1.252748e-65</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr2_93645657_93648162</td>
      <td>5.189615</td>
      <td>2.360560e-06</td>
      <td>4.324934e-05</td>
      <td>Alx4</td>
      <td>-10.292504</td>
      <td>5.458739e-18</td>
      <td>1.081356e-16</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr3_33140350_33142508</td>
      <td>5.534632</td>
      <td>6.455627e-08</td>
      <td>1.724615e-06</td>
      <td>Pex5l</td>
      <td>-7.590236</td>
      <td>2.351122e-24</td>
      <td>6.841972e-23</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr3_126498429_126499612</td>
      <td>5.337659</td>
      <td>5.594815e-07</td>
      <td>1.202787e-05</td>
      <td>Arsj</td>
      <td>-6.281259</td>
      <td>8.667831e-27</td>
      <td>2.783525e-25</td>
      <td>-58056</td>
    </tr>
    <tr>
      <th>5</th>
      <td>chr3_126657669_126660188</td>
      <td>-5.397757</td>
      <td>5.546104e-07</td>
      <td>1.202787e-05</td>
      <td>Arsj</td>
      <td>-6.281259</td>
      <td>8.667831e-27</td>
      <td>2.783525e-25</td>
      <td>-217296</td>
    </tr>
    <tr>
      <th>6</th>
      <td>chr3_132085669_132092999</td>
      <td>6.121213</td>
      <td>1.509244e-105</td>
      <td>2.850264e-102</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
      <td>5.513378e-175</td>
      <td>0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr3_132093815_132106295</td>
      <td>5.072335</td>
      <td>9.590057e-65</td>
      <td>6.924867e-62</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
      <td>5.513378e-175</td>
      <td>0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr3_141699368_141702608</td>
      <td>6.065565</td>
      <td>2.403341e-29</td>
      <td>5.413249e-27</td>
      <td>Bmpr1b</td>
      <td>-5.570116</td>
      <td>6.893076e-40</td>
      <td>3.655998e-38</td>
      <td>134528</td>
    </tr>
    <tr>
      <th>9</th>
      <td>chr4_68535893_68538867</td>
      <td>5.172571</td>
      <td>1.974087e-15</td>
      <td>1.648497e-13</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
      <td>5.726618e-09</td>
      <td>222505</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr4_68545353_68551019</td>
      <td>6.825275</td>
      <td>1.516093e-18</td>
      <td>1.684235e-16</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
      <td>5.726618e-09</td>
      <td>210353</td>
    </tr>
    <tr>
      <th>11</th>
      <td>chr4_118022206_118024522</td>
      <td>-5.332271</td>
      <td>1.094775e-06</td>
      <td>2.210347e-05</td>
      <td>Artn</td>
      <td>9.083650</td>
      <td>8.043160e-84</td>
      <td>1.066497e-81</td>
      <td>-92444</td>
    </tr>
    <tr>
      <th>12</th>
      <td>chr5_13001869_13003460</td>
      <td>5.957213</td>
      <td>2.037009e-10</td>
      <td>8.898685e-09</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
      <td>6.200968e-294</td>
      <td>393324</td>
    </tr>
    <tr>
      <th>13</th>
      <td>chr5_13276479_13283560</td>
      <td>7.727895</td>
      <td>3.536779e-34</td>
      <td>1.033708e-31</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
      <td>6.200968e-294</td>
      <td>113224</td>
    </tr>
    <tr>
      <th>14</th>
      <td>chr5_17080302_17082934</td>
      <td>5.024623</td>
      <td>9.959617e-06</td>
      <td>1.561421e-04</td>
      <td>Hgf</td>
      <td>-7.513756</td>
      <td>2.543270e-09</td>
      <td>2.523623e-08</td>
      <td>-460864</td>
    </tr>
    <tr>
      <th>15</th>
      <td>chr5_28465271_28467276</td>
      <td>-5.397757</td>
      <td>5.546104e-07</td>
      <td>1.202787e-05</td>
      <td>Shh</td>
      <td>5.565109</td>
      <td>2.158809e-156</td>
      <td>9.593279e-154</td>
      <td>0</td>
    </tr>
    <tr>
      <th>16</th>
      <td>chr5_38081825_38086331</td>
      <td>6.088429</td>
      <td>2.350343e-11</td>
      <td>1.192216e-09</td>
      <td>Nsg1</td>
      <td>-5.086635</td>
      <td>2.233941e-111</td>
      <td>4.963574e-109</td>
      <td>50862</td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr7_56523378_56528436</td>
      <td>6.721266</td>
      <td>2.699280e-17</td>
      <td>2.609056e-15</td>
      <td>Gabrg3</td>
      <td>-5.960560</td>
      <td>2.341610e-05</td>
      <td>1.323050e-04</td>
      <td>195806</td>
    </tr>
    <tr>
      <th>18</th>
      <td>chr7_111875691_111878313</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Galnt18</td>
      <td>-5.959958</td>
      <td>2.253335e-68</td>
      <td>2.218523e-66</td>
      <td>-95715</td>
    </tr>
    <tr>
      <th>19</th>
      <td>chr8_82329336_82331284</td>
      <td>5.189615</td>
      <td>2.360560e-06</td>
      <td>4.324934e-05</td>
      <td>Il15</td>
      <td>-5.185933</td>
      <td>8.096707e-04</td>
      <td>3.267016e-03</td>
      <td>340</td>
    </tr>
    <tr>
      <th>20</th>
      <td>chr8_102781569_102784308</td>
      <td>5.761340</td>
      <td>3.626346e-09</td>
      <td>1.222945e-07</td>
      <td>Cdh11</td>
      <td>-6.729272</td>
      <td>1.932941e-232</td>
      <td>1.986339e-229</td>
      <td>0</td>
    </tr>
    <tr>
      <th>21</th>
      <td>chr11_96341058_96343216</td>
      <td>5.337659</td>
      <td>5.594815e-07</td>
      <td>1.202787e-05</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
      <td>4.796719e-06</td>
      <td>0</td>
    </tr>
    <tr>
      <th>22</th>
      <td>chr11_96354206_96355627</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
      <td>4.796719e-06</td>
      <td>-6277</td>
    </tr>
    <tr>
      <th>23</th>
      <td>chr11_96354206_96355627</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Hoxb2</td>
      <td>-6.351656</td>
      <td>3.359549e-06</td>
      <td>2.173857e-05</td>
      <td>-193</td>
    </tr>
    <tr>
      <th>24</th>
      <td>chr12_56506841_56514912</td>
      <td>-9.259195</td>
      <td>2.083740e-95</td>
      <td>3.009288e-92</td>
      <td>Nkx2-1</td>
      <td>9.134857</td>
      <td>1.061953e-60</td>
      <td>9.189803e-59</td>
      <td>17023</td>
    </tr>
    <tr>
      <th>25</th>
      <td>chr13_117409406_117414817</td>
      <td>-5.127606</td>
      <td>3.397970e-15</td>
      <td>2.771547e-13</td>
      <td>Emb</td>
      <td>-6.021510</td>
      <td>3.917980e-06</td>
      <td>2.506593e-05</td>
      <td>-134992</td>
    </tr>
    <tr>
      <th>26</th>
      <td>chr14_66104432_66107303</td>
      <td>6.455571</td>
      <td>1.757505e-14</td>
      <td>1.344190e-12</td>
      <td>Ephx2</td>
      <td>-6.872942</td>
      <td>1.597944e-07</td>
      <td>1.247549e-06</td>
      <td>0</td>
    </tr>
    <tr>
      <th>27</th>
      <td>chr14_88460573_88463273</td>
      <td>5.594736</td>
      <td>3.142836e-08</td>
      <td>8.899628e-07</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
      <td>6.988068e-05</td>
      <td>1474</td>
    </tr>
    <tr>
      <th>28</th>
      <td>chr14_88898125_88900211</td>
      <td>-5.263671</td>
      <td>2.162151e-06</td>
      <td>4.039800e-05</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
      <td>6.988068e-05</td>
      <td>-426730</td>
    </tr>
    <tr>
      <th>29</th>
      <td>chr15_11692586_11694713</td>
      <td>6.046002</td>
      <td>4.827833e-11</td>
      <td>2.333231e-09</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>145183</td>
    </tr>
    <tr>
      <th>30</th>
      <td>chr15_11743634_11745456</td>
      <td>6.922288</td>
      <td>8.515240e-20</td>
      <td>1.034939e-17</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>94440</td>
    </tr>
    <tr>
      <th>31</th>
      <td>chr15_11782997_11784930</td>
      <td>5.957213</td>
      <td>2.037009e-10</td>
      <td>8.898685e-09</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>54966</td>
    </tr>
    <tr>
      <th>32</th>
      <td>chr18_53465649_53470088</td>
      <td>5.761340</td>
      <td>3.626346e-09</td>
      <td>1.222945e-07</td>
      <td>Prdm6</td>
      <td>-9.029699</td>
      <td>4.482978e-14</td>
      <td>6.960257e-13</td>
      <td>0</td>
    </tr>
    <tr>
      <th>33</th>
      <td>chr19_59159949_59161669</td>
      <td>-6.094880</td>
      <td>4.260481e-11</td>
      <td>2.067175e-09</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
      <td>3.071805e-29</td>
      <td>2816</td>
    </tr>
    <tr>
      <th>34</th>
      <td>chr19_59163812_59165938</td>
      <td>-5.788022</td>
      <td>4.812885e-09</td>
      <td>1.603272e-07</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
      <td>3.071805e-29</td>
      <td>0</td>
    </tr>
    <tr>
      <th>35</th>
      <td>chrX_93281556_93288401</td>
      <td>5.093695</td>
      <td>9.655833e-32</td>
      <td>2.469379e-29</td>
      <td>Arx</td>
      <td>-5.380834</td>
      <td>6.537245e-123</td>
      <td>1.853196e-120</td>
      <td>0</td>
    </tr>
    <tr>
      <th>36</th>
      <td>chrY_1009018_1011799</td>
      <td>-5.686862</td>
      <td>1.864054e-08</td>
      <td>5.507147e-07</td>
      <td>Eif2s3y</td>
      <td>11.645099</td>
      <td>9.414457e-23</td>
      <td>2.468780e-21</td>
      <td>0</td>
    </tr>
    <tr>
      <th>37</th>
      <td>chrY_1243715_1246316</td>
      <td>-5.520433</td>
      <td>1.425488e-07</td>
      <td>3.468498e-06</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
      <td>5.823136e-21</td>
      <td>14399</td>
    </tr>
    <tr>
      <th>38</th>
      <td>chrY_1243715_1246316</td>
      <td>-5.520433</td>
      <td>1.425488e-07</td>
      <td>3.468498e-06</td>
      <td>Uty</td>
      <td>8.369064</td>
      <td>1.224906e-22</td>
      <td>3.191744e-21</td>
      <td>0</td>
    </tr>
    <tr>
      <th>39</th>
      <td>chrY_1282482_1287504</td>
      <td>-5.191646</td>
      <td>4.272455e-06</td>
      <td>7.350598e-05</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
      <td>5.823136e-21</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Next, we visualize, for specific peaks, the magnitude of differential binding and the expression of nearby genes using the dotplot function. This function creates a dotplot with three columns (below). The first shows the relative binding of Brd4 at a chromsomal locus in males and females, the second and third columns show the expression levels of the nearest and second nearest genes in males and females.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot_bulk</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span> <span class="n">selected_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;Peak&quot;</span><span class="p">]))),</span> <span class="n">num_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">dotsize</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                   <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;plasma&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">16</span><span class="p">),</span> <span class="n">sort_by_chrom</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">topspace</span> <span class="o">=</span> <span class="mf">0.96</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Brd4_bulk_49_0.png" src="../../_images/tutorials_notebooks_Brd4_bulk_49_0.png" />
</div>
</div>
<p>We can see from these results that all of the peaks visualized here display differential binding, and either the nearest, or second nearest gene has differential expression. In most cases, increased Brd4 binding is associated with an increase in gene expression.</p>
<p>Pycallingcards provides functionality to explore relatinoships between differentially bound regions and nearby GWAS identified SNPs. Since the data analyzed here were collected in murine cells, we must first lift over any interesting peaks to the human genome before we can look for nearby disease associated SNPs. We use <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">liftover</a> to get the resuls.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">result_mapping</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="n">result</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Start mapping the peaks to the new genome.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 40/40 [00:00&lt;00:00, 69.43it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Start finding location of genes in the new genome.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 40/40 [00:00&lt;00:00, 287.78it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Peak</th>
      <th>logfoldchanges_peak</th>
      <th>Pvalue_peak</th>
      <th>Pvalue_adj_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
      <th>Pvalue_adj_gene</th>
      <th>Distance_peak_to_gene</th>
      <th>Chr_liftover</th>
      <th>Start_liftover</th>
      <th>End_liftover</th>
      <th>Chr_hg38</th>
      <th>Start_hg38</th>
      <th>End_hg38</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1_174659622_174662521</td>
      <td>5.652435</td>
      <td>1.530047e-08</td>
      <td>4.586591e-07</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
      <td>1.252748e-65</td>
      <td>171264</td>
      <td>chr1</td>
      <td>240280756</td>
      <td>240283835</td>
      <td>chr1</td>
      <td>240489572</td>
      <td>240612162</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1_174917639_174921559</td>
      <td>6.208699</td>
      <td>2.711848e-12</td>
      <td>1.562878e-10</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
      <td>1.252748e-65</td>
      <td>0</td>
      <td>chr1</td>
      <td>240608636</td>
      <td>240611870</td>
      <td>chr1</td>
      <td>240489572</td>
      <td>240612162</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr2_93645657_93648162</td>
      <td>5.189615</td>
      <td>2.360560e-06</td>
      <td>4.324934e-05</td>
      <td>Alx4</td>
      <td>-10.292504</td>
      <td>5.458739e-18</td>
      <td>1.081356e-16</td>
      <td>0</td>
      <td>chr11</td>
      <td>44303696</td>
      <td>44306415</td>
      <td>chr11</td>
      <td>44260727</td>
      <td>44310166</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr3_33140350_33142508</td>
      <td>5.534632</td>
      <td>6.455627e-08</td>
      <td>1.724615e-06</td>
      <td>Pex5l</td>
      <td>-7.590236</td>
      <td>2.351122e-24</td>
      <td>6.841972e-23</td>
      <td>0</td>
      <td>chr3</td>
      <td>180033883</td>
      <td>180036186</td>
      <td>chr3</td>
      <td>179794958</td>
      <td>180037053</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr3_126498429_126499612</td>
      <td>5.337659</td>
      <td>5.594815e-07</td>
      <td>1.202787e-05</td>
      <td>Arsj</td>
      <td>-6.281259</td>
      <td>8.667831e-27</td>
      <td>2.783525e-25</td>
      <td>-58056</td>
      <td>chr4</td>
      <td>113839066</td>
      <td>113844568</td>
      <td>chr4</td>
      <td>113900283</td>
      <td>113979722</td>
    </tr>
    <tr>
      <th>5</th>
      <td>chr3_126657669_126660188</td>
      <td>-5.397757</td>
      <td>5.546104e-07</td>
      <td>1.202787e-05</td>
      <td>Arsj</td>
      <td>-6.281259</td>
      <td>8.667831e-27</td>
      <td>2.783525e-25</td>
      <td>-217296</td>
      <td>chr4</td>
      <td>113683844</td>
      <td>113686758</td>
      <td>chr4</td>
      <td>113900283</td>
      <td>113979722</td>
    </tr>
    <tr>
      <th>6</th>
      <td>chr3_132085669_132092999</td>
      <td>6.121213</td>
      <td>1.509244e-105</td>
      <td>2.850264e-102</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
      <td>5.513378e-175</td>
      <td>0</td>
      <td>chr4</td>
      <td>107029387</td>
      <td>107035946</td>
      <td>chr4</td>
      <td>106921801</td>
      <td>107036296</td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr3_132093815_132106295</td>
      <td>5.072335</td>
      <td>9.590057e-65</td>
      <td>6.924867e-62</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
      <td>5.513378e-175</td>
      <td>0</td>
      <td>chr4</td>
      <td>107015456</td>
      <td>107028826</td>
      <td>chr4</td>
      <td>106921801</td>
      <td>107036296</td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr3_141699368_141702608</td>
      <td>6.065565</td>
      <td>2.403341e-29</td>
      <td>5.413249e-27</td>
      <td>Bmpr1b</td>
      <td>-5.570116</td>
      <td>6.893076e-40</td>
      <td>3.655998e-38</td>
      <td>134528</td>
      <td>chr4</td>
      <td>95306647</td>
      <td>95309431</td>
      <td>chr4</td>
      <td>94757976</td>
      <td>95158450</td>
    </tr>
    <tr>
      <th>9</th>
      <td>chr4_68535893_68538867</td>
      <td>5.172571</td>
      <td>1.974087e-15</td>
      <td>1.648497e-13</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
      <td>5.726618e-09</td>
      <td>222505</td>
      <td>chr9</td>
      <td>118993297</td>
      <td>118996641</td>
      <td>chr9</td>
      <td>119166629</td>
      <td>119369461</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr4_68545353_68551019</td>
      <td>6.825275</td>
      <td>1.516093e-18</td>
      <td>1.684235e-16</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
      <td>5.726618e-09</td>
      <td>210353</td>
      <td>chr9</td>
      <td>119002823</td>
      <td>119008509</td>
      <td>chr9</td>
      <td>119166629</td>
      <td>119369461</td>
    </tr>
    <tr>
      <th>11</th>
      <td>chr4_118022206_118024522</td>
      <td>-5.332271</td>
      <td>1.094775e-06</td>
      <td>2.210347e-05</td>
      <td>Artn</td>
      <td>9.083650</td>
      <td>8.043160e-84</td>
      <td>1.066497e-81</td>
      <td>-92444</td>
      <td>chr1</td>
      <td>43824383</td>
      <td>43825212</td>
      <td>chr1</td>
      <td>43933319</td>
      <td>43937240</td>
    </tr>
    <tr>
      <th>12</th>
      <td>chr5_13001869_13003460</td>
      <td>5.957213</td>
      <td>2.037009e-10</td>
      <td>8.898685e-09</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
      <td>6.200968e-294</td>
      <td>393324</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr7</td>
      <td>83958342</td>
      <td>84194901</td>
    </tr>
    <tr>
      <th>13</th>
      <td>chr5_13276479_13283560</td>
      <td>7.727895</td>
      <td>3.536779e-34</td>
      <td>1.033708e-31</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
      <td>6.200968e-294</td>
      <td>113224</td>
      <td>chr7</td>
      <td>84314176</td>
      <td>84318280</td>
      <td>chr7</td>
      <td>83958342</td>
      <td>84194901</td>
    </tr>
    <tr>
      <th>14</th>
      <td>chr5_17080302_17082934</td>
      <td>5.024623</td>
      <td>9.959617e-06</td>
      <td>1.561421e-04</td>
      <td>Hgf</td>
      <td>-7.513756</td>
      <td>2.543270e-09</td>
      <td>2.523623e-08</td>
      <td>-460864</td>
      <td>chr7</td>
      <td>81335127</td>
      <td>81335733</td>
      <td>chr7</td>
      <td>81699007</td>
      <td>81770198</td>
    </tr>
    <tr>
      <th>15</th>
      <td>chr5_28465271_28467276</td>
      <td>-5.397757</td>
      <td>5.546104e-07</td>
      <td>1.202787e-05</td>
      <td>Shh</td>
      <td>5.565109</td>
      <td>2.158809e-156</td>
      <td>9.593279e-154</td>
      <td>0</td>
      <td>chr7</td>
      <td>155810616</td>
      <td>155812624</td>
      <td>chr7</td>
      <td>155799983</td>
      <td>155812273</td>
    </tr>
    <tr>
      <th>16</th>
      <td>chr5_38081825_38086331</td>
      <td>6.088429</td>
      <td>2.350343e-11</td>
      <td>1.192216e-09</td>
      <td>Nsg1</td>
      <td>-5.086635</td>
      <td>2.233941e-111</td>
      <td>4.963574e-109</td>
      <td>50862</td>
      <td>chr4</td>
      <td>4488485</td>
      <td>4493470</td>
      <td>chr4</td>
      <td>4386255</td>
      <td>4419058</td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr7_56523378_56528436</td>
      <td>6.721266</td>
      <td>2.699280e-17</td>
      <td>2.609056e-15</td>
      <td>Gabrg3</td>
      <td>-5.960560</td>
      <td>2.341610e-05</td>
      <td>1.323050e-04</td>
      <td>195806</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr15</td>
      <td>26971281</td>
      <td>27533227</td>
    </tr>
    <tr>
      <th>18</th>
      <td>chr7_111875691_111878313</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Galnt18</td>
      <td>-5.959958</td>
      <td>2.253335e-68</td>
      <td>2.218523e-66</td>
      <td>-95715</td>
      <td>chr11</td>
      <td>11723801</td>
      <td>11725419</td>
      <td>chr11</td>
      <td>11270873</td>
      <td>11622014</td>
    </tr>
    <tr>
      <th>19</th>
      <td>chr8_82329336_82331284</td>
      <td>5.189615</td>
      <td>2.360560e-06</td>
      <td>4.324934e-05</td>
      <td>Il15</td>
      <td>-5.185933</td>
      <td>8.096707e-04</td>
      <td>3.267016e-03</td>
      <td>340</td>
      <td>chr4</td>
      <td>141733679</td>
      <td>141736013</td>
      <td>chr4</td>
      <td>141636595</td>
      <td>141733987</td>
    </tr>
    <tr>
      <th>20</th>
      <td>chr8_102781569_102784308</td>
      <td>5.761340</td>
      <td>3.626346e-09</td>
      <td>1.222945e-07</td>
      <td>Cdh11</td>
      <td>-6.729272</td>
      <td>1.932941e-232</td>
      <td>1.986339e-229</td>
      <td>0</td>
      <td>chr16</td>
      <td>65118621</td>
      <td>65121296</td>
      <td>chr16</td>
      <td>64943752</td>
      <td>65122137</td>
    </tr>
    <tr>
      <th>21</th>
      <td>chr11_96341058_96343216</td>
      <td>5.337659</td>
      <td>5.594815e-07</td>
      <td>1.202787e-05</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
      <td>4.796719e-06</td>
      <td>0</td>
      <td>chr17</td>
      <td>48553538</td>
      <td>48555622</td>
      <td>chr17</td>
      <td>48548869</td>
      <td>48582622</td>
    </tr>
    <tr>
      <th>22</th>
      <td>chr11_96354206_96355627</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
      <td>4.796719e-06</td>
      <td>-6277</td>
      <td>chr17</td>
      <td>48540790</td>
      <td>48542479</td>
      <td>chr17</td>
      <td>48548869</td>
      <td>48582622</td>
    </tr>
    <tr>
      <th>23</th>
      <td>chr11_96354206_96355627</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Hoxb2</td>
      <td>-6.351656</td>
      <td>3.359549e-06</td>
      <td>2.173857e-05</td>
      <td>-193</td>
      <td>chr17</td>
      <td>48540790</td>
      <td>48542479</td>
      <td>chr17</td>
      <td>48542655</td>
      <td>48545031</td>
    </tr>
    <tr>
      <th>24</th>
      <td>chr12_56506841_56514912</td>
      <td>-9.259195</td>
      <td>2.083740e-95</td>
      <td>3.009288e-92</td>
      <td>Nkx2-1</td>
      <td>9.134857</td>
      <td>1.061953e-60</td>
      <td>9.189803e-59</td>
      <td>17023</td>
      <td>chr14</td>
      <td>36491565</td>
      <td>36499712</td>
      <td>chr14</td>
      <td>36516398</td>
      <td>36520225</td>
    </tr>
    <tr>
      <th>25</th>
      <td>chr13_117409406_117414817</td>
      <td>-5.127606</td>
      <td>3.397970e-15</td>
      <td>2.771547e-13</td>
      <td>Emb</td>
      <td>-6.021510</td>
      <td>3.917980e-06</td>
      <td>2.506593e-05</td>
      <td>-134992</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr5</td>
      <td>50396196</td>
      <td>50441400</td>
    </tr>
    <tr>
      <th>26</th>
      <td>chr14_66104432_66107303</td>
      <td>6.455571</td>
      <td>1.757505e-14</td>
      <td>1.344190e-12</td>
      <td>Ephx2</td>
      <td>-6.872942</td>
      <td>1.597944e-07</td>
      <td>1.247549e-06</td>
      <td>0</td>
      <td>chr8</td>
      <td>27506903</td>
      <td>27511135</td>
      <td>chr8</td>
      <td>27491001</td>
      <td>27544922</td>
    </tr>
    <tr>
      <th>27</th>
      <td>chr14_88460573_88463273</td>
      <td>5.594736</td>
      <td>3.142836e-08</td>
      <td>8.899628e-07</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
      <td>6.988068e-05</td>
      <td>1474</td>
      <td>chr13</td>
      <td>61406547</td>
      <td>61407676</td>
      <td>chr13</td>
      <td>61409685</td>
      <td>61415522</td>
    </tr>
    <tr>
      <th>28</th>
      <td>chr14_88898125_88900211</td>
      <td>-5.263671</td>
      <td>2.162151e-06</td>
      <td>4.039800e-05</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
      <td>6.988068e-05</td>
      <td>-426730</td>
      <td>chr13</td>
      <td>61862674</td>
      <td>61862983</td>
      <td>chr13</td>
      <td>61409685</td>
      <td>61415522</td>
    </tr>
    <tr>
      <th>29</th>
      <td>chr15_11692586_11694713</td>
      <td>6.046002</td>
      <td>4.827833e-11</td>
      <td>2.333231e-09</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>145183</td>
      <td>chr5</td>
      <td>33023231</td>
      <td>33024567</td>
      <td>chr5</td>
      <td>32710636</td>
      <td>32791724</td>
    </tr>
    <tr>
      <th>30</th>
      <td>chr15_11743634_11745456</td>
      <td>6.922288</td>
      <td>8.515240e-20</td>
      <td>1.034939e-17</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>94440</td>
      <td>chr5</td>
      <td>32929090</td>
      <td>32932829</td>
      <td>chr5</td>
      <td>32710636</td>
      <td>32791724</td>
    </tr>
    <tr>
      <th>31</th>
      <td>chr15_11782997_11784930</td>
      <td>5.957213</td>
      <td>2.037009e-10</td>
      <td>8.898685e-09</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>54966</td>
      <td>chr5</td>
      <td>32864592</td>
      <td>32866245</td>
      <td>chr5</td>
      <td>32710636</td>
      <td>32791724</td>
    </tr>
    <tr>
      <th>32</th>
      <td>chr18_53465649_53470088</td>
      <td>5.761340</td>
      <td>3.626346e-09</td>
      <td>1.222945e-07</td>
      <td>Prdm6</td>
      <td>-9.029699</td>
      <td>4.482978e-14</td>
      <td>6.960257e-13</td>
      <td>0</td>
      <td>chr5</td>
      <td>123091173</td>
      <td>123096087</td>
      <td>chr5</td>
      <td>123089102</td>
      <td>123194266</td>
    </tr>
    <tr>
      <th>33</th>
      <td>chr19_59159949_59161669</td>
      <td>-6.094880</td>
      <td>4.260481e-11</td>
      <td>2.067175e-09</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
      <td>3.071805e-29</td>
      <td>2816</td>
      <td>chr10</td>
      <td>117127894</td>
      <td>117129270</td>
      <td>chr10</td>
      <td>117128520</td>
      <td>117138301</td>
    </tr>
    <tr>
      <th>34</th>
      <td>chr19_59163812_59165938</td>
      <td>-5.788022</td>
      <td>4.812885e-09</td>
      <td>1.603272e-07</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
      <td>3.071805e-29</td>
      <td>0</td>
      <td>chr10</td>
      <td>117131583</td>
      <td>117133776</td>
      <td>chr10</td>
      <td>117128520</td>
      <td>117138301</td>
    </tr>
    <tr>
      <th>35</th>
      <td>chrX_93281556_93288401</td>
      <td>5.093695</td>
      <td>9.655833e-32</td>
      <td>2.469379e-29</td>
      <td>Arx</td>
      <td>-5.380834</td>
      <td>6.537245e-123</td>
      <td>1.853196e-120</td>
      <td>0</td>
      <td>chrX</td>
      <td>25013967</td>
      <td>25020980</td>
      <td>chrX</td>
      <td>25003695</td>
      <td>25015948</td>
    </tr>
    <tr>
      <th>36</th>
      <td>chrY_1009018_1011799</td>
      <td>-5.686862</td>
      <td>1.864054e-08</td>
      <td>5.507147e-07</td>
      <td>Eif2s3y</td>
      <td>11.645099</td>
      <td>9.414457e-23</td>
      <td>2.468780e-21</td>
      <td>0</td>
      <td>chrX</td>
      <td>24054686</td>
      <td>24057370</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>37</th>
      <td>chrY_1243715_1246316</td>
      <td>-5.520433</td>
      <td>1.425488e-07</td>
      <td>3.468498e-06</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
      <td>5.823136e-21</td>
      <td>14399</td>
      <td>chrY</td>
      <td>13478375</td>
      <td>13480067</td>
      <td>chrY</td>
      <td>12903998</td>
      <td>12920478</td>
    </tr>
    <tr>
      <th>38</th>
      <td>chrY_1243715_1246316</td>
      <td>-5.520433</td>
      <td>1.425488e-07</td>
      <td>3.468498e-06</td>
      <td>Uty</td>
      <td>8.369064</td>
      <td>1.224906e-22</td>
      <td>3.191744e-21</td>
      <td>0</td>
      <td>chrY</td>
      <td>13478375</td>
      <td>13480067</td>
      <td>chrY</td>
      <td>13248378</td>
      <td>13480670</td>
    </tr>
    <tr>
      <th>39</th>
      <td>chrY_1282482_1287504</td>
      <td>-5.191646</td>
      <td>4.272455e-06</td>
      <td>7.350598e-05</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
      <td>5.823136e-21</td>
      <td>0</td>
      <td>chrY</td>
      <td>12903183</td>
      <td>12909538</td>
      <td>chrY</td>
      <td>12903998</td>
      <td>12920478</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">GWAS_result</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">GWAS</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="n">GWAS_result</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Peak</th>
      <th>logfoldchanges_peak</th>
      <th>Pvalue_peak</th>
      <th>Pvalue_adj_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
      <th>Pvalue_adj_gene</th>
      <th>Distance_peak_to_gene</th>
      <th>Chr_liftover</th>
      <th>Start_liftover</th>
      <th>End_liftover</th>
      <th>Chr_hg38</th>
      <th>Start_hg38</th>
      <th>End_hg38</th>
      <th>GWAS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1_174659622_174662521</td>
      <td>5.652435</td>
      <td>1.530047e-08</td>
      <td>4.586591e-07</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
      <td>1.252748e-65</td>
      <td>171264</td>
      <td>chr1</td>
      <td>240280756</td>
      <td>240283835</td>
      <td>chr1</td>
      <td>240489572</td>
      <td>240612162</td>
      <td>Coronary heart disease</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1_174917639_174921559</td>
      <td>6.208699</td>
      <td>2.711848e-12</td>
      <td>1.562878e-10</td>
      <td>Grem2</td>
      <td>-10.401831</td>
      <td>1.280025e-67</td>
      <td>1.252748e-65</td>
      <td>0</td>
      <td>chr1</td>
      <td>240608636</td>
      <td>240611870</td>
      <td>chr1</td>
      <td>240489572</td>
      <td>240612162</td>
      <td></td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr2_93645657_93648162</td>
      <td>5.189615</td>
      <td>2.360560e-06</td>
      <td>4.324934e-05</td>
      <td>Alx4</td>
      <td>-10.292504</td>
      <td>5.458739e-18</td>
      <td>1.081356e-16</td>
      <td>0</td>
      <td>chr11</td>
      <td>44303696</td>
      <td>44306415</td>
      <td>chr11</td>
      <td>44260727</td>
      <td>44310166</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr3_33140350_33142508</td>
      <td>5.534632</td>
      <td>6.455627e-08</td>
      <td>1.724615e-06</td>
      <td>Pex5l</td>
      <td>-7.590236</td>
      <td>2.351122e-24</td>
      <td>6.841972e-23</td>
      <td>0</td>
      <td>chr3</td>
      <td>180033883</td>
      <td>180036186</td>
      <td>chr3</td>
      <td>179794958</td>
      <td>180037053</td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr3_126498429_126499612</td>
      <td>5.337659</td>
      <td>5.594815e-07</td>
      <td>1.202787e-05</td>
      <td>Arsj</td>
      <td>-6.281259</td>
      <td>8.667831e-27</td>
      <td>2.783525e-25</td>
      <td>-58056</td>
      <td>chr4</td>
      <td>113839066</td>
      <td>113844568</td>
      <td>chr4</td>
      <td>113900283</td>
      <td>113979722</td>
      <td></td>
    </tr>
    <tr>
      <th>5</th>
      <td>chr3_126657669_126660188</td>
      <td>-5.397757</td>
      <td>5.546104e-07</td>
      <td>1.202787e-05</td>
      <td>Arsj</td>
      <td>-6.281259</td>
      <td>8.667831e-27</td>
      <td>2.783525e-25</td>
      <td>-217296</td>
      <td>chr4</td>
      <td>113683844</td>
      <td>113686758</td>
      <td>chr4</td>
      <td>113900283</td>
      <td>113979722</td>
      <td></td>
    </tr>
    <tr>
      <th>6</th>
      <td>chr3_132085669_132092999</td>
      <td>6.121213</td>
      <td>1.509244e-105</td>
      <td>2.850264e-102</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
      <td>5.513378e-175</td>
      <td>0</td>
      <td>chr4</td>
      <td>107029387</td>
      <td>107035946</td>
      <td>chr4</td>
      <td>106921801</td>
      <td>107036296</td>
      <td></td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr3_132093815_132106295</td>
      <td>5.072335</td>
      <td>9.590057e-65</td>
      <td>6.924867e-62</td>
      <td>Dkk2</td>
      <td>-6.916087</td>
      <td>9.389039e-178</td>
      <td>5.513378e-175</td>
      <td>0</td>
      <td>chr4</td>
      <td>107015456</td>
      <td>107028826</td>
      <td>chr4</td>
      <td>106921801</td>
      <td>107036296</td>
      <td></td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr3_141699368_141702608</td>
      <td>6.065565</td>
      <td>2.403341e-29</td>
      <td>5.413249e-27</td>
      <td>Bmpr1b</td>
      <td>-5.570116</td>
      <td>6.893076e-40</td>
      <td>3.655998e-38</td>
      <td>134528</td>
      <td>chr4</td>
      <td>95306647</td>
      <td>95309431</td>
      <td>chr4</td>
      <td>94757976</td>
      <td>95158450</td>
      <td></td>
    </tr>
    <tr>
      <th>9</th>
      <td>chr4_68535893_68538867</td>
      <td>5.172571</td>
      <td>1.974087e-15</td>
      <td>1.648497e-13</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
      <td>5.726618e-09</td>
      <td>222505</td>
      <td>chr9</td>
      <td>118993297</td>
      <td>118996641</td>
      <td>chr9</td>
      <td>119166629</td>
      <td>119369461</td>
      <td>Gut microbiota (bacterial taxa, hurdle binary ...</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr4_68545353_68551019</td>
      <td>6.825275</td>
      <td>1.516093e-18</td>
      <td>1.684235e-16</td>
      <td>Brinp1</td>
      <td>-7.758307</td>
      <td>5.384595e-10</td>
      <td>5.726618e-09</td>
      <td>210353</td>
      <td>chr9</td>
      <td>119002823</td>
      <td>119008509</td>
      <td>chr9</td>
      <td>119166629</td>
      <td>119369461</td>
      <td></td>
    </tr>
    <tr>
      <th>11</th>
      <td>chr4_118022206_118024522</td>
      <td>-5.332271</td>
      <td>1.094775e-06</td>
      <td>2.210347e-05</td>
      <td>Artn</td>
      <td>9.083650</td>
      <td>8.043160e-84</td>
      <td>1.066497e-81</td>
      <td>-92444</td>
      <td>chr1</td>
      <td>43824383</td>
      <td>43825212</td>
      <td>chr1</td>
      <td>43933319</td>
      <td>43937240</td>
      <td></td>
    </tr>
    <tr>
      <th>12</th>
      <td>chr5_13001869_13003460</td>
      <td>5.957213</td>
      <td>2.037009e-10</td>
      <td>8.898685e-09</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
      <td>6.200968e-294</td>
      <td>393324</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr7</td>
      <td>83958342</td>
      <td>84194901</td>
      <td></td>
    </tr>
    <tr>
      <th>13</th>
      <td>chr5_13276479_13283560</td>
      <td>7.727895</td>
      <td>3.536779e-34</td>
      <td>1.033708e-31</td>
      <td>Sema3a</td>
      <td>-5.771471</td>
      <td>3.017136e-297</td>
      <td>6.200968e-294</td>
      <td>113224</td>
      <td>chr7</td>
      <td>84314176</td>
      <td>84318280</td>
      <td>chr7</td>
      <td>83958342</td>
      <td>84194901</td>
      <td></td>
    </tr>
    <tr>
      <th>14</th>
      <td>chr5_17080302_17082934</td>
      <td>5.024623</td>
      <td>9.959617e-06</td>
      <td>1.561421e-04</td>
      <td>Hgf</td>
      <td>-7.513756</td>
      <td>2.543270e-09</td>
      <td>2.523623e-08</td>
      <td>-460864</td>
      <td>chr7</td>
      <td>81335127</td>
      <td>81335733</td>
      <td>chr7</td>
      <td>81699007</td>
      <td>81770198</td>
      <td></td>
    </tr>
    <tr>
      <th>15</th>
      <td>chr5_28465271_28467276</td>
      <td>-5.397757</td>
      <td>5.546104e-07</td>
      <td>1.202787e-05</td>
      <td>Shh</td>
      <td>5.565109</td>
      <td>2.158809e-156</td>
      <td>9.593279e-154</td>
      <td>0</td>
      <td>chr7</td>
      <td>155810616</td>
      <td>155812624</td>
      <td>chr7</td>
      <td>155799983</td>
      <td>155812273</td>
      <td>Urate levels</td>
    </tr>
    <tr>
      <th>16</th>
      <td>chr5_38081825_38086331</td>
      <td>6.088429</td>
      <td>2.350343e-11</td>
      <td>1.192216e-09</td>
      <td>Nsg1</td>
      <td>-5.086635</td>
      <td>2.233941e-111</td>
      <td>4.963574e-109</td>
      <td>50862</td>
      <td>chr4</td>
      <td>4488485</td>
      <td>4493470</td>
      <td>chr4</td>
      <td>4386255</td>
      <td>4419058</td>
      <td></td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr7_56523378_56528436</td>
      <td>6.721266</td>
      <td>2.699280e-17</td>
      <td>2.609056e-15</td>
      <td>Gabrg3</td>
      <td>-5.960560</td>
      <td>2.341610e-05</td>
      <td>1.323050e-04</td>
      <td>195806</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr15</td>
      <td>26971281</td>
      <td>27533227</td>
      <td></td>
    </tr>
    <tr>
      <th>18</th>
      <td>chr7_111875691_111878313</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Galnt18</td>
      <td>-5.959958</td>
      <td>2.253335e-68</td>
      <td>2.218523e-66</td>
      <td>-95715</td>
      <td>chr11</td>
      <td>11723801</td>
      <td>11725419</td>
      <td>chr11</td>
      <td>11270873</td>
      <td>11622014</td>
      <td></td>
    </tr>
    <tr>
      <th>19</th>
      <td>chr8_82329336_82331284</td>
      <td>5.189615</td>
      <td>2.360560e-06</td>
      <td>4.324934e-05</td>
      <td>Il15</td>
      <td>-5.185933</td>
      <td>8.096707e-04</td>
      <td>3.267016e-03</td>
      <td>340</td>
      <td>chr4</td>
      <td>141733679</td>
      <td>141736013</td>
      <td>chr4</td>
      <td>141636595</td>
      <td>141733987</td>
      <td></td>
    </tr>
    <tr>
      <th>20</th>
      <td>chr8_102781569_102784308</td>
      <td>5.761340</td>
      <td>3.626346e-09</td>
      <td>1.222945e-07</td>
      <td>Cdh11</td>
      <td>-6.729272</td>
      <td>1.932941e-232</td>
      <td>1.986339e-229</td>
      <td>0</td>
      <td>chr16</td>
      <td>65118621</td>
      <td>65121296</td>
      <td>chr16</td>
      <td>64943752</td>
      <td>65122137</td>
      <td></td>
    </tr>
    <tr>
      <th>21</th>
      <td>chr11_96341058_96343216</td>
      <td>5.337659</td>
      <td>5.594815e-07</td>
      <td>1.202787e-05</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
      <td>4.796719e-06</td>
      <td>0</td>
      <td>chr17</td>
      <td>48553538</td>
      <td>48555622</td>
      <td>chr17</td>
      <td>48548869</td>
      <td>48582622</td>
      <td>Migraine; Insomnia</td>
    </tr>
    <tr>
      <th>22</th>
      <td>chr11_96354206_96355627</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Hoxb3</td>
      <td>-6.433657</td>
      <td>6.669080e-07</td>
      <td>4.796719e-06</td>
      <td>-6277</td>
      <td>chr17</td>
      <td>48540790</td>
      <td>48542479</td>
      <td>chr17</td>
      <td>48548869</td>
      <td>48582622</td>
      <td>Cystatin C levels</td>
    </tr>
    <tr>
      <th>23</th>
      <td>chr11_96354206_96355627</td>
      <td>5.109476</td>
      <td>4.848742e-06</td>
      <td>8.221096e-05</td>
      <td>Hoxb2</td>
      <td>-6.351656</td>
      <td>3.359549e-06</td>
      <td>2.173857e-05</td>
      <td>-193</td>
      <td>chr17</td>
      <td>48540790</td>
      <td>48542479</td>
      <td>chr17</td>
      <td>48542655</td>
      <td>48545031</td>
      <td>Cystatin C levels</td>
    </tr>
    <tr>
      <th>24</th>
      <td>chr12_56506841_56514912</td>
      <td>-9.259195</td>
      <td>2.083740e-95</td>
      <td>3.009288e-92</td>
      <td>Nkx2-1</td>
      <td>9.134857</td>
      <td>1.061953e-60</td>
      <td>9.189803e-59</td>
      <td>17023</td>
      <td>chr14</td>
      <td>36491565</td>
      <td>36499712</td>
      <td>chr14</td>
      <td>36516398</td>
      <td>36520225</td>
      <td>Nonsyndromic orofacial cleft x sex interaction...</td>
    </tr>
    <tr>
      <th>25</th>
      <td>chr13_117409406_117414817</td>
      <td>-5.127606</td>
      <td>3.397970e-15</td>
      <td>2.771547e-13</td>
      <td>Emb</td>
      <td>-6.021510</td>
      <td>3.917980e-06</td>
      <td>2.506593e-05</td>
      <td>-134992</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr5</td>
      <td>50396196</td>
      <td>50441400</td>
      <td></td>
    </tr>
    <tr>
      <th>26</th>
      <td>chr14_66104432_66107303</td>
      <td>6.455571</td>
      <td>1.757505e-14</td>
      <td>1.344190e-12</td>
      <td>Ephx2</td>
      <td>-6.872942</td>
      <td>1.597944e-07</td>
      <td>1.247549e-06</td>
      <td>0</td>
      <td>chr8</td>
      <td>27506903</td>
      <td>27511135</td>
      <td>chr8</td>
      <td>27491001</td>
      <td>27544922</td>
      <td>Childhood ear infection</td>
    </tr>
    <tr>
      <th>27</th>
      <td>chr14_88460573_88463273</td>
      <td>5.594736</td>
      <td>3.142836e-08</td>
      <td>8.899628e-07</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
      <td>6.988068e-05</td>
      <td>1474</td>
      <td>chr13</td>
      <td>61406547</td>
      <td>61407676</td>
      <td>chr13</td>
      <td>61409685</td>
      <td>61415522</td>
      <td></td>
    </tr>
    <tr>
      <th>28</th>
      <td>chr14_88898125_88900211</td>
      <td>-5.263671</td>
      <td>2.162151e-06</td>
      <td>4.039800e-05</td>
      <td>Pcdh20</td>
      <td>-6.097424</td>
      <td>1.179837e-05</td>
      <td>6.988068e-05</td>
      <td>-426730</td>
      <td>chr13</td>
      <td>61862674</td>
      <td>61862983</td>
      <td>chr13</td>
      <td>61409685</td>
      <td>61415522</td>
      <td></td>
    </tr>
    <tr>
      <th>29</th>
      <td>chr15_11692586_11694713</td>
      <td>6.046002</td>
      <td>4.827833e-11</td>
      <td>2.333231e-09</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>145183</td>
      <td>chr5</td>
      <td>33023231</td>
      <td>33024567</td>
      <td>chr5</td>
      <td>32710636</td>
      <td>32791724</td>
      <td></td>
    </tr>
    <tr>
      <th>30</th>
      <td>chr15_11743634_11745456</td>
      <td>6.922288</td>
      <td>8.515240e-20</td>
      <td>1.034939e-17</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>94440</td>
      <td>chr5</td>
      <td>32929090</td>
      <td>32932829</td>
      <td>chr5</td>
      <td>32710636</td>
      <td>32791724</td>
      <td></td>
    </tr>
    <tr>
      <th>31</th>
      <td>chr15_11782997_11784930</td>
      <td>5.957213</td>
      <td>2.037009e-10</td>
      <td>8.898685e-09</td>
      <td>Npr3</td>
      <td>-8.356128</td>
      <td>2.957230e-242</td>
      <td>3.473055e-239</td>
      <td>54966</td>
      <td>chr5</td>
      <td>32864592</td>
      <td>32866245</td>
      <td>chr5</td>
      <td>32710636</td>
      <td>32791724</td>
      <td></td>
    </tr>
    <tr>
      <th>32</th>
      <td>chr18_53465649_53470088</td>
      <td>5.761340</td>
      <td>3.626346e-09</td>
      <td>1.222945e-07</td>
      <td>Prdm6</td>
      <td>-9.029699</td>
      <td>4.482978e-14</td>
      <td>6.960257e-13</td>
      <td>0</td>
      <td>chr5</td>
      <td>123091173</td>
      <td>123096087</td>
      <td>chr5</td>
      <td>123089102</td>
      <td>123194266</td>
      <td></td>
    </tr>
    <tr>
      <th>33</th>
      <td>chr19_59159949_59161669</td>
      <td>-6.094880</td>
      <td>4.260481e-11</td>
      <td>2.067175e-09</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
      <td>3.071805e-29</td>
      <td>2816</td>
      <td>chr10</td>
      <td>117127894</td>
      <td>117129270</td>
      <td>chr10</td>
      <td>117128520</td>
      <td>117138301</td>
      <td></td>
    </tr>
    <tr>
      <th>34</th>
      <td>chr19_59163812_59165938</td>
      <td>-5.788022</td>
      <td>4.812885e-09</td>
      <td>1.603272e-07</td>
      <td>Vax1</td>
      <td>7.636277</td>
      <td>8.070915e-31</td>
      <td>3.071805e-29</td>
      <td>0</td>
      <td>chr10</td>
      <td>117131583</td>
      <td>117133776</td>
      <td>chr10</td>
      <td>117128520</td>
      <td>117138301</td>
      <td>Age at first birth; Nonsyndromic cleft lip wit...</td>
    </tr>
    <tr>
      <th>35</th>
      <td>chrX_93281556_93288401</td>
      <td>5.093695</td>
      <td>9.655833e-32</td>
      <td>2.469379e-29</td>
      <td>Arx</td>
      <td>-5.380834</td>
      <td>6.537245e-123</td>
      <td>1.853196e-120</td>
      <td>0</td>
      <td>chrX</td>
      <td>25013967</td>
      <td>25020980</td>
      <td>chrX</td>
      <td>25003695</td>
      <td>25015948</td>
      <td></td>
    </tr>
    <tr>
      <th>36</th>
      <td>chrY_1009018_1011799</td>
      <td>-5.686862</td>
      <td>1.864054e-08</td>
      <td>5.507147e-07</td>
      <td>Eif2s3y</td>
      <td>11.645099</td>
      <td>9.414457e-23</td>
      <td>2.468780e-21</td>
      <td>0</td>
      <td>chrX</td>
      <td>24054686</td>
      <td>24057370</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>37</th>
      <td>chrY_1243715_1246316</td>
      <td>-5.520433</td>
      <td>1.425488e-07</td>
      <td>3.468498e-06</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
      <td>5.823136e-21</td>
      <td>14399</td>
      <td>chrY</td>
      <td>13478375</td>
      <td>13480067</td>
      <td>chrY</td>
      <td>12903998</td>
      <td>12920478</td>
      <td></td>
    </tr>
    <tr>
      <th>38</th>
      <td>chrY_1243715_1246316</td>
      <td>-5.520433</td>
      <td>1.425488e-07</td>
      <td>3.468498e-06</td>
      <td>Uty</td>
      <td>8.369064</td>
      <td>1.224906e-22</td>
      <td>3.191744e-21</td>
      <td>0</td>
      <td>chrY</td>
      <td>13478375</td>
      <td>13480067</td>
      <td>chrY</td>
      <td>13248378</td>
      <td>13480670</td>
      <td></td>
    </tr>
    <tr>
      <th>39</th>
      <td>chrY_1282482_1287504</td>
      <td>-5.191646</td>
      <td>4.272455e-06</td>
      <td>7.350598e-05</td>
      <td>Ddx3y</td>
      <td>11.525927</td>
      <td>2.291430e-22</td>
      <td>5.823136e-21</td>
      <td>0</td>
      <td>chrY</td>
      <td>12903183</td>
      <td>12909538</td>
      <td>chrY</td>
      <td>12903998</td>
      <td>12920478</td>
      <td></td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Save the file if needed.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Brd4_bindings_bulk.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="SP1_bulk.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tutorial: SP1 bindings in Cre-driver mouse lines.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="K562HCT116_SP1.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorial: K562 HCT116 SP1 single-cell calling cards data.</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js?v=0b546569"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../_static/scripts/furo.js?v=5fa4622c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>