<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Tutorial: K562 HCT116 SP1 single-cell calling cards data." />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pycallingcards.readthedocs.io/en/latest/tutorials/notebooks/K562HCT116_SP1.html" />
<meta property="og:site_name" content="Pycallingcards" />
<meta property="og:description" content="In this tutorial, we will analyze the binding of the transcription factor Sp1 in K562 and HCT116 cell lines. These data are generated using single-cell calling cards (scCCs). The data is from Moudg..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="https://pycallingcards.readthedocs.io/en/latest/index.html/_images/social_previews/summary_tutorials_notebooks_K562HCT116_SP1_6a24b26d.png" />
<meta property="og:image:alt" content="In this tutorial, we will analyze the binding of the transcription factor Sp1 in K562 and HCT116 cell lines. These data are generated using single-cell calli..." />
<meta name="description" content="In this tutorial, we will analyze the binding of the transcription factor Sp1 in K562 and HCT116 cell lines. These data are generated using single-cell calling cards (scCCs). The data is from Moudg..." />
<meta name="twitter:card" content="summary_large_image" />
<link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data." href="Brd4_bulk.html" /><link rel="prev" title="Tutorial: Analyzing single cell calling cards collected from the mouse cortex" href="Mouse_cortex_Example.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.08.06 -->
        <title>Tutorial: K562 HCT116 SP1 single-cell calling cards data. - pycallingcards-tools</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pycallingcards-tools</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Mouse_cortex_Example.html">Tutorial: Analyzing single cell calling cards collected from the mouse cortex</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Tutorial: K562 HCT116 SP1 single-cell calling cards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="Brd4_bulk.html">Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="SP1_bulk.html">Tutorial: SP1 bindings in Cre-driver mouse lines.</a></li>
<li class="toctree-l2"><a class="reference internal" href="yeast.html">Tutorial: Yeast TYE7 bulk calling cards data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Datasets</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mousecortex_data.html">pycallingcards.datasets.mousecortex_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_K562HCT116_data.html">pycallingcards.datasets.SP1_K562HCT116_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mouse_brd4_data.html">pycallingcards.datasets.mouse_brd4_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_Cre_data.html">pycallingcards.datasets.SP1_Cre_data</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/reading.html">Reading</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Reading</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_qbed.html">pycallingcards.reading.read_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_rna.html">pycallingcards.reading.read_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_barbodes.html">pycallingcards.reading.read_barbodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.combine_qbed.html">pycallingcards.reading.combine_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_qbed.html">pycallingcards.reading.save_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_bed.html">pycallingcards.reading.save_bed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/preprocessing.html">Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Preprocessing</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.clean_qbed.html">pycallingcards.preprocessing.clean_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.call_peaks.html">pycallingcards.preprocessing.call_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.combine_peaks.html">pycallingcards.preprocessing.combine_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.separate_peaks.html">pycallingcards.preprocessing.separate_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.annotation.html">pycallingcards.preprocessing.annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.make_Anndata.html">pycallingcards.preprocessing.make_Anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.filter_peaks.html">pycallingcards.preprocessing.filter_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.adata_insertions.html">pycallingcards.preprocessing.adata_insertions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.down_sample.html">pycallingcards.preprocessing.down_sample</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/tools.html">Tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of Tools</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.call_motif.html">pycallingcards.tools.call_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.compare_motif.html">pycallingcards.tools.compare_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups.html">pycallingcards.tools.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups_mu.html">pycallingcards.tools.rank_peak_groups_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups_df.html">pycallingcards.tools.rank_peak_groups_df</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bygroup.html">pycallingcards.tools.diff2group_bygroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bysample.html">pycallingcards.tools.diff2group_bysample</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.DE_pvalue.html">pycallingcards.tools.DE_pvalue</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_bulk.html">pycallingcards.tools.pair_peak_gene_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc.html">pycallingcards.tools.pair_peak_gene_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc_mu.html">pycallingcards.tools.pair_peak_gene_sc_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.liftover.html">pycallingcards.tools.liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.find_location.html">pycallingcards.tools.find_location</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.result_mapping.html">pycallingcards.tools.result_mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS.html">pycallingcards.tools.GWAS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS_adata_bulk.html">pycallingcards.tools.GWAS_adata_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS_adata_sc.html">pycallingcards.tools.GWAS_adata_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.footprint.html">pycallingcards.tools.footprint</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.find_fastq.html">pycallingcards.tools.find_fastq</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of Plotting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area.html">pycallingcards.plotting.draw_area</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area_mu.html">pycallingcards.plotting.draw_area_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.whole_peaks.html">pycallingcards.plotting.whole_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.calculate_signal.html">pycallingcards.plotting.calculate_signal</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_plot.html">pycallingcards.plotting.signal_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_heatmap.html">pycallingcards.plotting.signal_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.rank_peak_groups.html">pycallingcards.plotting.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.plot_matched.html">pycallingcards.plotting.plot_matched</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.volcano_plot.html">pycallingcards.plotting.volcano_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.heatmap.html">pycallingcards.plotting.heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_bulk.html">pycallingcards.plotting.dotplot_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc.html">pycallingcards.plotting.dotplot_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc_mu.html">pycallingcards.plotting.dotplot_sc_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.WashU_browser_url.html">pycallingcards.plotting.WashU_browser_url</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.GWAS_adata_sc.html">pycallingcards.plotting.GWAS_adata_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.GWAS_adata_bulk.html">pycallingcards.plotting.GWAS_adata_bulk</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../../_sources/tutorials/notebooks/K562HCT116_SP1.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="Tutorial:-K562-HCT116-SP1-single-cell-calling-cards-data.">
<h1>Tutorial: K562 HCT116 SP1 single-cell calling cards data.<a class="headerlink" href="#Tutorial:-K562-HCT116-SP1-single-cell-calling-cards-data." title="Permalink to this heading">¶</a></h1>
<p>In this tutorial, we will analyze the binding of the transcription factor Sp1 in K562 and HCT116 cell lines. These data are generated using single-cell calling cards (scCCs). The data is from <a class="reference external" href="https://doi.org/10.1016/j.cell.2020.06.037">Moudgil et al., Cell. (2020)</a> and can be downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE148448">GEO</a>.</p>
<p>We will cover how to call TF peaks using a background file, annotate these peaks, compare them with Chip-seq reference data, and perform a differential peak analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pycallingcards</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">150</span>
</pre></div>
</div>
</div>
<p>We start by reading the qbed datafile. In this file, each row represents a Sp1-directed insertion and columns indicate the chromosome, start point and end point, reads number, the direction and cell barcode of each insertion. For example, the first row tells us the first insertion is on Chromosome 1 in a TTAA site located at genomic coordinates 30116 to 30120. There are 12 reads supporting this insertion, which maps to the negative strand of the genome. The cell barcode is CCCAATCCATCGGTTA-1.
Note that the cell barcode allows us to connect this insertino to scRNA-seq data collected from the same cell.</p>
<p>Use <code class="docutils literal notranslate"><span class="pre">cc.rd.read_qbed(filename)</span></code> to read your own ccf data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read in experiement data</span>
<span class="n">HCT116_SP1</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">SP1_K562HCT116_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;HCT116_SP1_qbed&quot;</span><span class="p">)</span>
<span class="n">HCT116_SP1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>30116</td>
      <td>30120</td>
      <td>5</td>
      <td>-</td>
      <td>CCCAATCCATCGGTTA-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>34568</td>
      <td>34572</td>
      <td>3</td>
      <td>-</td>
      <td>CCTTCGAAGGGCTTCC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>36736</td>
      <td>36740</td>
      <td>29</td>
      <td>+</td>
      <td>ACGAGCCGTATAGGTA-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>42447</td>
      <td>42451</td>
      <td>3</td>
      <td>-</td>
      <td>CTCTACGTCGGAGCAA-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>89697</td>
      <td>89701</td>
      <td>119</td>
      <td>-</td>
      <td>AGCTCTCGTTTGTTTC-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77205</th>
      <td>chrY</td>
      <td>25518788</td>
      <td>25518792</td>
      <td>2</td>
      <td>+</td>
      <td>TGGGCGTTCGAACGGA-1</td>
    </tr>
    <tr>
      <th>77206</th>
      <td>chrY</td>
      <td>56987633</td>
      <td>56987637</td>
      <td>13</td>
      <td>+</td>
      <td>CAGTCCTAGGCACATG-1</td>
    </tr>
    <tr>
      <th>77207</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>17</td>
      <td>+</td>
      <td>CGGAGCTCATCGACGC-1</td>
    </tr>
    <tr>
      <th>77208</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>7</td>
      <td>+</td>
      <td>GTAACGTAGTTACGGG-1</td>
    </tr>
    <tr>
      <th>77209</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>9</td>
      <td>+</td>
      <td>TCAGCAAGTTGAACTC-1</td>
    </tr>
  </tbody>
</table>
<p>77210 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read in backgound data</span>
<span class="n">HCT116_brd4</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">SP1_K562HCT116_data</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s2">&quot;HCT116_brd4_qbed&quot;</span><span class="p">)</span>
<span class="n">HCT116_brd4</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>89697</td>
      <td>89701</td>
      <td>14</td>
      <td>+</td>
      <td>TCTGAGACAATGGTCT-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>89697</td>
      <td>89701</td>
      <td>8</td>
      <td>+</td>
      <td>CAGCGACCAAATACAG-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>203932</td>
      <td>203936</td>
      <td>99</td>
      <td>+</td>
      <td>TTCTCCTTCTACTTAC-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>204063</td>
      <td>204067</td>
      <td>5</td>
      <td>-</td>
      <td>TGTTCCGGTGTAAGTA-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>204063</td>
      <td>204067</td>
      <td>7</td>
      <td>-</td>
      <td>CAAGATCTCGACCAGC-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>37769</th>
      <td>chrY</td>
      <td>18037315</td>
      <td>18037319</td>
      <td>9</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37770</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>168</td>
      <td>+</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37771</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>508</td>
      <td>+</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
    <tr>
      <th>37772</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>13</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37773</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>32</td>
      <td>-</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
  </tbody>
</table>
<p>37774 rows × 6 columns</p>
</div></div>
</div>
<p>We first need to call peaks in order to find candidate SP1 binding sites. There are three different methods (<em>CCcaller, MACCs, Blockify</em>) available in Pycallingcards. Here, we will use <strong>MACCs</strong> to call peaks. The appropriate reference genome for these data is <em>human</em>(‘hg38’). The <em>window_size</em> parameter is the most important parameter for <em>MACCs</em>, it is highly related to the length of a peak. A value of 1000-2000 is recommended for most sequence-specific TFs. <em>step_size</em> is another important
parameter and it controls whether two nearby clusters of insertions are called as one peak or split into two peaks. 500-800 is good for <em>step_size</em>. <em>pvalue_cutoffTTAA</em> is the pvalue cutoff for TTAA data and <em>pvalue_cutoffbg</em> is the pvalue cutoff for the background ccf data. Normally, the setting for <em>pvalue_cutoffbg</em> is considerably higher than <em>pvalue_cutoffTTAA</em>. <em>pvalue_cutoffbg</em> is recommended to be 0.1 and <em>pvalue_cutoffTTAA</em> is recommended from 0.001 to 0.05. <em>pseudocounts</em> is advised to
be 0.1-1.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_data_HCT116</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">call_peaks</span><span class="p">(</span><span class="n">HCT116_SP1</span><span class="p">,</span> <span class="n">HCT116_brd4</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;MACCs&quot;</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span>  <span class="n">window_size</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">step_size</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                  <span class="n">pvalue_cutoffTTAA</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span> <span class="n">pvalue_cutoffbg</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">lam_win_size</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="s2">&quot;peak_HCT116.bed&quot;</span><span class="p">)</span>
<span class="n">peak_data_HCT116</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the MACCs method with background, [expdata, background, reference, pvalue_cutoffbg, pvalue_cutoffTTAA, lam_win_size, window_size, step_size, extend, pseudocounts, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 25/25 [00:43&lt;00:00,  1.75s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Center</th>
      <th>Experiment Insertions</th>
      <th>Background insertions</th>
      <th>Reference Insertions</th>
      <th>pvalue Reference</th>
      <th>pvalue Background</th>
      <th>Fraction Experiment</th>
      <th>TPH Experiment</th>
      <th>Fraction background</th>
      <th>TPH background</th>
      <th>TPH background subtracted</th>
      <th>pvalue_adj Reference</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>906689</td>
      <td>907160</td>
      <td>906957.0</td>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>3.099334e-09</td>
      <td>1.546531e-04</td>
      <td>0.000065</td>
      <td>6475.845098</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>6475.845098</td>
      <td>3.371990e-06</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>999921</td>
      <td>1000324</td>
      <td>1000121.0</td>
      <td>20</td>
      <td>0</td>
      <td>1</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>0.000259</td>
      <td>25903.380391</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>25903.380391</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1156947</td>
      <td>1157863</td>
      <td>1157660.0</td>
      <td>11</td>
      <td>0</td>
      <td>2</td>
      <td>0.000000e+00</td>
      <td>1.274899e-09</td>
      <td>0.000142</td>
      <td>14246.859215</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>14246.859215</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1692740</td>
      <td>1693542</td>
      <td>1693339.0</td>
      <td>6</td>
      <td>0</td>
      <td>3</td>
      <td>5.135270e-11</td>
      <td>1.546531e-04</td>
      <td>0.000078</td>
      <td>7771.014117</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>7771.014117</td>
      <td>7.604315e-08</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1744492</td>
      <td>1746808</td>
      <td>1746605.0</td>
      <td>11</td>
      <td>0</td>
      <td>7</td>
      <td>0.000000e+00</td>
      <td>1.274899e-09</td>
      <td>0.000142</td>
      <td>14246.859215</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>14246.859215</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3003</th>
      <td>chrY</td>
      <td>1245503</td>
      <td>1247177</td>
      <td>1245703.0</td>
      <td>5</td>
      <td>0</td>
      <td>7</td>
      <td>1.652897e-09</td>
      <td>4.678840e-03</td>
      <td>0.000065</td>
      <td>6475.845098</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>6475.845098</td>
      <td>2.056405e-06</td>
    </tr>
    <tr>
      <th>3004</th>
      <td>chrY</td>
      <td>1280372</td>
      <td>1281989</td>
      <td>1281786.0</td>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>1.426995e-09</td>
      <td>4.678840e-03</td>
      <td>0.000065</td>
      <td>6475.845098</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>6475.845098</td>
      <td>1.828173e-06</td>
    </tr>
    <tr>
      <th>3005</th>
      <td>chrY</td>
      <td>1586317</td>
      <td>1587733</td>
      <td>1587530.0</td>
      <td>7</td>
      <td>0</td>
      <td>8</td>
      <td>3.370637e-13</td>
      <td>1.546531e-04</td>
      <td>0.000091</td>
      <td>9066.183137</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>9066.183137</td>
      <td>6.766326e-10</td>
    </tr>
    <tr>
      <th>3006</th>
      <td>chrY</td>
      <td>2391936</td>
      <td>2392440</td>
      <td>2392237.0</td>
      <td>6</td>
      <td>1</td>
      <td>2</td>
      <td>1.985434e-11</td>
      <td>9.958372e-02</td>
      <td>0.000078</td>
      <td>7771.014117</td>
      <td>0.000026</td>
      <td>2647.323556</td>
      <td>5123.690561</td>
      <td>3.329699e-08</td>
    </tr>
    <tr>
      <th>3007</th>
      <td>chrY</td>
      <td>2608793</td>
      <td>2610054</td>
      <td>2608993.0</td>
      <td>8</td>
      <td>0</td>
      <td>2</td>
      <td>2.775558e-15</td>
      <td>1.546531e-04</td>
      <td>0.000104</td>
      <td>10361.352156</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>10361.352156</td>
      <td>6.730626e-12</td>
    </tr>
  </tbody>
</table>
<p>3008 rows × 15 columns</p>
</div></div>
</div>
<p>In order to tune parameters for peak calling, we advise looking at the data and evaluating the validity of the called peaks. The default settings are recommended, but for some TFs, adjacent peaks may be merged that should not be, or, alternatively, peaks that should be joined may be called separately.</p>
<p>Below, we plot the scCC data in HCT116 cells for a region in chromosome 1. The top track displays the locations of Sp1-directed (red) and background (grey) transpositions and their read counts. Each dot represents an insertion and the height is log(reads+1). The middle track plots the insertion density. The third track represents the reference genes and peaks. Finally, the last track represents peak calls. Below you can see that regions with high densities of insertions are accurately called as
Sp1 binding sites.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span><span class="mi">999921</span><span class="p">,</span><span class="mi">1000324</span><span class="p">,</span><span class="mi">20000</span><span class="p">,</span><span class="n">peak_data_HCT116</span><span class="p">,</span> <span class="n">HCT116_SP1</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">HCT116_brd4</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">example_length</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_9_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_9_0.png" />
</div>
</div>
<p>We can also visualize our data directly in the <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a>. This can be useful for overlaying your data with other published datasets. Notice that this link only valid for 24hrs, so please rerun it if you want to use it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">qbed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;SP1&quot;</span><span class="p">:</span><span class="n">HCT116_SP1</span><span class="p">,</span> <span class="s2">&quot;Brd4&quot;</span><span class="p">:</span> <span class="n">HCT116_brd4</span><span class="p">}</span>
<span class="n">bed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;peak&quot;</span><span class="p">:</span><span class="n">peak_data_HCT116</span><span class="p">}</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">WashU_browser_url</span><span class="p">(</span><span class="n">qbed</span> <span class="o">=</span> <span class="n">qbed</span><span class="p">,</span><span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span><span class="p">,</span><span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;hg38&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=hg38&amp;hub=https://companion.epigenomegateway.org//task/cb339e5b2d8c774345b7cf372cf4f97c/output//datahub.json
</pre></div></div>
</div>
<p>Pycallingcards can be used to visual peak locations acorss the genome to see that the distribution of peaks is unbiased and that all chromosomes are represented.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">whole_peaks</span><span class="p">(</span><span class="n">peak_data_HCT116</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">70</span><span class="p">),</span> <span class="n">height_scale</span> <span class="o">=</span> <span class="mf">1.7</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_13_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_13_0.png" />
</div>
</div>
<p>We can then analyze the scCC Sp1 peaks to see if there is an enrichment of ChIP-seq signal at these locations using a reference Chip-seq dataset of SP1 binding in HCT116 from <a class="reference external" href="https://www.encodeproject.org/experiments/ENCSR000BSF/">ENCSR000BSF</a> (use the bigWig file <a class="reference external" href="https://www.encodeproject.org/files/ENCFF587ZMX/">ENCFF587ZMX</a> generated by it).</p>
<p>Download the data, if needed, with:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!wget https://www.encodeproject.org/files/ENCFF587ZMX/@@download/ENCFF587ZMX.bigWig
</pre></div>
</div>
<p>We first calculate the signal of the Chip-seq signal around the peak.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mtx_HCT116</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">calculate_signal</span><span class="p">(</span><span class="n">peak_data_HCT116</span><span class="p">,</span>
                                    <span class="n">chipseq_signal</span> <span class="o">=</span> <span class="s2">&quot;ENCFF587ZMX.bigWig&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3008/3008 [00:14&lt;00:00, 200.91it/s]
</pre></div></div>
</div>
<p>Visualize it by the plotting the signal values.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">signal_plot</span><span class="p">(</span><span class="n">mtx_HCT116</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_18_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_18_0.png" />
</div>
</div>
<p>Visualized by the plotting the signal heatmap plot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">signal_heatmap</span><span class="p">(</span><span class="n">mtx_HCT116</span><span class="p">,</span><span class="n">pad</span> <span class="o">=</span> <span class="mf">0.035</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_20_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_20_0.png" />
</div>
</div>
<p>We can now use HOMER to call motifs. We hope to find the canonical Sp1 motif enriched under the call peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">call_motif</span><span class="p">(</span><span class="s2">&quot;peak_HCT116.bed&quot;</span><span class="p">,</span><span class="n">reference</span> <span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">,</span><span class="n">save_homer</span> <span class="o">=</span> <span class="s2">&quot;Homer/peak_HCT116&quot;</span><span class="p">,</span>
                 <span class="n">homer_path</span> <span class="o">=</span> <span class="s2">&quot;/home/juanru/miniconda3/bin/&quot;</span><span class="p">,</span> <span class="n">num_cores</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Peak data peak_HCT116.bed is used here.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

        Position file = peak_HCT116.bed
        Genome = hg38
        Output Directory = Homer/peak_HCT116
        Fragment size set to 1000
        Using 12 CPUs
        Will not run homer for de novo motifs
        Found mset for &#34;human&#34;, will check against vertebrates motifs
        Peak/BED file conversion summary:
                BED/Header formatted lines: 3008
                peakfile formatted lines: 0

        Peak File Statistics:
                Total Peaks: 3008
                Redundant Peak IDs: 0
                Peaks lacking information: 0 (need at least 5 columns per peak)
                Peaks with misformatted coordinates: 0 (should be integer)
                Peaks with misformatted strand: 0 (should be either +/- or 0/1)

        Peak file looks good!

        Background files for 1000 bp fragments found.
        Custom genome sequence directory: /home/juanru/miniconda3/share/homer/.//data/genomes/hg38//

        Extracting sequences from file: /home/juanru/miniconda3/share/homer/.//data/genomes/hg38///genome.fa
        Looking for peak sequences in a single file (/home/juanru/miniconda3/share/homer/.//data/genomes/hg38///genome.fa)
        Extracting 334 sequences from chr1
        Extracting 147 sequences from chr10
        Extracting 132 sequences from chr11
        Extracting 179 sequences from chr12
        Extracting 56 sequences from chr13
        Extracting 85 sequences from chr14
        Extracting 83 sequences from chr15
        Extracting 127 sequences from chr16
        Extracting 205 sequences from chr17
        Extracting 54 sequences from chr18
        Extracting 249 sequences from chr19
        Extracting 194 sequences from chr2
        Extracting 71 sequences from chr20
        Extracting 37 sequences from chr21
        Extracting 63 sequences from chr22
        Extracting 143 sequences from chr3
        Extracting 146 sequences from chr4
        Extracting 96 sequences from chr5
        Extracting 159 sequences from chr6
        Extracting 143 sequences from chr7
        Extracting 117 sequences from chr8
        Extracting 136 sequences from chr9
        Extracting 47 sequences from chrX
        Extracting 5 sequences from chrY

        Not removing redundant sequences


        Sequences processed:
                Auto detected maximum sequence length of 1001 bp
                3008 total

        Frequency Bins: 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.6 0.7 0.8
        Freq    Bin     Count
        0.3     2       1
        0.35    3       46
        0.4     4       183
        0.45    5       336
        0.5     6       476
        0.6     7       1116
        0.7     8       706
        0.8     9       142
        10      10      2

        Total sequences set to 50000

        Choosing background that matches in CpG/GC content...
        Bin     # Targets       # Background    Background Weight
        2       1       16      0.976
        3       46      719     0.999
        4       183     2859    1.000
        5       336     5249    1.000
        6       476     7436    1.000
        7       1116    17435   1.000
        8       706     11029   1.000
        9       142     2218    1.000
        10      2       31      1.008
        Assembling sequence file...
        Normalizing lower order oligos using homer2

        Reading input files...
        50000 total sequences read
        Autonormalization: 1-mers (4 total)
                A       22.98%  23.24%  0.989
                C       27.02%  26.76%  1.010
                G       27.02%  26.76%  1.010
                T       22.98%  23.24%  0.989
        Autonormalization: 2-mers (16 total)
                AA      6.71%   6.20%   1.082
                CA      6.45%   7.36%   0.876
                GA      5.97%   6.11%   0.977
                TA      3.85%   3.56%   1.081
                AC      4.73%   4.90%   0.965
                CC      8.95%   8.75%   1.023
                GC      7.37%   7.01%   1.051
                TC      5.97%   6.11%   0.977
                AG      7.19%   7.61%   0.944
                CG      4.43%   3.06%   1.451
                GG      8.95%   8.75%   1.023
                TG      6.45%   7.36%   0.876
                AT      4.36%   4.52%   0.964
                CT      7.19%   7.61%   0.944
                GT      4.73%   4.90%   0.965
                TT      6.71%   6.20%   1.082
        Autonormalization: 3-mers (64 total)
        Normalization weights can be found in file: Homer/peak_HCT116/seq.autonorm.tsv
        Converging on autonormalization solution:
        ...............................................................................
        Final normalization:    Autonormalization: 1-mers (4 total)
                A       22.98%  22.93%  1.002
                C       27.02%  27.07%  0.998
                G       27.02%  27.07%  0.998
                T       22.98%  22.93%  1.002
        Autonormalization: 2-mers (16 total)
                AA      6.71%   6.50%   1.032
                CA      6.45%   6.69%   0.965
                GA      5.97%   5.97%   1.000
                TA      3.85%   3.78%   1.020
                AC      4.73%   4.71%   1.003
                CC      8.95%   8.89%   1.007
                GC      7.37%   7.50%   0.983
                TC      5.97%   5.97%   1.000
                AG      7.19%   7.21%   0.997
                CG      4.43%   4.29%   1.035
                GG      8.95%   8.89%   1.007
                TG      6.45%   6.69%   0.965
                AT      4.36%   4.51%   0.966
                CT      7.19%   7.21%   0.997
                GT      4.73%   4.71%   1.003
                TT      6.71%   6.50%   1.032
        Autonormalization: 3-mers (64 total)
        Finished preparing sequence/group files

        ----------------------------------------------------------
        Known motif enrichment

        Reading input files...
        50000 total sequences read
        440 motifs loaded
        Cache length = 11180
        Using binomial scoring
        Checking enrichment of 440 motif(s)
        |0%                                    50%                                  100%|
        =================================================================================
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finished!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

        Preparing HTML output with sequence logos...
                1 of 440 (1e-105) Sp5(Zf)/mES-Sp5.Flag-ChIP-Seq(GSE72989)/Homer
                2 of 440 (1e-97) KLF3(Zf)/MEF-Klf3-ChIP-Seq(GSE44748)/Homer
                3 of 440 (1e-92) Sp1(Zf)/Promoter/Homer
                4 of 440 (1e-87) KLF1(Zf)/HUDEP2-KLF1-CutnRun(GSE136251)/Homer
                5 of 440 (1e-78) Sp2(Zf)/HEK293-Sp2.eGFP-ChIP-Seq(Encode)/Homer
                6 of 440 (1e-77) Klf9(Zf)/GBM-Klf9-ChIP-Seq(GSE62211)/Homer
                7 of 440 (1e-69) KLF6(Zf)/PDAC-KLF6-ChIP-Seq(GSE64557)/Homer
                8 of 440 (1e-68) KLF5(Zf)/LoVo-KLF5-ChIP-Seq(GSE49402)/Homer
                9 of 440 (1e-65) KLF14(Zf)/HEK293-KLF14.GFP-ChIP-Seq(GSE58341)/Homer
                10 of 440 (1e-62) Maz(Zf)/HepG2-Maz-ChIP-Seq(GSE31477)/Homer
                11 of 440 (1e-58) Klf4(Zf)/mES-Klf4-ChIP-Seq(GSE11431)/Homer
                12 of 440 (1e-51) Jun-AP1(bZIP)/K562-cJun-ChIP-Seq(GSE31477)/Homer
                13 of 440 (1e-50) Fosl2(bZIP)/3T3L1-Fosl2-ChIP-Seq(GSE56872)/Homer
                14 of 440 (1e-43) TATA-Box(TBP)/Promoter/Homer
                15 of 440 (1e-42) Fra2(bZIP)/Striatum-Fra2-ChIP-Seq(GSE43429)/Homer
                16 of 440 (1e-41) Fra1(bZIP)/BT549-Fra1-ChIP-Seq(GSE46166)/Homer
                17 of 440 (1e-41) Fos(bZIP)/TSC-Fos-ChIP-Seq(GSE110950)/Homer
                18 of 440 (1e-41) JunB(bZIP)/DendriticCells-Junb-ChIP-Seq(GSE36099)/Homer
                19 of 440 (1e-37) Atf3(bZIP)/GBM-ATF3-ChIP-Seq(GSE33912)/Homer
                20 of 440 (1e-35) KLF10(Zf)/HEK293-KLF10.GFP-ChIP-Seq(GSE58341)/Homer
                21 of 440 (1e-30) BATF(bZIP)/Th17-BATF-ChIP-Seq(GSE39756)/Homer
                22 of 440 (1e-28) AP-1(bZIP)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer
                23 of 440 (1e-25) Bach2(bZIP)/OCILy7-Bach2-ChIP-Seq(GSE44420)/Homer
                24 of 440 (1e-22) Pitx1(Homeobox)/Chicken-Pitx1-ChIP-Seq(GSE38910)/Homer
                25 of 440 (1e-20) Elk4(ETS)/Hela-Elk4-ChIP-Seq(GSE31477)/Homer
                26 of 440 (1e-19) E2F4(E2F)/K562-E2F4-ChIP-Seq(GSE31477)/Homer
                27 of 440 (1e-19) Hoxd11(Homeobox)/ChickenMSG-Hoxd11.Flag-ChIP-Seq(GSE86088)/Homer
                28 of 440 (1e-18) Nkx6.1(Homeobox)/Islet-Nkx6.1-ChIP-Seq(GSE40975)/Homer
                29 of 440 (1e-18) Hoxa13(Homeobox)/ChickenMSG-Hoxa13.Flag-ChIP-Seq(GSE86088)/Homer
                30 of 440 (1e-17) WT1(Zf)/Kidney-WT1-ChIP-Seq(GSE90016)/Homer
                31 of 440 (1e-16) Fli1(ETS)/CD8-FLI-ChIP-Seq(GSE20898)/Homer
                32 of 440 (1e-16) Hoxa11(Homeobox)/ChickenMSG-Hoxa11.Flag-ChIP-Seq(GSE86088)/Homer
                33 of 440 (1e-16) Isl1(Homeobox)/Neuron-Isl1-ChIP-Seq(GSE31456)/Homer
                34 of 440 (1e-15) Elk1(ETS)/Hela-Elk1-ChIP-Seq(GSE31477)/Homer
                35 of 440 (1e-15) Egr1(Zf)/K562-Egr1-ChIP-Seq(GSE32465)/Homer
                36 of 440 (1e-15) E2F6(E2F)/Hela-E2F6-ChIP-Seq(GSE31477)/Homer
                37 of 440 (1e-15) ETV4(ETS)/HepG2-ETV4-ChIP-Seq(ENCODE)/Homer
                38 of 440 (1e-15) EKLF(Zf)/Erythrocyte-Klf1-ChIP-Seq(GSE20478)/Homer
                39 of 440 (1e-14) Egr2(Zf)/Thymocytes-Egr2-ChIP-Seq(GSE34254)/Homer
                40 of 440 (1e-13) ZNF467(Zf)/HEK293-ZNF467.GFP-ChIP-Seq(GSE58341)/Homer
                41 of 440 (1e-12) E2F3(E2F)/MEF-E2F3-ChIP-Seq(GSE71376)/Homer
                42 of 440 (1e-11) ELF1(ETS)/Jurkat-ELF1-ChIP-Seq(SRA014231)/Homer
                43 of 440 (1e-11) Nrf2(bZIP)/Lymphoblast-Nrf2-ChIP-Seq(GSE37589)/Homer
                44 of 440 (1e-11) Unknown(Homeobox)/Limb-p300-ChIP-Seq/Homer
                45 of 440 (1e-10) Lhx1(Homeobox)/EmbryoCarcinoma-Lhx1-ChIP-Seq(GSE70957)/Homer
                46 of 440 (1e-10) E2F1(E2F)/Hela-E2F1-ChIP-Seq(GSE22478)/Homer
                47 of 440 (1e-10) En1(Homeobox)/SUM149-EN1-ChIP-Seq(GSE120957)/Homer
                48 of 440 (1e-9) Lhx3(Homeobox)/Neuron-Lhx3-ChIP-Seq(GSE31456)/Homer
                49 of 440 (1e-9) CRX(Homeobox)/Retina-Crx-ChIP-Seq(GSE20012)/Homer
                50 of 440 (1e-8) Zfp281(Zf)/ES-Zfp281-ChIP-Seq(GSE81042)/Homer
                51 of 440 (1e-8) GABPA(ETS)/Jurkat-GABPa-ChIP-Seq(GSE17954)/Homer
                52 of 440 (1e-8) NF-E2(bZIP)/K562-NFE2-ChIP-Seq(GSE31477)/Homer
                53 of 440 (1e-8) Hoxd13(Homeobox)/ChickenMSG-Hoxd13.Flag-ChIP-Seq(GSE86088)/Homer
                54 of 440 (1e-8) HIF-1b(HLH)/T47D-HIF1b-ChIP-Seq(GSE59937)/Homer
                55 of 440 (1e-8) ETV1(ETS)/GIST48-ETV1-ChIP-Seq(GSE22441)/Homer
                56 of 440 (1e-7) LHX9(Homeobox)/Hct116-LHX9.V5-ChIP-Seq(GSE116822)/Homer
                57 of 440 (1e-7) DLX2(Homeobox)/BasalGanglia-Dlx2-ChIP-seq(GSE124936)/Homer
                58 of 440 (1e-7) Bach1(bZIP)/K562-Bach1-ChIP-Seq(GSE31477)/Homer
                59 of 440 (1e-7) DLX1(Homeobox)/BasalGanglia-Dlx1-ChIP-seq(GSE124936)/Homer
                60 of 440 (1e-7) EHF(ETS)/LoVo-EHF-ChIP-Seq(GSE49402)/Homer
                61 of 440 (1e-7) BMYB(HTH)/Hela-BMYB-ChIP-Seq(GSE27030)/Homer
                62 of 440 (1e-7) NFE2L2(bZIP)/HepG2-NFE2L2-ChIP-Seq(Encode)/Homer
                63 of 440 (1e-7) Stat3(Stat)/mES-Stat3-ChIP-Seq(GSE11431)/Homer
                64 of 440 (1e-6) Hoxa9(Homeobox)/ChickenMSG-Hoxa9.Flag-ChIP-Seq(GSE86088)/Homer
                65 of 440 (1e-6) MYB(HTH)/ERMYB-Myb-ChIPSeq(GSE22095)/Homer
                66 of 440 (1e-6) Znf263(Zf)/K562-Znf263-ChIP-Seq(GSE31477)/Homer
                67 of 440 (1e-6) HNF1b(Homeobox)/PDAC-HNF1B-ChIP-Seq(GSE64557)/Homer
                68 of 440 (1e-6) CHR(?)/Hela-CellCycle-Expression/Homer
                69 of 440 (1e-6) CTCF(Zf)/CD4+-CTCF-ChIP-Seq(Barski_et_al.)/Homer
                70 of 440 (1e-6) ETS(ETS)/Promoter/Homer
                71 of 440 (1e-6) ETS1(ETS)/Jurkat-ETS1-ChIP-Seq(GSE17954)/Homer
                72 of 440 (1e-6) ZNF652/HepG2-ZNF652.Flag-ChIP-Seq(Encode)/Homer
                73 of 440 (1e-5) Nanog(Homeobox)/mES-Nanog-ChIP-Seq(GSE11724)/Homer
                74 of 440 (1e-5) Lhx2(Homeobox)/HFSC-Lhx2-ChIP-Seq(GSE48068)/Homer
                75 of 440 (1e-5) Elf4(ETS)/BMDM-Elf4-ChIP-Seq(GSE88699)/Homer
                76 of 440 (1e-5) NFY(CCAAT)/Promoter/Homer
                77 of 440 (1e-5) ELF5(ETS)/T47D-ELF5-ChIP-Seq(GSE30407)/Homer
                78 of 440 (1e-5) AP-2gamma(AP2)/MCF7-TFAP2C-ChIP-Seq(GSE21234)/Homer
                79 of 440 (1e-5) ERG(ETS)/VCaP-ERG-ChIP-Seq(GSE14097)/Homer
                80 of 440 (1e-5) EWS:ERG-fusion(ETS)/CADO_ES1-EWS:ERG-ChIP-Seq(SRA014231)/Homer
                81 of 440 (1e-5) AMYB(HTH)/Testes-AMYB-ChIP-Seq(GSE44588)/Homer
                82 of 440 (1e-5) ELF3(ETS)/PDAC-ELF3-ChIP-Seq(GSE64557)/Homer
                83 of 440 (1e-5) HOXB13(Homeobox)/ProstateTumor-HOXB13-ChIP-Seq(GSE56288)/Homer
                84 of 440 (1e-5) STAT4(Stat)/CD4-Stat4-ChIP-Seq(GSE22104)/Homer
                85 of 440 (1e-5) MafK(bZIP)/C2C12-MafK-ChIP-Seq(GSE36030)/Homer
                86 of 440 (1e-4) SPDEF(ETS)/VCaP-SPDEF-ChIP-Seq(SRA014231)/Homer
                87 of 440 (1e-4) Ets1-distal(ETS)/CD4+-PolII-ChIP-Seq(Barski_et_al.)/Homer
                88 of 440 (1e-4) DLX5(Homeobox)/BasalGanglia-Dlx5-ChIP-seq(GSE124936)/Homer
                89 of 440 (1e-4) Dlx3(Homeobox)/Kerainocytes-Dlx3-ChIP-Seq(GSE89884)/Homer
                90 of 440 (1e-4) PRDM15(Zf)/ESC-Prdm15-ChIP-Seq(GSE73694)/Homer
                91 of 440 (1e-4) Stat3+il21(Stat)/CD4-Stat3-ChIP-Seq(GSE19198)/Homer
                92 of 440 (1e-4) Etv2(ETS)/ES-ER71-ChIP-Seq(GSE59402)/Homer
                93 of 440 (1e-4) HIF2a(bHLH)/785_O-HIF2a-ChIP-Seq(GSE34871)/Homer
                94 of 440 (1e-4) TEAD1(TEAD)/HepG2-TEAD1-ChIP-Seq(Encode)/Homer
                95 of 440 (1e-3) ZNF189(Zf)/HEK293-ZNF189.GFP-ChIP-Seq(GSE58341)/Homer
                96 of 440 (1e-3) STAT5(Stat)/mCD4+-Stat5-ChIP-Seq(GSE12346)/Homer
                97 of 440 (1e-3) AP-2alpha(AP2)/Hela-AP2alpha-ChIP-Seq(GSE31477)/Homer
                98 of 440 (1e-3) ZFX(Zf)/mES-Zfx-ChIP-Seq(GSE11431)/Homer
                99 of 440 (1e-3) NFkB-p65-Rel(RHD)/ThioMac-LPS-Expression(GSE23622)/Homer
                100 of 440 (1e-3) BORIS(Zf)/K562-CTCFL-ChIP-Seq(GSE32465)/Homer
                101 of 440 (1e-3) E2F7(E2F)/Hela-E2F7-ChIP-Seq(GSE32673)/Homer
                102 of 440 (1e-3) EWS:FLI1-fusion(ETS)/SK_N_MC-EWS:FLI1-ChIP-Seq(SRA014231)/Homer
                103 of 440 (1e-3) c-Myc(bHLH)/LNCAP-cMyc-ChIP-Seq(Unpublished)/Homer
                104 of 440 (1e-3) Otx2(Homeobox)/EpiLC-Otx2-ChIP-Seq(GSE56098)/Homer
                105 of 440 (1e-3) TEAD4(TEA)/Tropoblast-Tead4-ChIP-Seq(GSE37350)/Homer
                106 of 440 (1e-2) Npas4(bHLH)/Neuron-Npas4-ChIP-Seq(GSE127793)/Homer
                107 of 440 (1e-2) CArG(MADS)/PUER-Srf-ChIP-Seq(Sullivan_et_al.)/Homer
                108 of 440 (1e-2) ZNF711(Zf)/SHSY5Y-ZNF711-ChIP-Seq(GSE20673)/Homer
                109 of 440 (1e-2) Prop1(Homeobox)/GHFT1-PROP1.biotin-ChIP-Seq(GSE77302)/Homer
                110 of 440 (1e-2) GSC(Homeobox)/FrogEmbryos-GSC-ChIP-Seq(DRA000576)/Homer
                111 of 440 (1e-2) Hnf1(Homeobox)/Liver-Foxa2-Chip-Seq(GSE25694)/Homer
                112 of 440 (1e-2) TRPS1(Zf)/MCF7-TRPS1-ChIP-Seq(GSE107013)/Homer
                113 of 440 (1e-2) Tbr1(T-box)/Cortex-Tbr1-ChIP-Seq(GSE71384)/Homer
                114 of 440 (1e-2) NFAT(RHD)/Jurkat-NFATC1-ChIP-Seq(Jolma_et_al.)/Homer
                115 of 440 (1e-2) CLOCK(bHLH)/Liver-Clock-ChIP-Seq(GSE39860)/Homer
                116 of 440 (1e-2) STAT1(Stat)/HelaS3-STAT1-ChIP-Seq(GSE12782)/Homer
                117 of 440 (1e-2) Barx1(Homeobox)/Stomach-Barx1.3xFlag-ChIP-Seq(GSE69483)/Homer
                118 of 440 (1e-2) NFkB-p65(RHD)/GM12787-p65-ChIP-Seq(GSE19485)/Homer
                119 of 440 (1e-2) CEBP(bZIP)/ThioMac-CEBPb-ChIP-Seq(GSE21512)/Homer
                120 of 440 (1e-2) Arnt:Ahr(bHLH)/MCF7-Arnt-ChIP-Seq(Lo_et_al.)/Homer
                121 of 440 (1e-2) HINFP(Zf)/K562-HINFP.eGFP-ChIP-Seq(Encode)/Homer
                122 of 440 (1e-2) CRE(bZIP)/Promoter/Homer
                123 of 440 (1e-2) E-box(bHLH)/Promoter/Homer
                124 of 440 (1e-2) Hoxd12(Homeobox)/ChickenMSG-Hoxd12.Flag-ChIP-Seq(GSE86088)/Homer
                125 of 440 (1e-2) TEAD(TEA)/Fibroblast-PU.1-ChIP-Seq(Unpublished)/Homer
                126 of 440 (1e-2) ZSCAN22(Zf)/HEK293-ZSCAN22.GFP-ChIP-Seq(GSE58341)/Homer
                127 of 440 (1e-2) IRF:BATF(IRF:bZIP)/pDC-Irf8-ChIP-Seq(GSE66899)/Homer
                128 of 440 (1e-2) CDX4(Homeobox)/ZebrafishEmbryos-Cdx4.Myc-ChIP-Seq(GSE48254)/Homer
        Skipping...
        Job finished - if results look good, please send beer to ..

        Cleaning up tmp files...

</pre></div></div>
</div>
<p>In the motif analysis result, SP1 motif and many other family members rank top.</p>
<p><img alt="drawing" class="no-scaled-link" src="../../_images/peak_HCT116.png" style="width: 800px;" /></p>
<p>Do the exact same thing for K562 SP1 data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read experiment data</span>
<span class="n">K562_SP1</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">SP1_K562HCT116_data</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s2">&quot;K562_SP1_qbed&quot;</span><span class="p">)</span>
<span class="n">K562_SP1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>16529</td>
      <td>16533</td>
      <td>163</td>
      <td>-</td>
      <td>GCTCCTAAGTACGTTC-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>10</td>
      <td>+</td>
      <td>CTCACACCAGACGCTC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>155</td>
      <td>+</td>
      <td>TGGCCAGCACCCATTC-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>285</td>
      <td>+</td>
      <td>GTGGGTCCACGGCCAT-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>7</td>
      <td>+</td>
      <td>CGTCTACTCAACACGT-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>327460</th>
      <td>chrY</td>
      <td>57061562</td>
      <td>57061566</td>
      <td>6</td>
      <td>+</td>
      <td>CTCATTATCATCATTC-1</td>
    </tr>
    <tr>
      <th>327461</th>
      <td>chrY</td>
      <td>57061562</td>
      <td>57061566</td>
      <td>67</td>
      <td>+</td>
      <td>TGCGTGGCATTAGGCT-1</td>
    </tr>
    <tr>
      <th>327462</th>
      <td>chrY</td>
      <td>57145084</td>
      <td>57145088</td>
      <td>2</td>
      <td>-</td>
      <td>ACATACGTCGCGCCAA-1</td>
    </tr>
    <tr>
      <th>327463</th>
      <td>chrY</td>
      <td>57148630</td>
      <td>57148634</td>
      <td>2</td>
      <td>-</td>
      <td>TATGCCCGTACAGTTC-1</td>
    </tr>
    <tr>
      <th>327464</th>
      <td>chrY</td>
      <td>57183913</td>
      <td>57183917</td>
      <td>228</td>
      <td>-</td>
      <td>AAACCTGGTCCTGCTT-1</td>
    </tr>
  </tbody>
</table>
<p>327465 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read background data</span>
<span class="n">K562_brd4</span> <span class="o">=</span><span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">SP1_K562HCT116_data</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s2">&quot;K562_brd4_qbed&quot;</span><span class="p">)</span>
<span class="n">K562_brd4</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>30238</td>
      <td>30242</td>
      <td>3</td>
      <td>+</td>
      <td>TTTACTGCATAAAGGT-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>2</td>
      <td>-</td>
      <td>ATCACGAAGAGTAATC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>70</td>
      <td>+</td>
      <td>TTGAACGCAAATCCGT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>31101</td>
      <td>31105</td>
      <td>2</td>
      <td>+</td>
      <td>CCTCAGTCATCAGTAC-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>32116</td>
      <td>32120</td>
      <td>5</td>
      <td>+</td>
      <td>CTAGTGAAGACAAAGG-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>107380</th>
      <td>chrY</td>
      <td>57080210</td>
      <td>57080214</td>
      <td>9</td>
      <td>-</td>
      <td>AAGGAGCCAGTATAAG-1</td>
    </tr>
    <tr>
      <th>107381</th>
      <td>chrY</td>
      <td>57087785</td>
      <td>57087789</td>
      <td>24</td>
      <td>-</td>
      <td>CGAGCCAGTCTCTCTG-1</td>
    </tr>
    <tr>
      <th>107382</th>
      <td>chrY</td>
      <td>57144853</td>
      <td>57144857</td>
      <td>5</td>
      <td>+</td>
      <td>GAAGCAGTCCCATTTA-1</td>
    </tr>
    <tr>
      <th>107383</th>
      <td>chrY</td>
      <td>57183772</td>
      <td>57183776</td>
      <td>2</td>
      <td>-</td>
      <td>TCTTTCCTCTTGCCGT-1</td>
    </tr>
    <tr>
      <th>107384</th>
      <td>chrY</td>
      <td>57204853</td>
      <td>57204857</td>
      <td>369</td>
      <td>-</td>
      <td>ATAACGCAGTTTGCGT-1</td>
    </tr>
  </tbody>
</table>
<p>107385 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_data_K562</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">call_peaks</span><span class="p">(</span><span class="n">K562_SP1</span><span class="p">,</span> <span class="n">K562_brd4</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;MACCs&quot;</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">window_size</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">step_size</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                  <span class="n">pvalue_cutoffTTAA</span> <span class="o">=</span> <span class="mf">0.0001</span><span class="p">,</span>  <span class="n">pvalue_cutoffbg</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">lam_win_size</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="s2">&quot;peak_k562.bed&quot;</span><span class="p">)</span>
<span class="n">peak_data_K562</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the MACCs method with background, [expdata, background, reference, pvalue_cutoffbg, pvalue_cutoffTTAA, lam_win_size, window_size, step_size, extend, pseudocounts, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 24/24 [01:34&lt;00:00,  3.96s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Center</th>
      <th>Experiment Insertions</th>
      <th>Background insertions</th>
      <th>Reference Insertions</th>
      <th>pvalue Reference</th>
      <th>pvalue Background</th>
      <th>Fraction Experiment</th>
      <th>TPH Experiment</th>
      <th>Fraction background</th>
      <th>TPH background</th>
      <th>TPH background subtracted</th>
      <th>pvalue_adj Reference</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>29684</td>
      <td>30087</td>
      <td>29884.0</td>
      <td>6</td>
      <td>0</td>
      <td>1</td>
      <td>8.878753e-11</td>
      <td>1.546531e-04</td>
      <td>0.000018</td>
      <td>1832.256882</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1832.256882</td>
      <td>3.323285e-08</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>36239</td>
      <td>38107</td>
      <td>37578.0</td>
      <td>24</td>
      <td>2</td>
      <td>15</td>
      <td>0.000000e+00</td>
      <td>1.486029e-03</td>
      <td>0.000073</td>
      <td>7329.027530</td>
      <td>0.000019</td>
      <td>1862.457513</td>
      <td>5466.570017</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>198893</td>
      <td>201208</td>
      <td>200869.0</td>
      <td>28</td>
      <td>2</td>
      <td>11</td>
      <td>0.000000e+00</td>
      <td>6.927041e-05</td>
      <td>0.000086</td>
      <td>8550.532118</td>
      <td>0.000019</td>
      <td>1862.457513</td>
      <td>6688.074605</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>203351</td>
      <td>207161</td>
      <td>205004.0</td>
      <td>92</td>
      <td>13</td>
      <td>22</td>
      <td>0.000000e+00</td>
      <td>4.337485e-05</td>
      <td>0.000281</td>
      <td>28094.605530</td>
      <td>0.000121</td>
      <td>12105.973832</td>
      <td>15988.631698</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>265549</td>
      <td>266336</td>
      <td>265749.0</td>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>3.731359e-08</td>
      <td>4.678840e-03</td>
      <td>0.000015</td>
      <td>1526.880735</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1526.880735</td>
      <td>1.056034e-05</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>9404</th>
      <td>chrY</td>
      <td>15158250</td>
      <td>15158653</td>
      <td>15158450.0</td>
      <td>11</td>
      <td>0</td>
      <td>1</td>
      <td>0.000000e+00</td>
      <td>1.546531e-04</td>
      <td>0.000034</td>
      <td>3359.137618</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>3359.137618</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>9405</th>
      <td>chrY</td>
      <td>16985442</td>
      <td>16985845</td>
      <td>16985642.0</td>
      <td>5</td>
      <td>0</td>
      <td>2</td>
      <td>1.806731e-09</td>
      <td>4.678840e-03</td>
      <td>0.000015</td>
      <td>1526.880735</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1526.880735</td>
      <td>6.060202e-07</td>
    </tr>
    <tr>
      <th>9406</th>
      <td>chrY</td>
      <td>19753311</td>
      <td>19753714</td>
      <td>19753511.0</td>
      <td>33</td>
      <td>0</td>
      <td>1</td>
      <td>0.000000e+00</td>
      <td>2.269296e-13</td>
      <td>0.000101</td>
      <td>10077.412853</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>10077.412853</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>9407</th>
      <td>chrY</td>
      <td>21011133</td>
      <td>21011828</td>
      <td>21011333.0</td>
      <td>5</td>
      <td>0</td>
      <td>4</td>
      <td>2.510448e-09</td>
      <td>4.678840e-03</td>
      <td>0.000015</td>
      <td>1526.880735</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1526.880735</td>
      <td>8.296735e-07</td>
    </tr>
    <tr>
      <th>9408</th>
      <td>chrY</td>
      <td>56952574</td>
      <td>56957328</td>
      <td>56953707.0</td>
      <td>40</td>
      <td>1</td>
      <td>37</td>
      <td>0.000000e+00</td>
      <td>2.427052e-06</td>
      <td>0.000122</td>
      <td>12215.045883</td>
      <td>0.000009</td>
      <td>931.228756</td>
      <td>11283.817126</td>
      <td>0.000000e+00</td>
    </tr>
  </tbody>
</table>
<p>9409 rows × 15 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr10&quot;</span><span class="p">,</span><span class="mi">3048452</span><span class="p">,</span><span class="mi">3049913</span><span class="p">,</span><span class="mi">60000</span><span class="p">,</span><span class="n">peak_data_K562</span><span class="p">,</span><span class="n">K562_SP1</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">K562_brd4</span> <span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">15</span><span class="p">),</span><span class="n">peak_line</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span><span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">bins</span> <span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">example_length</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_28_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_28_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">qbed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;SP1&quot;</span><span class="p">:</span><span class="n">K562_SP1</span><span class="p">,</span> <span class="s2">&quot;Brd4&quot;</span><span class="p">:</span> <span class="n">K562_brd4</span><span class="p">}</span>
<span class="n">bed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;peak&quot;</span><span class="p">:</span><span class="n">peak_data_K562</span><span class="p">}</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">WashU_browser_url</span><span class="p">(</span><span class="n">qbed</span> <span class="o">=</span> <span class="n">qbed</span><span class="p">,</span><span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span><span class="p">,</span><span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;hg38&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=hg38&amp;hub=https://companion.epigenomegateway.org//task/db6498ca05ea546a0938a34cd8de001c/output//datahub.json
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">whole_peaks</span><span class="p">(</span><span class="n">peak_data_K562</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">70</span><span class="p">),</span><span class="n">height_scale</span> <span class="o">=</span> <span class="mf">1.7</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_30_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_30_0.png" />
</div>
</div>
<p>We can see that SP1 binds much more frequently in K562 than HCT116.</p>
<p>We can then check with reference Chip-seq data of SP1 in K562 from <a class="reference external" href="https://www.encodeproject.org/experiments/ENCSR372IML/">ENCSR372IML</a> (and use the bigWig file <a class="reference external" href="https://www.encodeproject.org/files/ENCFF588UII/">ENCFF588UII</a> generated by it）.</p>
<p>Download the data if needed:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!wget https://www.encodeproject.org/files/ENCFF588UII/@@download/ENCFF588UII.bigWig
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mtx_K562</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">calculate_signal</span><span class="p">(</span><span class="n">peak_data_K562</span><span class="p">,</span> <span class="n">chipseq_signal</span> <span class="o">=</span> <span class="s2">&quot;ENCFF588UII.bigWig&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 9409/9409 [00:41&lt;00:00, 225.34it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">signal_plot</span><span class="p">(</span><span class="n">mtx_K562</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_34_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_34_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">signal_heatmap</span><span class="p">(</span><span class="n">mtx_K562</span><span class="p">,</span><span class="n">pad</span> <span class="o">=</span>  <span class="mf">0.023</span><span class="p">,</span> <span class="n">belowlength</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_35_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_35_0.png" />
</div>
</div>
<p>We can see that calling cards peaks are consistent with Chip-seq data. Peak centers tend to have a higher signal and the signal goes lower as the distance increases.</p>
<p>Call motif to check the peak results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">call_motif</span><span class="p">(</span><span class="s2">&quot;peak_k562.bed&quot;</span><span class="p">,</span><span class="n">reference</span> <span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">,</span><span class="n">save_homer</span> <span class="o">=</span> <span class="s2">&quot;Homer/peak_k562&quot;</span><span class="p">,</span>
                 <span class="n">homer_path</span> <span class="o">=</span> <span class="s2">&quot;/home/juanru/miniconda3/bin/&quot;</span><span class="p">,</span> <span class="n">num_cores</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Peak data peak_k562.bed is used here.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

        Position file = peak_k562.bed
        Genome = hg38
        Output Directory = Homer/peak_k562
        Fragment size set to 1000
        Using 12 CPUs
        Will not run homer for de novo motifs
        Found mset for &#34;human&#34;, will check against vertebrates motifs
        Peak/BED file conversion summary:
                BED/Header formatted lines: 9409
                peakfile formatted lines: 0

        Peak File Statistics:
                Total Peaks: 9409
                Redundant Peak IDs: 2
                Peaks lacking information: 0 (need at least 5 columns per peak)
                Peaks with misformatted coordinates: 0 (should be integer)
                Peaks with misformatted strand: 0 (should be either +/- or 0/1)

        Redunant Peaks found: Remove or rename these or some programs may have trouble...

        2 duplicate peak IDs out of 9409 total peaks
        Background files for 1000 bp fragments found.
        Custom genome sequence directory: /home/juanru/miniconda3/share/homer/.//data/genomes/hg38//

        Extracting sequences from file: /home/juanru/miniconda3/share/homer/.//data/genomes/hg38///genome.fa
        Looking for peak sequences in a single file (/home/juanru/miniconda3/share/homer/.//data/genomes/hg38///genome.fa)
        Extracting 1062 sequences from chr1
        Extracting 426 sequences from chr10
        Extracting 475 sequences from chr11
        Extracting 443 sequences from chr12
        Extracting 182 sequences from chr13
        Extracting 227 sequences from chr14
        Extracting 302 sequences from chr15
        Extracting 342 sequences from chr16
        Extracting 409 sequences from chr17
        Extracting 169 sequences from chr18
        Extracting 505 sequences from chr19
        Extracting 660 sequences from chr2
        Extracting 249 sequences from chr20
        Extracting 154 sequences from chr21
        Extracting 186 sequences from chr22
        Extracting 508 sequences from chr3
        Extracting 428 sequences from chr4
        Extracting 408 sequences from chr5
        Extracting 620 sequences from chr6
        Extracting 572 sequences from chr7
        Extracting 375 sequences from chr8
        Extracting 408 sequences from chr9
        Extracting 281 sequences from chrX
        Extracting 18 sequences from chrY

        Not removing redundant sequences


        Sequences processed:
                Auto detected maximum sequence length of 1001 bp
                9409 total

        Frequency Bins: 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.6 0.7 0.8
        Freq    Bin     Count
        0.3     2       52
        0.35    3       412
        0.4     4       1127
        0.45    5       1812
        0.5     6       1961
        0.6     7       2657
        0.7     8       1164
        0.8     9       219
        10      10      5

        Total sequences set to 50000

        Choosing background that matches in CpG/GC content...
        Bin     # Targets       # Background    Background Weight
        2       52      224     1.001
        3       412     1777    1.000
        4       1127    4862    1.000
        5       1812    7817    1.000
        6       1961    8460    1.000
        7       2657    11462   1.000
        8       1164    5022    1.000
        9       219     945     1.000
        10      5       22      0.980
        Assembling sequence file...
        Normalizing lower order oligos using homer2

        Reading input files...
        50000 total sequences read
        Autonormalization: 1-mers (4 total)
                A       25.36%  25.49%  0.995
                C       24.64%  24.51%  1.005
                G       24.64%  24.51%  1.005
                T       25.36%  25.49%  0.995
        Autonormalization: 2-mers (16 total)
                AA      7.80%   7.45%   1.047
                CA      6.85%   7.40%   0.925
                GA      6.01%   6.09%   0.986
                TA      4.70%   4.54%   1.036
                AC      4.92%   4.99%   0.986
                CC      7.54%   7.44%   1.013
                GC      6.17%   5.99%   1.029
                TC      6.01%   6.09%   0.986
                AG      7.26%   7.47%   0.972
                CG      2.98%   2.20%   1.356
                GG      7.54%   7.44%   1.013
                TG      6.85%   7.40%   0.925
                AT      5.37%   5.57%   0.966
                CT      7.26%   7.47%   0.972
                GT      4.92%   4.99%   0.986
                TT      7.80%   7.45%   1.047
        Autonormalization: 3-mers (64 total)
        Normalization weights can be found in file: Homer/peak_k562/seq.autonorm.tsv
        Converging on autonormalization solution:
        ...............................................................................
        Final normalization:    Autonormalization: 1-mers (4 total)
                A       25.36%  25.34%  1.001
                C       24.64%  24.66%  0.999
                G       24.64%  24.66%  0.999
                T       25.36%  25.34%  1.001
        Autonormalization: 2-mers (16 total)
                AA      7.80%   7.68%   1.016
                CA      6.85%   6.98%   0.982
                GA      6.01%   6.00%   1.000
                TA      4.70%   4.67%   1.006
                AC      4.92%   4.88%   1.008
                CC      7.54%   7.52%   1.002
                GC      6.17%   6.25%   0.987
                TC      6.01%   6.00%   1.000
                AG      7.26%   7.24%   1.003
                CG      2.98%   2.92%   1.023
                GG      7.54%   7.52%   1.002
                TG      6.85%   6.98%   0.982
                AT      5.37%   5.53%   0.971
                CT      7.26%   7.24%   1.003
                GT      4.92%   4.88%   1.008
                TT      7.80%   7.68%   1.016
        Autonormalization: 3-mers (64 total)
        Finished preparing sequence/group files

        ----------------------------------------------------------
        Known motif enrichment

        Reading input files...
        50000 total sequences read
        440 motifs loaded
        Cache length = 11180
        Using binomial scoring
        Checking enrichment of 440 motif(s)
        |0%                                    50%                                  100%|
        =================================================================================
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finished!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

        Preparing HTML output with sequence logos...
                1 of 440 (1e-169) Sp2(Zf)/HEK293-Sp2.eGFP-ChIP-Seq(Encode)/Homer
                2 of 440 (1e-148) Sp5(Zf)/mES-Sp5.Flag-ChIP-Seq(GSE72989)/Homer
                3 of 440 (1e-137) KLF1(Zf)/HUDEP2-KLF1-CutnRun(GSE136251)/Homer
                4 of 440 (1e-125) KLF6(Zf)/PDAC-KLF6-ChIP-Seq(GSE64557)/Homer
                5 of 440 (1e-124) KLF5(Zf)/LoVo-KLF5-ChIP-Seq(GSE49402)/Homer
                6 of 440 (1e-123) Sp1(Zf)/Promoter/Homer
                7 of 440 (1e-114) KLF3(Zf)/MEF-Klf3-ChIP-Seq(GSE44748)/Homer
                8 of 440 (1e-114) Maz(Zf)/HepG2-Maz-ChIP-Seq(GSE31477)/Homer
                9 of 440 (1e-100) KLF14(Zf)/HEK293-KLF14.GFP-ChIP-Seq(GSE58341)/Homer
                10 of 440 (1e-96) Klf9(Zf)/GBM-Klf9-ChIP-Seq(GSE62211)/Homer
                11 of 440 (1e-74) Jun-AP1(bZIP)/K562-cJun-ChIP-Seq(GSE31477)/Homer
                12 of 440 (1e-67) Gata2(Zf)/K562-GATA2-ChIP-Seq(GSE18829)/Homer
                13 of 440 (1e-65) Fosl2(bZIP)/3T3L1-Fosl2-ChIP-Seq(GSE56872)/Homer
                14 of 440 (1e-61) TATA-Box(TBP)/Promoter/Homer
                15 of 440 (1e-60) Gata4(Zf)/Heart-Gata4-ChIP-Seq(GSE35151)/Homer
                16 of 440 (1e-59) Gata6(Zf)/HUG1N-GATA6-ChIP-Seq(GSE51936)/Homer
                17 of 440 (1e-57) Bach2(bZIP)/OCILy7-Bach2-ChIP-Seq(GSE44420)/Homer
                18 of 440 (1e-52) BMYB(HTH)/Hela-BMYB-ChIP-Seq(GSE27030)/Homer
                19 of 440 (1e-52) GATA3(Zf)/iTreg-Gata3-ChIP-Seq(GSE20898)/Homer
                20 of 440 (1e-48) Fra2(bZIP)/Striatum-Fra2-ChIP-Seq(GSE43429)/Homer
                21 of 440 (1e-47) Klf4(Zf)/mES-Klf4-ChIP-Seq(GSE11431)/Homer
                22 of 440 (1e-46) Gata1(Zf)/K562-GATA1-ChIP-Seq(GSE18829)/Homer
                23 of 440 (1e-45) Fos(bZIP)/TSC-Fos-ChIP-Seq(GSE110950)/Homer
                24 of 440 (1e-45) JunB(bZIP)/DendriticCells-Junb-ChIP-Seq(GSE36099)/Homer
                25 of 440 (1e-43) Fra1(bZIP)/BT549-Fra1-ChIP-Seq(GSE46166)/Homer
                26 of 440 (1e-41) AMYB(HTH)/Testes-AMYB-ChIP-Seq(GSE44588)/Homer
                27 of 440 (1e-39) TRPS1(Zf)/MCF7-TRPS1-ChIP-Seq(GSE107013)/Homer
                28 of 440 (1e-36) MYB(HTH)/ERMYB-Myb-ChIPSeq(GSE22095)/Homer
                29 of 440 (1e-36) KLF10(Zf)/HEK293-KLF10.GFP-ChIP-Seq(GSE58341)/Homer
                30 of 440 (1e-36) Atf3(bZIP)/GBM-ATF3-ChIP-Seq(GSE33912)/Homer
                31 of 440 (1e-35) Bach1(bZIP)/K562-Bach1-ChIP-Seq(GSE31477)/Homer
                32 of 440 (1e-34) BATF(bZIP)/Th17-BATF-ChIP-Seq(GSE39756)/Homer
                33 of 440 (1e-33) NF-E2(bZIP)/K562-NFE2-ChIP-Seq(GSE31477)/Homer
                34 of 440 (1e-33) Pitx1(Homeobox)/Chicken-Pitx1-ChIP-Seq(GSE38910)/Homer
                35 of 440 (1e-32) WT1(Zf)/Kidney-WT1-ChIP-Seq(GSE90016)/Homer
                36 of 440 (1e-32) Elk4(ETS)/Hela-Elk4-ChIP-Seq(GSE31477)/Homer
                37 of 440 (1e-28) Isl1(Homeobox)/Neuron-Isl1-ChIP-Seq(GSE31456)/Homer
                38 of 440 (1e-27) AP-1(bZIP)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer
                39 of 440 (1e-27) Nrf2(bZIP)/Lymphoblast-Nrf2-ChIP-Seq(GSE37589)/Homer
                40 of 440 (1e-26) Elk1(ETS)/Hela-Elk1-ChIP-Seq(GSE31477)/Homer
                41 of 440 (1e-24) ZNF467(Zf)/HEK293-ZNF467.GFP-ChIP-Seq(GSE58341)/Homer
                42 of 440 (1e-23) NFE2L2(bZIP)/HepG2-NFE2L2-ChIP-Seq(Encode)/Homer
                43 of 440 (1e-23) E2F6(E2F)/Hela-E2F6-ChIP-Seq(GSE31477)/Homer
                44 of 440 (1e-23) CTCF(Zf)/CD4+-CTCF-ChIP-Seq(Barski_et_al.)/Homer
                45 of 440 (1e-23) ETV4(ETS)/HepG2-ETV4-ChIP-Seq(ENCODE)/Homer
                46 of 440 (1e-23) Fli1(ETS)/CD8-FLI-ChIP-Seq(GSE20898)/Homer
                47 of 440 (1e-23) Egr1(Zf)/K562-Egr1-ChIP-Seq(GSE32465)/Homer
                48 of 440 (1e-22) GATA:SCL(Zf,bHLH)/Ter119-SCL-ChIP-Seq(GSE18720)/Homer
                49 of 440 (1e-21) Egr2(Zf)/Thymocytes-Egr2-ChIP-Seq(GSE34254)/Homer
                50 of 440 (1e-21) Zfp281(Zf)/ES-Zfp281-ChIP-Seq(GSE81042)/Homer
                51 of 440 (1e-20) ETV1(ETS)/GIST48-ETV1-ChIP-Seq(GSE22441)/Homer
                52 of 440 (1e-19) Elf4(ETS)/BMDM-Elf4-ChIP-Seq(GSE88699)/Homer
                53 of 440 (1e-19) Nkx6.1(Homeobox)/Islet-Nkx6.1-ChIP-Seq(GSE40975)/Homer
                54 of 440 (1e-18) ZNF652/HepG2-ZNF652.Flag-ChIP-Seq(Encode)/Homer
                55 of 440 (1e-18) Hoxa13(Homeobox)/ChickenMSG-Hoxa13.Flag-ChIP-Seq(GSE86088)/Homer
                56 of 440 (1e-16) Znf263(Zf)/K562-Znf263-ChIP-Seq(GSE31477)/Homer
                57 of 440 (1e-15) E2F4(E2F)/K562-E2F4-ChIP-Seq(GSE31477)/Homer
                58 of 440 (1e-15) ELF1(ETS)/Jurkat-ELF1-ChIP-Seq(SRA014231)/Homer
                59 of 440 (1e-15) EKLF(Zf)/Erythrocyte-Klf1-ChIP-Seq(GSE20478)/Homer
                60 of 440 (1e-14) GABPA(ETS)/Jurkat-GABPa-ChIP-Seq(GSE17954)/Homer
                61 of 440 (1e-13) CRX(Homeobox)/Retina-Crx-ChIP-Seq(GSE20012)/Homer
                62 of 440 (1e-13) Lhx3(Homeobox)/Neuron-Lhx3-ChIP-Seq(GSE31456)/Homer
                63 of 440 (1e-12) Hoxa11(Homeobox)/ChickenMSG-Hoxa11.Flag-ChIP-Seq(GSE86088)/Homer
                64 of 440 (1e-11) STAT5(Stat)/mCD4+-Stat5-ChIP-Seq(GSE12346)/Homer
                65 of 440 (1e-11) ETS1(ETS)/Jurkat-ETS1-ChIP-Seq(GSE17954)/Homer
                66 of 440 (1e-11) ZNF317(Zf)/HEK293-ZNF317.GFP-ChIP-Seq(GSE58341)/Homer
                67 of 440 (1e-11) En1(Homeobox)/SUM149-EN1-ChIP-Seq(GSE120957)/Homer
                68 of 440 (1e-10) BORIS(Zf)/K562-CTCFL-ChIP-Seq(GSE32465)/Homer
                69 of 440 (1e-10) MITF(bHLH)/MastCells-MITF-ChIP-Seq(GSE48085)/Homer
                70 of 440 (1e-10) RORg(NR)/Liver-Rorc-ChIP-Seq(GSE101115)/Homer
                71 of 440 (1e-10) Hoxd13(Homeobox)/ChickenMSG-Hoxd13.Flag-ChIP-Seq(GSE86088)/Homer
                72 of 440 (1e-10) EHF(ETS)/LoVo-EHF-ChIP-Seq(GSE49402)/Homer
                73 of 440 (1e-10) EWS:FLI1-fusion(ETS)/SK_N_MC-EWS:FLI1-ChIP-Seq(SRA014231)/Homer
                74 of 440 (1e-9) RXR(NR),DR1/3T3L1-RXR-ChIP-Seq(GSE13511)/Homer
                75 of 440 (1e-9) ELF5(ETS)/T47D-ELF5-ChIP-Seq(GSE30407)/Homer
                76 of 440 (1e-9) PU.1(ETS)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer
                77 of 440 (1e-9) FOXK2(Forkhead)/U2OS-FOXK2-ChIP-Seq(E-MTAB-2204)/Homer
                78 of 440 (1e-9) Hoxd11(Homeobox)/ChickenMSG-Hoxd11.Flag-ChIP-Seq(GSE86088)/Homer
                79 of 440 (1e-9) PPARa(NR),DR1/Liver-Ppara-ChIP-Seq(GSE47954)/Homer
                80 of 440 (1e-9) PU.1-IRF(ETS:IRF)/Bcell-PU.1-ChIP-Seq(GSE21512)/Homer
                81 of 440 (1e-9) Stat3(Stat)/mES-Stat3-ChIP-Seq(GSE11431)/Homer
                82 of 440 (1e-9) Unknown(Homeobox)/Limb-p300-ChIP-Seq/Homer
                83 of 440 (1e-9) Hoxa9(Homeobox)/ChickenMSG-Hoxa9.Flag-ChIP-Seq(GSE86088)/Homer
                84 of 440 (1e-9) ETS(ETS)/Promoter/Homer
                85 of 440 (1e-8) HIF-1b(HLH)/T47D-HIF1b-ChIP-Seq(GSE59937)/Homer
                86 of 440 (1e-8) Hnf1(Homeobox)/Liver-Foxa2-Chip-Seq(GSE25694)/Homer
                87 of 440 (1e-8) STAT1(Stat)/HelaS3-STAT1-ChIP-Seq(GSE12782)/Homer
                88 of 440 (1e-8) PPARE(NR),DR1/3T3L1-Pparg-ChIP-Seq(GSE13511)/Homer
                89 of 440 (1e-8) Smad4(MAD)/ESC-SMAD4-ChIP-Seq(GSE29422)/Homer
                90 of 440 (1e-8) PR(NR)/T47D-PR-ChIP-Seq(GSE31130)/Homer
                91 of 440 (1e-8) ERG(ETS)/VCaP-ERG-ChIP-Seq(GSE14097)/Homer
                92 of 440 (1e-8) MafK(bZIP)/C2C12-MafK-ChIP-Seq(GSE36030)/Homer
                93 of 440 (1e-8) MafB(bZIP)/BMM-Mafb-ChIP-Seq(GSE75722)/Homer
                94 of 440 (1e-8) FOXP1(Forkhead)/H9-FOXP1-ChIP-Seq(GSE31006)/Homer
                95 of 440 (1e-7) E2F3(E2F)/MEF-E2F3-ChIP-Seq(GSE71376)/Homer
                96 of 440 (1e-7) E2F7(E2F)/Hela-E2F7-ChIP-Seq(GSE32673)/Homer
                97 of 440 (1e-7) Lhx1(Homeobox)/EmbryoCarcinoma-Lhx1-ChIP-Seq(GSE70957)/Homer
                98 of 440 (1e-7) Smad2(MAD)/ES-SMAD2-ChIP-Seq(GSE29422)/Homer
                99 of 440 (1e-7) Nanog(Homeobox)/mES-Nanog-ChIP-Seq(GSE11724)/Homer
                100 of 440 (1e-7) Max(bHLH)/K562-Max-ChIP-Seq(GSE31477)/Homer
                101 of 440 (1e-7) DLX2(Homeobox)/BasalGanglia-Dlx2-ChIP-seq(GSE124936)/Homer
                102 of 440 (1e-7) MafA(bZIP)/Islet-MafA-ChIP-Seq(GSE30298)/Homer
                103 of 440 (1e-7) ZFX(Zf)/mES-Zfx-ChIP-Seq(GSE11431)/Homer
                104 of 440 (1e-7) DLX1(Homeobox)/BasalGanglia-Dlx1-ChIP-seq(GSE124936)/Homer
                105 of 440 (1e-7) E2F1(E2F)/Hela-E2F1-ChIP-Seq(GSE22478)/Homer
                106 of 440 (1e-7) CHR(?)/Hela-CellCycle-Expression/Homer
                107 of 440 (1e-6) Ets1-distal(ETS)/CD4+-PolII-ChIP-Seq(Barski_et_al.)/Homer
                108 of 440 (1e-6) MafF(bZIP)/HepG2-MafF-ChIP-Seq(GSE31477)/Homer
                109 of 440 (1e-6) USF1(bHLH)/GM12878-Usf1-ChIP-Seq(GSE32465)/Homer
                110 of 440 (1e-6) LRF(Zf)/Erythroblasts-ZBTB7A-ChIP-Seq(GSE74977)/Homer
                111 of 440 (1e-6) Foxo3(Forkhead)/U2OS-Foxo3-ChIP-Seq(E-MTAB-2701)/Homer
                112 of 440 (1e-6) SpiB(ETS)/OCILY3-SPIB-ChIP-Seq(GSE56857)/Homer
                113 of 440 (1e-6) p53(p53)/mES-cMyc-ChIP-Seq(GSE11431)/Homer
                114 of 440 (1e-6) Tbox:Smad(T-box,MAD)/ESCd5-Smad2_3-ChIP-Seq(GSE29422)/Homer
                115 of 440 (1e-6) Otx2(Homeobox)/EpiLC-Otx2-ChIP-Seq(GSE56098)/Homer
                116 of 440 (1e-6) MYNN(Zf)/HEK293-MYNN.eGFP-ChIP-Seq(Encode)/Homer
                117 of 440 (1e-6) YY1(Zf)/Promoter/Homer
                118 of 440 (1e-6) SPDEF(ETS)/VCaP-SPDEF-ChIP-Seq(SRA014231)/Homer
                119 of 440 (1e-5) NeuroD1(bHLH)/Islet-NeuroD1-ChIP-Seq(GSE30298)/Homer
                120 of 440 (1e-5) SCL(bHLH)/HPC7-Scl-ChIP-Seq(GSE13511)/Homer
                121 of 440 (1e-5) IRF3(IRF)/BMDM-Irf3-ChIP-Seq(GSE67343)/Homer
                122 of 440 (1e-5) GFY-Staf(?,Zf)/Promoter/Homer
                123 of 440 (1e-5) Foxf1(Forkhead)/Lung-Foxf1-ChIP-Seq(GSE77951)/Homer
                124 of 440 (1e-5) Foxo1(Forkhead)/RAW-Foxo1-ChIP-Seq(Fan_et_al.)/Homer
                125 of 440 (1e-5) CLOCK(bHLH)/Liver-Clock-ChIP-Seq(GSE39860)/Homer
                126 of 440 (1e-5) ETS:E-box(ETS,bHLH)/HPC7-Scl-ChIP-Seq(GSE22178)/Homer
                127 of 440 (1e-5) Atoh1(bHLH)/Cerebellum-Atoh1-ChIP-Seq(GSE22111)/Homer
                128 of 440 (1e-5) Tbr1(T-box)/Cortex-Tbr1-ChIP-Seq(GSE71384)/Homer
                129 of 440 (1e-4) STAT4(Stat)/CD4-Stat4-ChIP-Seq(GSE22104)/Homer
                130 of 440 (1e-4) Sox3(HMG)/NPC-Sox3-ChIP-Seq(GSE33059)/Homer
                131 of 440 (1e-4) c-Myc(bHLH)/LNCAP-cMyc-ChIP-Seq(Unpublished)/Homer
                132 of 440 (1e-4) LHX9(Homeobox)/Hct116-LHX9.V5-ChIP-Seq(GSE116822)/Homer
                133 of 440 (1e-4) ELF3(ETS)/PDAC-ELF3-ChIP-Seq(GSE64557)/Homer
                134 of 440 (1e-4) Stat3+il21(Stat)/CD4-Stat3-ChIP-Seq(GSE19198)/Homer
                135 of 440 (1e-4) DLX5(Homeobox)/BasalGanglia-Dlx5-ChIP-seq(GSE124936)/Homer
                136 of 440 (1e-4) CEBP:AP1(bZIP)/ThioMac-CEBPb-ChIP-Seq(GSE21512)/Homer
                137 of 440 (1e-4) AR-halfsite(NR)/LNCaP-AR-ChIP-Seq(GSE27824)/Homer
                138 of 440 (1e-4) Six1(Homeobox)/Myoblast-Six1-ChIP-Chip(GSE20150)/Homer
                139 of 440 (1e-4) Tcf12(bHLH)/GM12878-Tcf12-ChIP-Seq(GSE32465)/Homer
                140 of 440 (1e-4) CArG(MADS)/PUER-Srf-ChIP-Seq(Sullivan_et_al.)/Homer
                141 of 440 (1e-3) Etv2(ETS)/ES-ER71-ChIP-Seq(GSE59402)/Homer
                142 of 440 (1e-3) Zfp57(Zf)/H1-ZFP57.HA-ChIP-Seq(GSE115387)/Homer
                143 of 440 (1e-3) NFY(CCAAT)/Promoter/Homer
                144 of 440 (1e-3) PU.1:IRF8(ETS:IRF)/pDC-Irf8-ChIP-Seq(GSE66899)/Homer
                145 of 440 (1e-3) NPAS(bHLH)/Liver-NPAS-ChIP-Seq(GSE39860)/Homer
                146 of 440 (1e-3) EWS:ERG-fusion(ETS)/CADO_ES1-EWS:ERG-ChIP-Seq(SRA014231)/Homer
                147 of 440 (1e-3) FOXK1(Forkhead)/HEK293-FOXK1-ChIP-Seq(GSE51673)/Homer
                148 of 440 (1e-3) Zic2(Zf)/ESC-Zic2-ChIP-Seq(SRP197560)/Homer
                149 of 440 (1e-3) MNT(bHLH)/HepG2-MNT-ChIP-Seq(Encode)/Homer
                150 of 440 (1e-3) IRF2(IRF)/Erythroblas-IRF2-ChIP-Seq(GSE36985)/Homer
                151 of 440 (1e-3) Tcf21(bHLH)/ArterySmoothMuscle-Tcf21-ChIP-Seq(GSE61369)/Homer
                152 of 440 (1e-3) BHLHA15(bHLH)/NIH3T3-BHLHB8.HA-ChIP-Seq(GSE119782)/Homer
                153 of 440 (1e-3) Ap4(bHLH)/AML-Tfap4-ChIP-Seq(GSE45738)/Homer
                154 of 440 (1e-3) IRF8(IRF)/BMDM-IRF8-ChIP-Seq(GSE77884)/Homer
                155 of 440 (1e-3) NFAT(RHD)/Jurkat-NFATC1-ChIP-Seq(Jolma_et_al.)/Homer
                156 of 440 (1e-3) Unknown-ESC-element(?)/mES-Nanog-ChIP-Seq(GSE11724)/Homer
                157 of 440 (1e-3) Sox9(HMG)/Limb-SOX9-ChIP-Seq(GSE73225)/Homer
                158 of 440 (1e-3) MyoG(bHLH)/C2C12-MyoG-ChIP-Seq(GSE36024)/Homer
                159 of 440 (1e-3) Six2(Homeobox)/NephronProgenitor-Six2-ChIP-Seq(GSE39837)/Homer
                160 of 440 (1e-3) Lhx2(Homeobox)/HFSC-Lhx2-ChIP-Seq(GSE48068)/Homer
                161 of 440 (1e-3) GSC(Homeobox)/FrogEmbryos-GSC-ChIP-Seq(DRA000576)/Homer
                162 of 440 (1e-3) HNF1b(Homeobox)/PDAC-HNF1B-ChIP-Seq(GSE64557)/Homer
                163 of 440 (1e-3) IRF4(IRF)/GM12878-IRF4-ChIP-Seq(GSE32465)/Homer
                164 of 440 (1e-3) Prop1(Homeobox)/GHFT1-PROP1.biotin-ChIP-Seq(GSE77302)/Homer
                165 of 440 (1e-3) NeuroG2(bHLH)/Fibroblast-NeuroG2-ChIP-Seq(GSE75910)/Homer
                166 of 440 (1e-3) BMAL1(bHLH)/Liver-Bmal1-ChIP-Seq(GSE39860)/Homer
                167 of 440 (1e-3) ZNF189(Zf)/HEK293-ZNF189.GFP-ChIP-Seq(GSE58341)/Homer
                168 of 440 (1e-3) EBF1(EBF)/Near-E2A-ChIP-Seq(GSE21512)/Homer
                169 of 440 (1e-3) HIF2a(bHLH)/785_O-HIF2a-ChIP-Seq(GSE34871)/Homer
                170 of 440 (1e-3) E2F(E2F)/Hela-CellCycle-Expression/Homer
                171 of 440 (1e-2) Rfx6(HTH)/Min6b1-Rfx6.HA-ChIP-Seq(GSE62844)/Homer
                172 of 440 (1e-2) EBF2(EBF)/BrownAdipose-EBF2-ChIP-Seq(GSE97114)/Homer
                173 of 440 (1e-2) Zac1(Zf)/Neuro2A-Plagl1-ChIP-Seq(GSE75942)/Homer
                174 of 440 (1e-2) VDR(NR),DR3/GM10855-VDR+vitD-ChIP-Seq(GSE22484)/Homer
                175 of 440 (1e-2) Rbpj1(?)/Panc1-Rbpj1-ChIP-Seq(GSE47459)/Homer
                176 of 440 (1e-2) STAT6(Stat)/Macrophage-Stat6-ChIP-Seq(GSE38377)/Homer
                177 of 440 (1e-2) Ascl1(bHLH)/NeuralTubes-Ascl1-ChIP-Seq(GSE55840)/Homer
                178 of 440 (1e-2) n-Myc(bHLH)/mES-nMyc-ChIP-Seq(GSE11431)/Homer
                179 of 440 (1e-2) MyoD(bHLH)/Myotube-MyoD-ChIP-Seq(GSE21614)/Homer
                180 of 440 (1e-2) FOXA1(Forkhead)/MCF7-FOXA1-ChIP-Seq(GSE26831)/Homer
                181 of 440 (1e-2) Zfp809(Zf)/ES-Zfp809-ChIP-Seq(GSE70799)/Homer
                182 of 440 (1e-2) Sox21(HMG)/ESC-SOX21-ChIP-Seq(GSE110505)/Homer
                183 of 440 (1e-2) CRE(bZIP)/Promoter/Homer
                184 of 440 (1e-2) RUNX1(Runt)/Jurkat-RUNX1-ChIP-Seq(GSE29180)/Homer
                185 of 440 (1e-2) JunD(bZIP)/K562-JunD-ChIP-Seq/Homer
                186 of 440 (1e-2) Foxa3(Forkhead)/Liver-Foxa3-ChIP-Seq(GSE77670)/Homer
                187 of 440 (1e-2) FOXA1(Forkhead)/LNCAP-FOXA1-ChIP-Seq(GSE27824)/Homer
                188 of 440 (1e-2) ZNF711(Zf)/SHSY5Y-ZNF711-ChIP-Seq(GSE20673)/Homer
                189 of 440 (1e-2) TCF4(bHLH)/SHSY5Y-TCF4-ChIP-Seq(GSE96915)/Homer
                190 of 440 (1e-2) PRDM15(Zf)/ESC-Prdm15-ChIP-Seq(GSE73694)/Homer
                191 of 440 (1e-2) Esrrb(NR)/mES-Esrrb-ChIP-Seq(GSE11431)/Homer
                192 of 440 (1e-2) ZNF692(Zf)/HEK293-ZNF692.GFP-ChIP-Seq(GSE58341)/Homer
                193 of 440 (1e-2) PRDM1(Zf)/Hela-PRDM1-ChIP-Seq(GSE31477)/Homer
                194 of 440 (1e-2) Sox10(HMG)/SciaticNerve-Sox3-ChIP-Seq(GSE35132)/Homer
                195 of 440 (1e-2) Zic(Zf)/Cerebellum-ZIC1.2-ChIP-Seq(GSE60731)/Homer
                196 of 440 (1e-2) PRDM14(Zf)/H1-PRDM14-ChIP-Seq(GSE22767)/Homer
                197 of 440 (1e-2) Myf5(bHLH)/GM-Myf5-ChIP-Seq(GSE24852)/Homer
                198 of 440 (1e-2) Barx1(Homeobox)/Stomach-Barx1.3xFlag-ChIP-Seq(GSE69483)/Homer
                199 of 440 (1e-2) IRF1(IRF)/PBMC-IRF1-ChIP-Seq(GSE43036)/Homer
                200 of 440 (1e-2) AP-2gamma(AP2)/MCF7-TFAP2C-ChIP-Seq(GSE21234)/Homer
                201 of 440 (1e-2) bHLHE40(bHLH)/HepG2-BHLHE40-ChIP-Seq(GSE31477)/Homer
                202 of 440 (1e-2) ZBTB18(Zf)/HEK293-ZBTB18.GFP-ChIP-Seq(GSE58341)/Homer
                203 of 440 (1e-2) TR4(NR),DR1/Hela-TR4-ChIP-Seq(GSE24685)/Homer
                204 of 440 (1e-2) Olig2(bHLH)/Neuron-Olig2-ChIP-Seq(GSE30882)/Homer
                205 of 440 (1e-2) HNF4a(NR),DR1/HepG2-HNF4a-ChIP-Seq(GSE25021)/Homer
                206 of 440 (1e-2) Twist2(bHLH)/Myoblast-Twist2.Ty1-ChIP-Seq(GSE127998)/Homer
                207 of 440 (1e-2) FOXM1(Forkhead)/MCF7-FOXM1-ChIP-Seq(GSE72977)/Homer
                208 of 440 (1e-2) RORgt(NR)/EL4-RORgt.Flag-ChIP-Seq(GSE56019)/Homer
                209 of 440 (1e-2) RORgt(NR)/EL4-RORgt.Flag-ChIP-Seq(GSE56019)/Homer
                210 of 440 (1e-2) STAT6(Stat)/CD4-Stat6-ChIP-Seq(GSE22104)/Homer
                211 of 440 (1e-2) RUNX2(Runt)/PCa-RUNX2-ChIP-Seq(GSE33889)/Homer
                212 of 440 (1e-2) Eomes(T-box)/H9-Eomes-ChIP-Seq(GSE26097)/Homer
                213 of 440 (1e-2) Six4(Homeobox)/MCF7-SIX4-ChIP-Seq(Encode)/Homer
                214 of 440 (1e-2) ZNF341(Zf)/EBV-ZNF341-ChIP-Seq(GSE113194)/Homer
                215 of 440 (1e-2) Foxa2(Forkhead)/Liver-Foxa2-ChIP-Seq(GSE25694)/Homer
        Skipping...
        Job finished - if results look good, please send beer to ..

        Cleaning up tmp files...

</pre></div></div>
</div>
<p>In the motif analysis result, SP1 motif and many other family members rank top.</p>
<p><img alt="drawing" class="no-scaled-link" src="../../_images/peak_k562.png" style="width: 800px;" /></p>
<p>Next we want to identify binding sites that are differentially bound in K562 and Hct-116 cells. This can be challenging as the two samples may have slightly shifted peaks centers at a given genomic region, leading to false positive differential peak calls. To handle this, Pycallingcards first combines the insertions from the two samples and calls peaks on the joint dataset. We do this using <a class="reference external" href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and
<a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pybedtools</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">rd</span><span class="o">.</span><span class="n">combine_qbed</span><span class="p">([</span><span class="n">peak_data_HCT116</span><span class="p">,</span> <span class="n">peak_data_K562</span><span class="p">])</span>
<span class="n">peak</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="o">.</span><span class="n">from_dataframe</span><span class="p">(</span><span class="n">peak</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">peak_data</span> <span class="o">=</span> <span class="n">peak</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chrom&quot;</span><span class="p">:</span><span class="s2">&quot;Chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span><span class="s2">&quot;End&quot;</span><span class="p">})</span>
<span class="n">peak_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>29684</td>
      <td>30087</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>36239</td>
      <td>38107</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>198893</td>
      <td>201208</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>203351</td>
      <td>207161</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>265549</td>
      <td>266336</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>10445</th>
      <td>chrY</td>
      <td>15158250</td>
      <td>15158653</td>
    </tr>
    <tr>
      <th>10446</th>
      <td>chrY</td>
      <td>16985442</td>
      <td>16985845</td>
    </tr>
    <tr>
      <th>10447</th>
      <td>chrY</td>
      <td>19753311</td>
      <td>19753714</td>
    </tr>
    <tr>
      <th>10448</th>
      <td>chrY</td>
      <td>21011133</td>
      <td>21011828</td>
    </tr>
    <tr>
      <th>10449</th>
      <td>chrY</td>
      <td>56952574</td>
      <td>56957328</td>
    </tr>
  </tbody>
</table>
<p>10450 rows × 3 columns</p>
</div></div>
</div>
<p>We can now visualize the peaks called on the joint dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span><span class="mi">999921</span><span class="p">,</span><span class="mi">1000324</span><span class="p">,</span><span class="mi">15000</span><span class="p">,</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">HCT116_SP1</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">HCT116_brd4</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">peak_line</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span><span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">example_length</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
                <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;HCT116_SP1&quot;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span><span class="mi">999921</span><span class="p">,</span><span class="mi">1000324</span><span class="p">,</span><span class="mi">15000</span><span class="p">,</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">K562_SP1</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">K562_brd4</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">peak_line</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span><span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">example_length</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
                <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;K562_SP1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_43_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_43_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_43_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_43_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr10&quot;</span><span class="p">,</span><span class="mi">3048452</span><span class="p">,</span><span class="mi">3049913</span><span class="p">,</span><span class="mi">60000</span><span class="p">,</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">HCT116_SP1</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">HCT116_brd4</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">14</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span>
                <span class="n">example_length</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;HCT116_SP1&quot;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr10&quot;</span><span class="p">,</span><span class="mi">3048452</span><span class="p">,</span><span class="mi">3049913</span><span class="p">,</span><span class="mi">60000</span><span class="p">,</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">K562_SP1</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">K562_brd4</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">14</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span>
                <span class="n">example_length</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;K562_SP1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_44_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_44_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_44_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_44_1.png" />
</div>
</div>
<p>The results seem to be good! Congratulations! Now we can annotate the peaks using bedtools.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_annotation</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">annotation</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">)</span>
<span class="n">peak_annotation</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">combine_annotation</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span><span class="n">peak_annotation</span><span class="p">)</span>
<span class="n">peak_annotation</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In the bedtools method, we would use bedtools in the default path. Set bedtools path by &#39;bedtools_path&#39; if needed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Nearest Refseq1</th>
      <th>Gene Name1</th>
      <th>Direction1</th>
      <th>Distance1</th>
      <th>Nearest Refseq2</th>
      <th>Gene Name2</th>
      <th>Direction2</th>
      <th>Distance2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>29684</td>
      <td>30087</td>
      <td>NR_036051</td>
      <td>MIR1302-2</td>
      <td>+</td>
      <td>279</td>
      <td>NR_024540</td>
      <td>WASH7P</td>
      <td>-</td>
      <td>-315</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>36239</td>
      <td>38107</td>
      <td>NR_026818</td>
      <td>FAM138A</td>
      <td>-</td>
      <td>-159</td>
      <td>NR_036051</td>
      <td>MIR1302-2</td>
      <td>+</td>
      <td>-5737</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>198893</td>
      <td>201208</td>
      <td>NR_026823</td>
      <td>FAM138D</td>
      <td>-</td>
      <td>3921</td>
      <td>NR_107063</td>
      <td>MIR6859-3</td>
      <td>-</td>
      <td>-10936</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>203351</td>
      <td>207161</td>
      <td>NR_026823</td>
      <td>FAM138D</td>
      <td>-</td>
      <td>0</td>
      <td>NR_107063</td>
      <td>MIR6859-3</td>
      <td>-</td>
      <td>-15394</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>265549</td>
      <td>266336</td>
      <td>NR_026823</td>
      <td>FAM138D</td>
      <td>-</td>
      <td>-58953</td>
      <td>NR_107063</td>
      <td>MIR6859-3</td>
      <td>-</td>
      <td>-77592</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>10445</th>
      <td>chrY</td>
      <td>15158250</td>
      <td>15158653</td>
      <td>NM_001206850</td>
      <td>NLGN4Y</td>
      <td>+</td>
      <td>-314283</td>
      <td>NR_046504</td>
      <td>NLGN4Y-AS1</td>
      <td>-</td>
      <td>-354218</td>
    </tr>
    <tr>
      <th>10446</th>
      <td>chrY</td>
      <td>16985442</td>
      <td>16985845</td>
      <td>NR_028083</td>
      <td>FAM41AY1</td>
      <td>+</td>
      <td>515113</td>
      <td>NR_002160</td>
      <td>FAM224B</td>
      <td>-</td>
      <td>589316</td>
    </tr>
    <tr>
      <th>10447</th>
      <td>chrY</td>
      <td>19753311</td>
      <td>19753714</td>
      <td>NM_001146706</td>
      <td>KDM5D</td>
      <td>-</td>
      <td>-8373</td>
      <td>NR_045128</td>
      <td>TXLNGY</td>
      <td>+</td>
      <td>-146142</td>
    </tr>
    <tr>
      <th>10448</th>
      <td>chrY</td>
      <td>21011133</td>
      <td>21011828</td>
      <td>NM_001039567</td>
      <td>RPS4Y2</td>
      <td>+</td>
      <td>-230102</td>
      <td>NM_001282471</td>
      <td>PRORY</td>
      <td>-</td>
      <td>371146</td>
    </tr>
    <tr>
      <th>10449</th>
      <td>chrY</td>
      <td>56952574</td>
      <td>56957328</td>
      <td>NM_005840</td>
      <td>SPRY3</td>
      <td>+</td>
      <td>0</td>
      <td>NM_001145149</td>
      <td>VAMP7</td>
      <td>+</td>
      <td>110472</td>
    </tr>
  </tbody>
</table>
<p>10450 rows × 11 columns</p>
</div></div>
</div>
<p>Combine the two experiment qbed files to make anndata object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">exp_qbed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">K562_SP1</span><span class="p">,</span><span class="n">HCT116_SP1</span><span class="p">])</span>
<span class="n">exp_qbed</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>16529</td>
      <td>16533</td>
      <td>163</td>
      <td>-</td>
      <td>GCTCCTAAGTACGTTC-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>10</td>
      <td>+</td>
      <td>CTCACACCAGACGCTC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>155</td>
      <td>+</td>
      <td>TGGCCAGCACCCATTC-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>285</td>
      <td>+</td>
      <td>GTGGGTCCACGGCCAT-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>29884</td>
      <td>29888</td>
      <td>7</td>
      <td>+</td>
      <td>CGTCTACTCAACACGT-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77205</th>
      <td>chrY</td>
      <td>25518788</td>
      <td>25518792</td>
      <td>2</td>
      <td>+</td>
      <td>TGGGCGTTCGAACGGA-1</td>
    </tr>
    <tr>
      <th>77206</th>
      <td>chrY</td>
      <td>56987633</td>
      <td>56987637</td>
      <td>13</td>
      <td>+</td>
      <td>CAGTCCTAGGCACATG-1</td>
    </tr>
    <tr>
      <th>77207</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>17</td>
      <td>+</td>
      <td>CGGAGCTCATCGACGC-1</td>
    </tr>
    <tr>
      <th>77208</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>7</td>
      <td>+</td>
      <td>GTAACGTAGTTACGGG-1</td>
    </tr>
    <tr>
      <th>77209</th>
      <td>chrY</td>
      <td>57080855</td>
      <td>57080859</td>
      <td>9</td>
      <td>+</td>
      <td>TCAGCAAGTTGAACTC-1</td>
    </tr>
  </tbody>
</table>
<p>404675 rows × 6 columns</p>
</div></div>
</div>
<p>Read the barcode file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">barcodes</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">SP1_K562HCT116_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;barcodes&quot;</span><span class="p">)</span>
<span class="n">barcodes</span> <span class="o">=</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Index&#39;</span><span class="p">])</span>
<span class="n">barcodes</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Index</th>
      <th>cluster</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AAACCTGAGAAAGTGG-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AAACCTGAGACCGGAT-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAACCTGAGACTAGAT-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAACCTGAGAGCTTCT-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AAACCTGAGAGTACCG-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>52206</th>
      <td>TTTGTCATCTCCGGTT-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>52207</th>
      <td>TTTGTCATCTCGATGA-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>52208</th>
      <td>TTTGTCATCTCTAAGG-1</td>
      <td>K562</td>
    </tr>
    <tr>
      <th>52209</th>
      <td>TTTGTCATCTGGAGCC-1</td>
      <td>HCT116</td>
    </tr>
    <tr>
      <th>52210</th>
      <td>TTTGTCATCTTGGGTA-1</td>
      <td>HCT116</td>
    </tr>
  </tbody>
</table>
<p>51079 rows × 2 columns</p>
</div></div>
</div>
<p>Now we will connect the peaks (and insertions under the peaks) to the cell barcode data. To do so, we will use the qbed data, peak data and barcodes data to make a cell by peak anndata object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">make_Anndata</span><span class="p">(</span><span class="n">exp_qbed</span><span class="p">,</span> <span class="n">peak_annotation</span><span class="p">,</span> <span class="n">barcodes</span><span class="p">)</span>
<span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 24/24 [00:02&lt;00:00, 10.77it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 51079 × 10450
    obs: &#39;cluster&#39;
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;
</pre></div></div>
</div>
<p>Although one peak should have many insertions, there is a chance that all the cells from the peak were filtered by the RNA preprocesssing. In this case, we advise to filter the peaks. Additionally, we also recommend to filter cells that have very few insertions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_peaks</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">min_counts</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_peaks</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">min_cells</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>
<span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 51079 × 10448
    obs: &#39;cluster&#39;
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;, &#39;n_counts&#39;, &#39;n_cells&#39;
</pre></div></div>
</div>
<p>Next we can perform differential peak analysis to determine which peaks are cell type specific. In this example, we use the <em>fisher exact test</em> to find peaks enriched in K562 versus Hct116 cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="s2">&quot;cluster&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2/2 [00:55&lt;00:00, 27.99s/it]
</pre></div></div>
</div>
<p>We can plot the results for differential peak analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;fisher_exact&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_58_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_58_0.png" />
</div>
</div>
<p>Now let’s visualize some peaks that are differentially bound. The colored ones are the insertions for the cluster of interest (i.e. cell type) and the grey ones are insertions in the rest of the clusters. In this case there are only two clusters, HCT116 and K562. We observe large differences in Sp1 binding in HCT116 and K562 cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bg_qbed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">K562_brd4</span><span class="p">,</span> <span class="n">HCT116_brd4</span><span class="p">])</span>
<span class="n">bg_qbed</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>30238</td>
      <td>30242</td>
      <td>3</td>
      <td>+</td>
      <td>TTTACTGCATAAAGGT-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>2</td>
      <td>-</td>
      <td>ATCACGAAGAGTAATC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>30355</td>
      <td>30359</td>
      <td>70</td>
      <td>+</td>
      <td>TTGAACGCAAATCCGT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>31101</td>
      <td>31105</td>
      <td>2</td>
      <td>+</td>
      <td>CCTCAGTCATCAGTAC-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>32116</td>
      <td>32120</td>
      <td>5</td>
      <td>+</td>
      <td>CTAGTGAAGACAAAGG-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>37769</th>
      <td>chrY</td>
      <td>18037315</td>
      <td>18037319</td>
      <td>9</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37770</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>168</td>
      <td>+</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37771</th>
      <td>chrY</td>
      <td>24036504</td>
      <td>24036508</td>
      <td>508</td>
      <td>+</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
    <tr>
      <th>37772</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>13</td>
      <td>-</td>
      <td>GCAGTTAAGATCTGAA-1</td>
    </tr>
    <tr>
      <th>37773</th>
      <td>chrY</td>
      <td>25633622</td>
      <td>25633626</td>
      <td>32</td>
      <td>-</td>
      <td>CATATGGCAGCCAGAA-1</td>
    </tr>
  </tbody>
</table>
<p>145159 rows × 6 columns</p>
</div></div>
</div>
<p>In the tracks above, we see a strong peak on Chr 15 in HCT116 cell (purple) that is not present in K562 cells (red)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr7&quot;</span><span class="p">,</span> <span class="mi">143706539</span><span class="p">,</span> <span class="mi">143718962</span><span class="p">,</span> <span class="mi">100000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">exp_qbed</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;K562&quot;</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">13</span><span class="p">),</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;K562 Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;K562 Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;chr7_143706539_143718962_K562&quot;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr7&quot;</span><span class="p">,</span><span class="mi">143706539</span><span class="p">,</span><span class="mi">143718962</span><span class="p">,</span><span class="mi">100000</span><span class="p">,</span><span class="n">peak_data</span><span class="p">,</span><span class="n">exp_qbed</span><span class="p">,</span><span class="s2">&quot;hg38&quot;</span><span class="p">,</span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;HCT116&quot;</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">13</span><span class="p">),</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;HCT116 Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;HCT116 Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;purple&quot;</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;chr7_143706539_143718962_HCT116&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_62_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_62_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_62_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_62_1.png" />
</div>
</div>
<p>In the tracks above we see a number of Sp1 peaks on chr7 that are tightly bound in K562, but not in HCT116.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr8&quot;</span><span class="p">,</span> <span class="mi">17799370</span><span class="p">,</span> <span class="mi">17802353</span><span class="p">,</span> <span class="mi">100000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">exp_qbed</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;K562&quot;</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">13</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;K562 Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;K562 Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;chr19_9818643_9820060&quot;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr8&quot;</span><span class="p">,</span> <span class="mi">17799370</span><span class="p">,</span> <span class="mi">17802353</span><span class="p">,</span> <span class="mi">100000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">exp_qbed</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">adata_cc</span><span class="p">,</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;HCT116&quot;</span><span class="p">,</span><span class="n">key</span> <span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">13</span><span class="p">),</span><span class="n">peak_line</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
                <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;HCT116 Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;HCT116 Insertion Density&#39;</span><span class="p">,</span>
                <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;purple&quot;</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;chr19_9818643_9820060&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_64_0.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_64_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_K562HCT116_SP1_64_1.png" src="../../_images/tutorials_notebooks_K562HCT116_SP1_64_1.png" />
</div>
</div>
<p>Here we find peaks on Chr8 that are bound in HCT116 but not in K562 cells.</p>
<p>Saved the file if needed.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;SP1_qbed.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="Brd4_bulk.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="Mouse_cortex_Example.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorial: Analyzing single cell calling cards collected from the mouse cortex</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js?v=0b546569"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../_static/scripts/furo.js?v=5fa4622c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>