<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Tutorial: Analyzing single cell calling cards collected from the mouse cortex" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pycallingcards.readthedocs.io/en/latest/tutorials/notebooks/Mouse_cortex_Example.html" />
<meta property="og:site_name" content="Pycallingcards" />
<meta property="og:description" content="In this tutorial, we will analyze single cell calling data that describes Brd4 binding in the mouse cortex. The dataset comes from Moudgil et al., Cell. (2020) and can be downloaded from GEO. In th..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="https://pycallingcards.readthedocs.io/en/latest/index.html/_images/social_previews/summary_tutorials_notebooks_Mouse_cortex_Example_1a8bebd3.png" />
<meta property="og:image:alt" content="In this tutorial, we will analyze single cell calling data that describes Brd4 binding in the mouse cortex. The dataset comes from Moudgil et al., Cell. (202..." />
<meta name="description" content="In this tutorial, we will analyze single cell calling data that describes Brd4 binding in the mouse cortex. The dataset comes from Moudgil et al., Cell. (2020) and can be downloaded from GEO. In th..." />
<meta name="twitter:card" content="summary_large_image" />
<link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Tutorial: K562 HCT116 SP1 single-cell calling cards data." href="K562HCT116_SP1.html" /><link rel="prev" title="Tutorials" href="../index.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.08.06 -->
        <title>Tutorial: Analyzing single cell calling cards collected from the mouse cortex - pycallingcards-tools</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">pycallingcards-tools</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Tutorial: Analyzing single cell calling cards collected from the mouse cortex</a></li>
<li class="toctree-l2"><a class="reference internal" href="K562HCT116_SP1.html">Tutorial: K562 HCT116 SP1 single-cell calling cards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="Brd4_bulk.html">Tutorial: Brd4 bindings for Male/Female mouse bulk calling cards data.</a></li>
<li class="toctree-l2"><a class="reference internal" href="SP1_bulk.html">Tutorial: SP1 bindings in Cre-driver mouse lines.</a></li>
<li class="toctree-l2"><a class="reference internal" href="yeast.html">Tutorial: Yeast TYE7 bulk calling cards data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User guide</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Datasets</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mousecortex_data.html">pycallingcards.datasets.mousecortex_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_K562HCT116_data.html">pycallingcards.datasets.SP1_K562HCT116_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.mouse_brd4_data.html">pycallingcards.datasets.mouse_brd4_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.datasets.SP1_Cre_data.html">pycallingcards.datasets.SP1_Cre_data</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/reading.html">Reading</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Reading</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_qbed.html">pycallingcards.reading.read_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_rna.html">pycallingcards.reading.read_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.read_barbodes.html">pycallingcards.reading.read_barbodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.combine_qbed.html">pycallingcards.reading.combine_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_qbed.html">pycallingcards.reading.save_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.reading.save_bed.html">pycallingcards.reading.save_bed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/preprocessing.html">Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Preprocessing</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.clean_qbed.html">pycallingcards.preprocessing.clean_qbed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.call_peaks.html">pycallingcards.preprocessing.call_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.combine_peaks.html">pycallingcards.preprocessing.combine_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.separate_peaks.html">pycallingcards.preprocessing.separate_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.annotation.html">pycallingcards.preprocessing.annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.make_Anndata.html">pycallingcards.preprocessing.make_Anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.filter_peaks.html">pycallingcards.preprocessing.filter_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.adata_insertions.html">pycallingcards.preprocessing.adata_insertions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.preprocessing.down_sample.html">pycallingcards.preprocessing.down_sample</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/tools.html">Tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of Tools</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.call_motif.html">pycallingcards.tools.call_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.compare_motif.html">pycallingcards.tools.compare_motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups.html">pycallingcards.tools.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups_mu.html">pycallingcards.tools.rank_peak_groups_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.rank_peak_groups_df.html">pycallingcards.tools.rank_peak_groups_df</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bygroup.html">pycallingcards.tools.diff2group_bygroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.diff2group_bysample.html">pycallingcards.tools.diff2group_bysample</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.DE_pvalue.html">pycallingcards.tools.DE_pvalue</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_bulk.html">pycallingcards.tools.pair_peak_gene_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc.html">pycallingcards.tools.pair_peak_gene_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.pair_peak_gene_sc_mu.html">pycallingcards.tools.pair_peak_gene_sc_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.liftover.html">pycallingcards.tools.liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.find_location.html">pycallingcards.tools.find_location</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.result_mapping.html">pycallingcards.tools.result_mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS.html">pycallingcards.tools.GWAS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS_adata_bulk.html">pycallingcards.tools.GWAS_adata_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.GWAS_adata_sc.html">pycallingcards.tools.GWAS_adata_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.footprint.html">pycallingcards.tools.footprint</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.tools.find_fastq.html">pycallingcards.tools.find_fastq</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of Plotting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area.html">pycallingcards.plotting.draw_area</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.draw_area_mu.html">pycallingcards.plotting.draw_area_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.whole_peaks.html">pycallingcards.plotting.whole_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.calculate_signal.html">pycallingcards.plotting.calculate_signal</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_plot.html">pycallingcards.plotting.signal_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.signal_heatmap.html">pycallingcards.plotting.signal_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.rank_peak_groups.html">pycallingcards.plotting.rank_peak_groups</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.plot_matched.html">pycallingcards.plotting.plot_matched</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.volcano_plot.html">pycallingcards.plotting.volcano_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.heatmap.html">pycallingcards.plotting.heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_bulk.html">pycallingcards.plotting.dotplot_bulk</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc.html">pycallingcards.plotting.dotplot_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.dotplot_sc_mu.html">pycallingcards.plotting.dotplot_sc_mu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.WashU_browser_url.html">pycallingcards.plotting.WashU_browser_url</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.GWAS_adata_sc.html">pycallingcards.plotting.GWAS_adata_sc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/pycallingcards.plotting.GWAS_adata_bulk.html">pycallingcards.plotting.GWAS_adata_bulk</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/pycallingcards">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../../_sources/tutorials/notebooks/Mouse_cortex_Example.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="Tutorial:-Analyzing-single-cell-calling-cards-collected-from-the-mouse-cortex">
<h1>Tutorial: Analyzing single cell calling cards collected from the mouse cortex<a class="headerlink" href="#Tutorial:-Analyzing-single-cell-calling-cards-collected-from-the-mouse-cortex" title="Permalink to this heading">¶</a></h1>
<p>In this tutorial, we will analyze single cell calling data that describes Brd4 binding in the mouse cortex. The dataset comes from <a class="reference external" href="https://doi.org/10.1016/j.cell.2020.06.037">Moudgil et al., Cell. (2020)</a> and can be downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE148448">GEO</a>.</p>
<p>In this tutorial, e will cover how to call peaks, annotatate them, perform a differential peak analysis, and pair peaks with genes. In this dataset, there are 111382 insertions and 35950 cells in total. However, many cells are filtered in scRNA-seq analysis. It uses Mudata for calling cards and RNA data. If you want to use Anndata only, please check <a class="reference external" href="https://nbviewer.org/github/The-Mitra-Lab/pycallingcards_data/blob/main/Mouse_cortex_Example_adata.ipynb">Github</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pycallingcards</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">mudata</span><span class="w"> </span><span class="kn">import</span> <span class="n">MuData</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">150</span>
</pre></div>
</div>
</div>
<p>We start by reading the qbed datafile. In this file, each row represents a Brd4-directed insertion and columns indicate the chromosome, start point and end point, reads number, the direction and cell barcode of each insertion. For example, the first row indicates one transposition is located on Chromosome 1, and starts from 3112541 and ends on 3112545. The reads number is 12 with the direction going from 3’ to 5’. The barcode of the cell is GATGAAAAGAGTTGGC-1. Note that the barcodes of cells
should be consistent with scRNA-seq data.</p>
<p>Use <code class="docutils literal notranslate"><span class="pre">cc.rd.read_qbed(filename)</span></code> to read your own qbed data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">qbed_data</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">mousecortex_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;qbed&quot;</span><span class="p">)</span>
<span class="n">qbed_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Reads</th>
      <th>Direction</th>
      <th>Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>3112541</td>
      <td>3112545</td>
      <td>12</td>
      <td>+</td>
      <td>GATGAAAAGAGTTGGC-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>3121337</td>
      <td>3121341</td>
      <td>6</td>
      <td>-</td>
      <td>CGATCGGCACATTTCT-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>3199281</td>
      <td>3199285</td>
      <td>7</td>
      <td>+</td>
      <td>GTCCTCATCTCCGGTT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3211433</td>
      <td>3211437</td>
      <td>22</td>
      <td>-</td>
      <td>CGAGAAGAGGAATCGC-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>3245859</td>
      <td>3245863</td>
      <td>149</td>
      <td>+</td>
      <td>TTTACTGCATCCGCGA-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>111377</th>
      <td>chrY</td>
      <td>90807968</td>
      <td>90807972</td>
      <td>200</td>
      <td>-</td>
      <td>ACGGAGAGTCGCATAT-1</td>
    </tr>
    <tr>
      <th>111378</th>
      <td>chrY</td>
      <td>90833531</td>
      <td>90833535</td>
      <td>51</td>
      <td>-</td>
      <td>TAGCCGGTCCTGTACC-1</td>
    </tr>
    <tr>
      <th>111379</th>
      <td>chrY</td>
      <td>90833600</td>
      <td>90833604</td>
      <td>13</td>
      <td>-</td>
      <td>TTGGCAAAGAATTGTG-1</td>
    </tr>
    <tr>
      <th>111380</th>
      <td>chrY</td>
      <td>90840262</td>
      <td>90840266</td>
      <td>8</td>
      <td>-</td>
      <td>GTGCATAGTACCAGTT-1</td>
    </tr>
    <tr>
      <th>111381</th>
      <td>chrY</td>
      <td>90840262</td>
      <td>90840266</td>
      <td>95</td>
      <td>+</td>
      <td>TTGTAGGTCGAATCCA-1</td>
    </tr>
  </tbody>
</table>
<p>111382 rows × 6 columns</p>
</div></div>
</div>
<p>We next need to call peaks in order to find Brd4 binding sites. Three different methods (<em>CCcaller, cc_tools, Blockify</em>) are available along with three different species (<em>hg38, mm10, sacCer3</em>).</p>
<p>In this setting, we will use <strong>CCcaller</strong> in <em>mouse</em>(‘mm10’) data. <em>maxbetween</em> is the most important parameter for <em>CCcaller</em>. It controls the maximum distance between two nearby insertions, or, in other words, the minimum distance between two peaks. 1000-2000 is a good parameter for <em>maxbetween</em>. <em>pvalue_cutoff</em> is also an important parameter, and a number between 0.001 and 0.05 is strongly advised. <em>pseudocounts</em> is advised to be 0.1-1.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_data</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">call_peaks</span><span class="p">(</span><span class="n">qbed_data</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;CCcaller&quot;</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span>  <span class="n">maxbetween</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
                             <span class="n">pvalue_cutoff</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">lam_win_size</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span>  <span class="n">pseudocounts</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">record</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                             <span class="n">save</span> <span class="o">=</span> <span class="s1">&#39;mouse_cortex.bed&#39;</span><span class="p">)</span>
<span class="n">peak_data</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
For the CCcaller method without background, [expdata, reference, pvalue_cutoff, lam_win_size, pseudocounts, minlen, extend, maxbetween, test_method, min_insertions, record] would be utilized.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 21/21 [00:16&lt;00:00,  1.24it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
      <th>pvalue_adj</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>4806673</td>
      <td>4809049</td>
      <td>12</td>
      <td>20</td>
      <td>1.120541</td>
      <td>2.498336e-10</td>
      <td>6.563055e-07</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>14302176</td>
      <td>14310895</td>
      <td>14</td>
      <td>92</td>
      <td>1.523252</td>
      <td>1.015569e-10</td>
      <td>2.845105e-07</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>15287495</td>
      <td>15288141</td>
      <td>8</td>
      <td>4</td>
      <td>1.029800</td>
      <td>1.427167e-06</td>
      <td>2.002598e-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>18307949</td>
      <td>18310271</td>
      <td>8</td>
      <td>31</td>
      <td>1.151983</td>
      <td>3.511291e-06</td>
      <td>4.395841e-03</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>18976012</td>
      <td>18982286</td>
      <td>13</td>
      <td>62</td>
      <td>1.452142</td>
      <td>5.507813e-10</td>
      <td>1.357847e-06</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>696</th>
      <td>chrX</td>
      <td>158919208</td>
      <td>158925514</td>
      <td>9</td>
      <td>39</td>
      <td>1.272189</td>
      <td>9.683526e-07</td>
      <td>1.413241e-03</td>
    </tr>
    <tr>
      <th>697</th>
      <td>chrX</td>
      <td>165325630</td>
      <td>165327490</td>
      <td>8</td>
      <td>17</td>
      <td>1.047701</td>
      <td>1.640240e-06</td>
      <td>2.254207e-03</td>
    </tr>
    <tr>
      <th>698</th>
      <td>chrX</td>
      <td>166241178</td>
      <td>166243587</td>
      <td>8</td>
      <td>18</td>
      <td>1.257844</td>
      <td>7.050309e-06</td>
      <td>8.151358e-03</td>
    </tr>
    <tr>
      <th>699</th>
      <td>chrX</td>
      <td>166345453</td>
      <td>166350005</td>
      <td>11</td>
      <td>35</td>
      <td>1.507674</td>
      <td>7.204418e-08</td>
      <td>1.263100e-04</td>
    </tr>
    <tr>
      <th>700</th>
      <td>chrX</td>
      <td>169842873</td>
      <td>169845831</td>
      <td>9</td>
      <td>28</td>
      <td>1.314399</td>
      <td>1.292113e-06</td>
      <td>1.842126e-03</td>
    </tr>
  </tbody>
</table>
<p>701 rows × 8 columns</p>
</div></div>
</div>
<p>In order to tune parameters for peak calling, we advise looking at the data and evaluating the validity of the called peaks. The default settings are recommended, but for some TFs, adjacent peaks may be merged that should not be, or, alternatively, peaks that should be joined may be called separately.</p>
<p>In this plot, the top section is insertions and their read counts. One dot is an insertion and the height is log(reads+1). The middle section is the distribution of insertions. The bottom section represents the reference genes and peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area</span><span class="p">(</span><span class="s2">&quot;chr12&quot;</span><span class="p">,</span> <span class="mi">50102917</span><span class="p">,</span> <span class="mi">50124960</span><span class="p">,</span> <span class="mi">400000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">qbed_data</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">],</span>
                <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;chr12_50102917_50124960&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_8_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_8_0.png" />
</div>
</div>
<p>We can also visualize our data directly in the <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a>. This can be useful for overlaying your data with other published datasets. Please note that this link only valid for 24hrs, so you will have to rerun it if you want to use it after this time period.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">qbed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;qbed_data&quot;</span><span class="p">:</span><span class="n">qbed_data</span><span class="p">}</span>
<span class="n">bed</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;peak&quot;</span><span class="p">:</span><span class="n">peak_data</span><span class="p">}</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">WashU_browser_url</span><span class="p">(</span><span class="n">qbed</span> <span class="o">=</span> <span class="n">qbed</span><span class="p">,</span> <span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span><span class="p">,</span> <span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;mm10&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All qbed addressed
All bed addressed
Uploading files
Please click the following link to see the data on WashU Epigenome Browser directly.
https://epigenomegateway.wustl.edu/browser/?genome=mm10&amp;hub=https://companion.epigenomegateway.org//task/ce3f3ef6f66e901b7fc00587e885fd07/output//datahub.json
</pre></div></div>
</div>
<p>Pycallingcards can be used to visual peak locations acorss the genome to see that the distribution of peaks is unbiased and that all chromosomes are represented.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">whole_peaks</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_12_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_12_0.png" />
</div>
</div>
<p>In the next step, we annotate the peaks by their closest genes using <a class="reference external" href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a>. Make sure these programs are properly installed before using.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_annotation</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">annotation</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span><span class="p">)</span>
<span class="n">peak_annotation</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">combine_annotation</span><span class="p">(</span><span class="n">peak_data</span><span class="p">,</span> <span class="n">peak_annotation</span><span class="p">)</span>
<span class="n">peak_annotation</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In the bedtools method, we would use bedtools in the default path. Set bedtools path by &#39;bedtools_path&#39; if needed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Experiment Insertions</th>
      <th>Reference Insertions</th>
      <th>Expected Insertions</th>
      <th>pvalue</th>
      <th>pvalue_adj</th>
      <th>Nearest Refseq1</th>
      <th>Gene Name1</th>
      <th>Direction1</th>
      <th>Distance1</th>
      <th>Nearest Refseq2</th>
      <th>Gene Name2</th>
      <th>Direction2</th>
      <th>Distance2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>4806673</td>
      <td>4809049</td>
      <td>12</td>
      <td>20</td>
      <td>1.120541</td>
      <td>2.498336e-10</td>
      <td>6.563055e-07</td>
      <td>NM_008866</td>
      <td>Lypla1</td>
      <td>+</td>
      <td>0</td>
      <td>NR_033530</td>
      <td>Mrpl15</td>
      <td>-</td>
      <td>-20948</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>14302176</td>
      <td>14310895</td>
      <td>14</td>
      <td>92</td>
      <td>1.523252</td>
      <td>1.015569e-10</td>
      <td>2.845105e-07</td>
      <td>NM_010164</td>
      <td>Eya1</td>
      <td>-</td>
      <td>0</td>
      <td>NM_010827</td>
      <td>Msc</td>
      <td>-</td>
      <td>442451</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>15287495</td>
      <td>15288141</td>
      <td>8</td>
      <td>4</td>
      <td>1.029800</td>
      <td>1.427167e-06</td>
      <td>2.002598e-03</td>
      <td>NM_001098528</td>
      <td>Kcnb2</td>
      <td>+</td>
      <td>24311</td>
      <td>NM_177781</td>
      <td>Trpa1</td>
      <td>-</td>
      <td>-368634</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>18307949</td>
      <td>18310271</td>
      <td>8</td>
      <td>31</td>
      <td>1.151983</td>
      <td>3.511291e-06</td>
      <td>4.395841e-03</td>
      <td>NM_183124</td>
      <td>Defb41</td>
      <td>-</td>
      <td>-42812</td>
      <td>NM_001039123</td>
      <td>Defb18</td>
      <td>-</td>
      <td>-70507</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>18976012</td>
      <td>18982286</td>
      <td>13</td>
      <td>62</td>
      <td>1.452142</td>
      <td>5.507813e-10</td>
      <td>1.357847e-06</td>
      <td>NM_153154</td>
      <td>Tfap2d</td>
      <td>+</td>
      <td>120736</td>
      <td>NM_001286340</td>
      <td>Tfap2b</td>
      <td>+</td>
      <td>226628</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>696</th>
      <td>chrX</td>
      <td>158919208</td>
      <td>158925514</td>
      <td>9</td>
      <td>39</td>
      <td>1.272189</td>
      <td>9.683526e-07</td>
      <td>1.413241e-03</td>
      <td>NM_001346675</td>
      <td>Rps6ka3</td>
      <td>+</td>
      <td>330268</td>
      <td>NM_025437</td>
      <td>Eif1ax</td>
      <td>+</td>
      <td>446681</td>
    </tr>
    <tr>
      <th>697</th>
      <td>chrX</td>
      <td>165325630</td>
      <td>165327490</td>
      <td>8</td>
      <td>17</td>
      <td>1.047701</td>
      <td>1.640240e-06</td>
      <td>2.254207e-03</td>
      <td>NM_183427</td>
      <td>Glra2</td>
      <td>-</td>
      <td>0</td>
      <td>NM_175027</td>
      <td>Fancb</td>
      <td>+</td>
      <td>-328359</td>
    </tr>
    <tr>
      <th>698</th>
      <td>chrX</td>
      <td>166241178</td>
      <td>166243587</td>
      <td>8</td>
      <td>18</td>
      <td>1.257844</td>
      <td>7.050309e-06</td>
      <td>8.151358e-03</td>
      <td>NM_023122</td>
      <td>Gpm6b</td>
      <td>+</td>
      <td>0</td>
      <td>NM_001310724</td>
      <td>Gemin8</td>
      <td>+</td>
      <td>-50667</td>
    </tr>
    <tr>
      <th>699</th>
      <td>chrX</td>
      <td>166345453</td>
      <td>166350005</td>
      <td>11</td>
      <td>35</td>
      <td>1.507674</td>
      <td>7.204418e-08</td>
      <td>1.263100e-04</td>
      <td>NM_023122</td>
      <td>Gpm6b</td>
      <td>+</td>
      <td>0</td>
      <td>NM_177429</td>
      <td>Ofd1</td>
      <td>-</td>
      <td>40028</td>
    </tr>
    <tr>
      <th>700</th>
      <td>chrX</td>
      <td>169842873</td>
      <td>169845831</td>
      <td>9</td>
      <td>28</td>
      <td>1.314399</td>
      <td>1.292113e-06</td>
      <td>1.842126e-03</td>
      <td>NM_010797</td>
      <td>Mid1</td>
      <td>+</td>
      <td>0</td>
      <td>NM_001290506</td>
      <td>Mid1</td>
      <td>+</td>
      <td>33788</td>
    </tr>
  </tbody>
</table>
<p>701 rows × 16 columns</p>
</div></div>
</div>
<p>Then, we read the barcode file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">barcodes</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">mousecortex_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;barcodes&quot;</span><span class="p">)</span>
<span class="n">barcodes</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AAACCTGAGAACTCGG-1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AAACCTGAGCAATCTC-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAACCTGAGCCGTCGT-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAACCTGAGTAGCGGT-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AAACCTGAGTGGAGTC-1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>35945</th>
      <td>TTTGTCAAGTCCCACG-1</td>
    </tr>
    <tr>
      <th>35946</th>
      <td>TTTGTCACAGCGTCCA-1</td>
    </tr>
    <tr>
      <th>35947</th>
      <td>TTTGTCACATTTCACT-1</td>
    </tr>
    <tr>
      <th>35948</th>
      <td>TTTGTCAGTCGCATCG-1</td>
    </tr>
    <tr>
      <th>35949</th>
      <td>TTTGTCATCTTTACAC-1</td>
    </tr>
  </tbody>
</table>
<p>35950 rows × 1 columns</p>
</div></div>
</div>
<p>Use qbed data, peak data and barcode data to make a cell by peak <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">Anndata</a> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">make_Anndata</span><span class="p">(</span><span class="n">qbed_data</span><span class="p">,</span> <span class="n">peak_annotation</span><span class="p">,</span> <span class="n">barcodes</span><span class="p">)</span>
<span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 20/20 [00:00&lt;00:00, 89.37it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 35950 × 701
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;pvalue_adj&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;
</pre></div></div>
</div>
<p>In single cell calling cards, for each cell, RNA-seq data is collected simultaneously with TF binding information. For the following steps, we are going to read scRNA-seq data and analyze them together. <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a> is recommended to load and analyze scRNA-seq data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">mousecortex_data</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/miniconda3/lib/python3.9/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 30300 × 2638
    obs: &#39;batch&#39;, &#39;n_genes&#39;, &#39;total_counts&#39;, &#39;cluster&#39;
    var: &#39;n_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;
</pre></div></div>
</div>
<p>In scRNA-seq analysis, many cells are filtered out because of low quality. We need to make the cells in qbed anndata to be the exactly same as RNA-seq anndata.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_cc</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_adata_sc</span><span class="p">(</span><span class="n">adata_cc</span><span class="p">,</span> <span class="n">adata</span><span class="p">)</span>
<span class="n">adata_cc</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
View of AnnData object with n_obs × n_vars = 30300 × 701
    var: &#39;Chr&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Experiment Insertions&#39;, &#39;Reference Insertions&#39;, &#39;Expected Insertions&#39;, &#39;pvalue&#39;, &#39;pvalue_adj&#39;, &#39;Nearest Refseq1&#39;, &#39;Gene Name1&#39;, &#39;Direction1&#39;, &#39;Distance1&#39;, &#39;Nearest Refseq2&#39;, &#39;Gene Name2&#39;, &#39;Direction2&#39;, &#39;Distance2&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;RNA&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;CC&quot;</span><span class="p">:</span> <span class="n">adata_cc</span><span class="p">})</span>
<span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 30300 × 3339
  2 modalities
    RNA:    30300 x 2638
      obs:  &#x27;batch&#x27;, &#x27;n_genes&#x27;, &#x27;total_counts&#x27;, &#x27;cluster&#x27;
      var:  &#x27;n_counts&#x27;, &#x27;n_cells&#x27;, &#x27;highly_variable&#x27;
    CC:     30300 x 701
      var:  &#x27;Chr&#x27;, &#x27;Start&#x27;, &#x27;End&#x27;, &#x27;Experiment Insertions&#x27;, &#x27;Reference Insertions&#x27;, &#x27;Expected Insertions&#x27;, &#x27;pvalue&#x27;, &#x27;pvalue_adj&#x27;, &#x27;Nearest Refseq1&#x27;, &#x27;Gene Name1&#x27;, &#x27;Direction1&#x27;, &#x27;Distance1&#x27;, &#x27;Nearest Refseq2&#x27;, &#x27;Gene Name2&#x27;, &#x27;Direction2&#x27;, &#x27;Distance2&#x27;</pre></div>
</div>
<p>Next, we will cluster cells by the RNA-seq data to identify cell types so that we can identify differentially bound peaks between different cell types.</p>
<p>Although one peak should have many insertions, but there is a chance that all the cells from the peak were filtered by the RNA preprocesssing. In this case, we advise to filter peaks by the minimum number of cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_peaks</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">],</span> <span class="n">min_counts</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/miniconda3/lib/python3.9/site-packages/scanpy/preprocessing/_simple.py:248: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.
  adata.var[&#39;n_counts&#39;] = number
</pre></div></div>
</div>
<p>Differential peak analysis shows significant bindings for each cluster. In this example, we use <em>binomial test</em> to find out.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups_mu</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;RNA:cluster&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">)</span>
<span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 18/18 [00:21&lt;00:00,  1.19s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 30300 × 3339
  2 modalities
    RNA:    30300 x 2638
      obs:  &#x27;batch&#x27;, &#x27;n_genes&#x27;, &#x27;total_counts&#x27;, &#x27;cluster&#x27;
      var:  &#x27;n_counts&#x27;, &#x27;n_cells&#x27;, &#x27;highly_variable&#x27;
    CC:     30300 x 701
      var:  &#x27;Chr&#x27;, &#x27;Start&#x27;, &#x27;End&#x27;, &#x27;Experiment Insertions&#x27;, &#x27;Reference Insertions&#x27;, &#x27;Expected Insertions&#x27;, &#x27;pvalue&#x27;, &#x27;pvalue_adj&#x27;, &#x27;Nearest Refseq1&#x27;, &#x27;Gene Name1&#x27;, &#x27;Direction1&#x27;, &#x27;Distance1&#x27;, &#x27;Nearest Refseq2&#x27;, &#x27;Gene Name2&#x27;, &#x27;Direction2&#x27;, &#x27;Distance2&#x27;, &#x27;n_counts&#x27;
      uns:  &#x27;binomtest&#x27;</pre></div>
</div>
<p>Plot the results for differential peak analysis.</p>
<p>Currently, the peaks are ranked by pvalues. It could also be ranked by logfoldchanges by the following code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_peak_groups_mu</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;RNA:cluster&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">,</span> <span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
<span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">],</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">,</span> <span class="n">rankby</span> <span class="o">=</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_peak_groups</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">],</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;binomtest&#39;</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_30_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_30_0.png" />
</div>
</div>
<p>Next, we will visualize differentially bound peaks. The colored datapoints are the insertions specific to a cluster and the grey ones are the total insertions across the rest of the clusters. We can see that most of the insertions are from Astrocyte in the following peaks.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_area_mu</span><span class="p">(</span><span class="s2">&quot;chr3&quot;</span><span class="p">,</span> <span class="mi">34638588</span><span class="p">,</span> <span class="mi">34656047</span><span class="p">,</span> <span class="mi">400000</span><span class="p">,</span> <span class="n">peak_data</span><span class="p">,</span> <span class="n">qbed_data</span><span class="p">,</span> <span class="s2">&quot;mm10&quot;</span><span class="p">,</span> <span class="n">mdata</span> <span class="o">=</span> <span class="n">mdata</span><span class="p">,</span> <span class="n">font_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
                   <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Astrocyte&#39;</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;RNA:cluster&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">12</span><span class="p">),</span> <span class="n">peak_line</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span>
                   <span class="n">name_insertion1</span> <span class="o">=</span> <span class="s1">&#39;Astrocyte Insertions&#39;</span><span class="p">,</span> <span class="n">name_density1</span> <span class="o">=</span> <span class="s1">&#39;Astrocyte Insertion Density&#39;</span><span class="p">,</span>
                   <span class="n">name_insertion2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertions&#39;</span><span class="p">,</span> <span class="n">name_density2</span> <span class="o">=</span> <span class="s1">&#39;Total Insertion Density&#39;</span><span class="p">,</span>
                   <span class="n">plotsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;chr3_34638588_34656047&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_32_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_32_0.png" />
</div>
</div>
<p>Next we can ask if differentially bound peaks are near differentially expressed genes, which might suggest the idenified enhancer regulates the nearby gene.</p>
<p>We first perform a differential expression analysis for the scRNA-seq data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">],</span> <span class="s1">&#39;cluster&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Default of the method has been changed to &#39;t-test&#39; from &#39;t-test_overestim_var&#39;
</pre></div></div>
</div>
<p>Next, we find peak-gene pairs that are differentially bound/regulated in the specified cell-type. We look at all differential peaks in each cluster and see if the annotated genes are significantly expressed in the cluster. We can then set the pvalue and score/log foldchange cutoff.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pair_peak_gene_sc_mu</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">pvalue_adj_cutoff_cc</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">pvalue_adj_cutoff_rna</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
                           <span class="n">lfc_cutoff</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">score_cutoff</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;pair&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cluster</th>
      <th>Peak</th>
      <th>Logfoldchanges</th>
      <th>Pvalue_peak</th>
      <th>Pvalue_adj_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
      <th>Pvalue_adj_gene</th>
      <th>Distance_peak_to_gene</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Astrocyte</td>
      <td>chr16_43501178_43518253</td>
      <td>3.077262</td>
      <td>1.462410e-16</td>
      <td>2.562874e-14</td>
      <td>Zbtb20</td>
      <td>79.234276</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Astrocyte</td>
      <td>chr8_64645834_64659215</td>
      <td>4.623955</td>
      <td>3.114365e-14</td>
      <td>2.949104e-12</td>
      <td>Msmo1</td>
      <td>61.964111</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>58930</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Astrocyte</td>
      <td>chr3_141928409_141939733</td>
      <td>4.876938</td>
      <td>3.786296e-14</td>
      <td>2.949104e-12</td>
      <td>Bmpr1b</td>
      <td>24.131538</td>
      <td>2.475603e-121</td>
      <td>1.659753e-120</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>4.140867e-12</td>
      <td>Nfia</td>
      <td>45.417362</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>188838</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>4.140867e-12</td>
      <td>E130114P18Rik</td>
      <td>16.893749</td>
      <td>3.380841e-62</td>
      <td>1.514206e-61</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>57</th>
      <td>Neuron_Excit_L2-4</td>
      <td>chr13_83141353_83148478</td>
      <td>3.495487</td>
      <td>3.141216e-06</td>
      <td>2.446659e-04</td>
      <td>Mef2c</td>
      <td>182.292236</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>355556</td>
    </tr>
    <tr>
      <th>58</th>
      <td>Neuron_Excit_L5_Mixed</td>
      <td>chr7_66014128_66014532</td>
      <td>3.737143</td>
      <td>2.786909e-04</td>
      <td>2.858012e-02</td>
      <td>Pcsk6</td>
      <td>-16.587570</td>
      <td>2.263189e-61</td>
      <td>3.313281e-60</td>
      <td>0</td>
    </tr>
    <tr>
      <th>59</th>
      <td>Neuron_Granule_DG</td>
      <td>chr3_125405334_125410657</td>
      <td>5.806064</td>
      <td>1.011677e-04</td>
      <td>1.251507e-02</td>
      <td>Ugt8a</td>
      <td>-23.040115</td>
      <td>5.201659e-89</td>
      <td>3.379286e-87</td>
      <td>454686</td>
    </tr>
    <tr>
      <th>60</th>
      <td>Neuron_Granule_DG</td>
      <td>chr3_125405334_125410657</td>
      <td>5.806064</td>
      <td>1.011677e-04</td>
      <td>1.251507e-02</td>
      <td>Ndst4</td>
      <td>4.497158</td>
      <td>8.449939e-06</td>
      <td>3.032315e-05</td>
      <td>0</td>
    </tr>
    <tr>
      <th>61</th>
      <td>Neuron_Granule_DG</td>
      <td>chr1_50860999_50861403</td>
      <td>4.356917</td>
      <td>1.071190e-04</td>
      <td>1.251507e-02</td>
      <td>Tmeff2</td>
      <td>3.399986</td>
      <td>7.231398e-04</td>
      <td>1.952972e-03</td>
      <td>66120</td>
    </tr>
  </tbody>
</table>
<p>62 rows × 10 columns</p>
</div></div>
</div>
<p>Plot the results above to find out the potential relationship between TF bindings and gene expression.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot_sc_mu</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">60</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_39_0.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_39_0.png" />
</div>
</div>
<p>After seeing the dotplot above, bring some peaks to the RNA Anndata object and see the distribution in the UMAP plot.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>The first plot is the average insertions of the peak in each cluster, the second plot is the gene expression Gou3f2 (one of its nearest genes for the peak); the third plot is the cluster information.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_matched</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s1">&#39;chr4_22969921_22973019&#39;</span><span class="p">,</span> <span class="s1">&#39;Pou3f2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/miniconda3/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:391: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_43_1.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_43_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_matched</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s1">&#39;chr1_42253068_42265192&#39;</span><span class="p">,</span> <span class="s1">&#39;Pou3f3&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The history saving thread hit an unexpected error (OperationalError(&#39;database is locked&#39;)).History will not be written to the database.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/miniconda3/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:391: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Mouse_cortex_Example_44_2.png" src="../../_images/tutorials_notebooks_Mouse_cortex_Example_44_2.png" />
</div>
</div>
<p>We can see a potential relationship between Brd4 binding and gene expression.</p>
<p>If we map these enhancers to the human genome, are there disease associated SNPs nearby? To answer this question, we can map binding sites and genes to the human genome. We use <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">liftover</a> to get the resuls.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pair&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">result_mapping</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pair&quot;</span><span class="p">])</span>
<span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pair&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Start mapping the peaks to the new genome.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 62/62 [00:00&lt;00:00, 222.90it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Start finding location of genes in the new genome.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 62/62 [00:00&lt;00:00, 257.11it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cluster</th>
      <th>Peak</th>
      <th>Logfoldchanges</th>
      <th>Pvalue_peak</th>
      <th>Pvalue_adj_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
      <th>Pvalue_adj_gene</th>
      <th>Distance_peak_to_gene</th>
      <th>Chr_liftover</th>
      <th>Start_liftover</th>
      <th>End_liftover</th>
      <th>Chr_hg38</th>
      <th>Start_hg38</th>
      <th>End_hg38</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Astrocyte</td>
      <td>chr16_43501178_43518253</td>
      <td>3.077262</td>
      <td>1.462410e-16</td>
      <td>2.562874e-14</td>
      <td>Zbtb20</td>
      <td>79.234276</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>0</td>
      <td>chr3</td>
      <td>114439800</td>
      <td>114457200</td>
      <td>chr3</td>
      <td>114314499</td>
      <td>115147280</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Astrocyte</td>
      <td>chr8_64645834_64659215</td>
      <td>4.623955</td>
      <td>3.114365e-14</td>
      <td>2.949104e-12</td>
      <td>Msmo1</td>
      <td>61.964111</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>58930</td>
      <td>chr4</td>
      <td>165418445</td>
      <td>165438029</td>
      <td>chr4</td>
      <td>165327665</td>
      <td>165343162</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Astrocyte</td>
      <td>chr3_141928409_141939733</td>
      <td>4.876938</td>
      <td>3.786296e-14</td>
      <td>2.949104e-12</td>
      <td>Bmpr1b</td>
      <td>24.131538</td>
      <td>2.475603e-121</td>
      <td>1.659753e-120</td>
      <td>0</td>
      <td>chr4</td>
      <td>95040417</td>
      <td>95054959</td>
      <td>chr4</td>
      <td>94757976</td>
      <td>95158450</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>4.140867e-12</td>
      <td>Nfia</td>
      <td>45.417362</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>188838</td>
      <td>chr1</td>
      <td>60858298</td>
      <td>60872361</td>
      <td>chr1</td>
      <td>61077273</td>
      <td>61462788</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>4.140867e-12</td>
      <td>E130114P18Rik</td>
      <td>16.893749</td>
      <td>3.380841e-62</td>
      <td>1.514206e-61</td>
      <td>0</td>
      <td>chr1</td>
      <td>60858298</td>
      <td>60872361</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>57</th>
      <td>Neuron_Excit_L2-4</td>
      <td>chr13_83141353_83148478</td>
      <td>3.495487</td>
      <td>3.141216e-06</td>
      <td>2.446659e-04</td>
      <td>Mef2c</td>
      <td>182.292236</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>355556</td>
      <td>chr5</td>
      <td>89235156</td>
      <td>89249323</td>
      <td>chr5</td>
      <td>88718240</td>
      <td>88904105</td>
    </tr>
    <tr>
      <th>58</th>
      <td>Neuron_Excit_L5_Mixed</td>
      <td>chr7_66014128_66014532</td>
      <td>3.737143</td>
      <td>2.786909e-04</td>
      <td>2.858012e-02</td>
      <td>Pcsk6</td>
      <td>-16.587570</td>
      <td>2.263189e-61</td>
      <td>3.313281e-60</td>
      <td>0</td>
      <td>chr15</td>
      <td>101337048</td>
      <td>101337565</td>
      <td>chr15</td>
      <td>101303927</td>
      <td>101489984</td>
    </tr>
    <tr>
      <th>59</th>
      <td>Neuron_Granule_DG</td>
      <td>chr3_125405334_125410657</td>
      <td>5.806064</td>
      <td>1.011677e-04</td>
      <td>1.251507e-02</td>
      <td>Ugt8a</td>
      <td>-23.040115</td>
      <td>5.201659e-89</td>
      <td>3.379286e-87</td>
      <td>454686</td>
      <td>chr4</td>
      <td>115106106</td>
      <td>115112630</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>60</th>
      <td>Neuron_Granule_DG</td>
      <td>chr3_125405334_125410657</td>
      <td>5.806064</td>
      <td>1.011677e-04</td>
      <td>1.251507e-02</td>
      <td>Ndst4</td>
      <td>4.497158</td>
      <td>8.449939e-06</td>
      <td>3.032315e-05</td>
      <td>0</td>
      <td>chr4</td>
      <td>115106106</td>
      <td>115112630</td>
      <td>chr4</td>
      <td>114827772</td>
      <td>115113876</td>
    </tr>
    <tr>
      <th>61</th>
      <td>Neuron_Granule_DG</td>
      <td>chr1_50860999_50861403</td>
      <td>4.356917</td>
      <td>1.071190e-04</td>
      <td>1.251507e-02</td>
      <td>Tmeff2</td>
      <td>3.399986</td>
      <td>7.231398e-04</td>
      <td>1.952972e-03</td>
      <td>66120</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr2</td>
      <td>191949045</td>
      <td>192194933</td>
    </tr>
  </tbody>
</table>
<p>62 rows × 16 columns</p>
</div></div>
</div>
<p>We search the <a class="reference external" href="https://www.ebi.ac.uk/gwas/docs/file-downloads">GWAS Catalog database</a> and find out related SNPs in the binding areas.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pair&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">GWAS</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pair&quot;</span><span class="p">])</span>
<span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;CC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pair&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cluster</th>
      <th>Peak</th>
      <th>Logfoldchanges</th>
      <th>Pvalue_peak</th>
      <th>Pvalue_adj_peak</th>
      <th>Gene</th>
      <th>Score_gene</th>
      <th>Pvalue_gene</th>
      <th>Pvalue_adj_gene</th>
      <th>Distance_peak_to_gene</th>
      <th>Chr_liftover</th>
      <th>Start_liftover</th>
      <th>End_liftover</th>
      <th>Chr_hg38</th>
      <th>Start_hg38</th>
      <th>End_hg38</th>
      <th>GWAS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Astrocyte</td>
      <td>chr16_43501178_43518253</td>
      <td>3.077262</td>
      <td>1.462410e-16</td>
      <td>2.562874e-14</td>
      <td>Zbtb20</td>
      <td>79.234276</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>0</td>
      <td>chr3</td>
      <td>114439800</td>
      <td>114457200</td>
      <td>chr3</td>
      <td>114314499</td>
      <td>115147280</td>
      <td>Schizophrenia; Smoking status (ever vs never s...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Astrocyte</td>
      <td>chr8_64645834_64659215</td>
      <td>4.623955</td>
      <td>3.114365e-14</td>
      <td>2.949104e-12</td>
      <td>Msmo1</td>
      <td>61.964111</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>58930</td>
      <td>chr4</td>
      <td>165418445</td>
      <td>165438029</td>
      <td>chr4</td>
      <td>165327665</td>
      <td>165343162</td>
      <td>Atopic dermatitis (moderate to severe)</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Astrocyte</td>
      <td>chr3_141928409_141939733</td>
      <td>4.876938</td>
      <td>3.786296e-14</td>
      <td>2.949104e-12</td>
      <td>Bmpr1b</td>
      <td>24.131538</td>
      <td>2.475603e-121</td>
      <td>1.659753e-120</td>
      <td>0</td>
      <td>chr4</td>
      <td>95040417</td>
      <td>95054959</td>
      <td>chr4</td>
      <td>94757976</td>
      <td>95158450</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>4.140867e-12</td>
      <td>Nfia</td>
      <td>45.417362</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>188838</td>
      <td>chr1</td>
      <td>60858298</td>
      <td>60872361</td>
      <td>chr1</td>
      <td>61077273</td>
      <td>61462788</td>
      <td>Refractive error</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Astrocyte</td>
      <td>chr4_97575305_97588788</td>
      <td>3.099679</td>
      <td>5.907086e-14</td>
      <td>4.140867e-12</td>
      <td>E130114P18Rik</td>
      <td>16.893749</td>
      <td>3.380841e-62</td>
      <td>1.514206e-61</td>
      <td>0</td>
      <td>chr1</td>
      <td>60858298</td>
      <td>60872361</td>
      <td></td>
      <td></td>
      <td></td>
      <td>Refractive error</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>57</th>
      <td>Neuron_Excit_L2-4</td>
      <td>chr13_83141353_83148478</td>
      <td>3.495487</td>
      <td>3.141216e-06</td>
      <td>2.446659e-04</td>
      <td>Mef2c</td>
      <td>182.292236</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>355556</td>
      <td>chr5</td>
      <td>89235156</td>
      <td>89249323</td>
      <td>chr5</td>
      <td>88718240</td>
      <td>88904105</td>
      <td>Macular thickness; Waist circumference adjuste...</td>
    </tr>
    <tr>
      <th>58</th>
      <td>Neuron_Excit_L5_Mixed</td>
      <td>chr7_66014128_66014532</td>
      <td>3.737143</td>
      <td>2.786909e-04</td>
      <td>2.858012e-02</td>
      <td>Pcsk6</td>
      <td>-16.587570</td>
      <td>2.263189e-61</td>
      <td>3.313281e-60</td>
      <td>0</td>
      <td>chr15</td>
      <td>101337048</td>
      <td>101337565</td>
      <td>chr15</td>
      <td>101303927</td>
      <td>101489984</td>
      <td></td>
    </tr>
    <tr>
      <th>59</th>
      <td>Neuron_Granule_DG</td>
      <td>chr3_125405334_125410657</td>
      <td>5.806064</td>
      <td>1.011677e-04</td>
      <td>1.251507e-02</td>
      <td>Ugt8a</td>
      <td>-23.040115</td>
      <td>5.201659e-89</td>
      <td>3.379286e-87</td>
      <td>454686</td>
      <td>chr4</td>
      <td>115106106</td>
      <td>115112630</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>60</th>
      <td>Neuron_Granule_DG</td>
      <td>chr3_125405334_125410657</td>
      <td>5.806064</td>
      <td>1.011677e-04</td>
      <td>1.251507e-02</td>
      <td>Ndst4</td>
      <td>4.497158</td>
      <td>8.449939e-06</td>
      <td>3.032315e-05</td>
      <td>0</td>
      <td>chr4</td>
      <td>115106106</td>
      <td>115112630</td>
      <td>chr4</td>
      <td>114827772</td>
      <td>115113876</td>
      <td></td>
    </tr>
    <tr>
      <th>61</th>
      <td>Neuron_Granule_DG</td>
      <td>chr1_50860999_50861403</td>
      <td>4.356917</td>
      <td>1.071190e-04</td>
      <td>1.251507e-02</td>
      <td>Tmeff2</td>
      <td>3.399986</td>
      <td>7.231398e-04</td>
      <td>1.952972e-03</td>
      <td>66120</td>
      <td></td>
      <td></td>
      <td></td>
      <td>chr2</td>
      <td>191949045</td>
      <td>192194933</td>
      <td></td>
    </tr>
  </tbody>
</table>
<p>62 rows × 17 columns</p>
</div></div>
</div>
<p>Save the file if needed.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Mouse-Cortex.h5mu&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="K562HCT116_SP1.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tutorial: K562 HCT116 SP1 single-cell calling cards data.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorials</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js?v=0b546569"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../_static/scripts/furo.js?v=5fa4622c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>